Search.setIndex({"docnames": ["api_reference/assessment/index", "api_reference/assessment/summary_stats", "api_reference/assessment/validity_checks", "api_reference/index", "api_reference/resources/config", "api_reference/resources/grch37/gnomad", "api_reference/resources/grch37/gnomad_ld", "api_reference/resources/grch37/index", "api_reference/resources/grch37/reference_data", "api_reference/resources/grch38/gnomad", "api_reference/resources/grch38/index", "api_reference/resources/grch38/reference_data", "api_reference/resources/import_resources", "api_reference/resources/index", "api_reference/resources/resource_utils", "api_reference/sample_qc/ancestry", "api_reference/sample_qc/filtering", "api_reference/sample_qc/index", "api_reference/sample_qc/pipeline", "api_reference/sample_qc/platform", "api_reference/sample_qc/relatedness", "api_reference/sample_qc/sex", "api_reference/utils/annotations", "api_reference/utils/constraint", "api_reference/utils/file_utils", "api_reference/utils/filtering", "api_reference/utils/gen_stats", "api_reference/utils/index", "api_reference/utils/intervals", "api_reference/utils/liftover", "api_reference/utils/parse", "api_reference/utils/plotting", "api_reference/utils/reference_genome", "api_reference/utils/release", "api_reference/utils/slack", "api_reference/utils/sparse_mt", "api_reference/utils/transcript_annotation", "api_reference/utils/vcf", "api_reference/utils/vep", "api_reference/variant_qc/evaluation", "api_reference/variant_qc/index", "api_reference/variant_qc/ld", "api_reference/variant_qc/pipeline", "api_reference/variant_qc/random_forest", "api_reference/variant_qc/training", "examples/index", "examples/vep", "getting_started", "index", "resource_sources"], "filenames": ["api_reference/assessment/index.rst", "api_reference/assessment/summary_stats.rst", "api_reference/assessment/validity_checks.rst", "api_reference/index.rst", "api_reference/resources/config.rst", "api_reference/resources/grch37/gnomad.rst", "api_reference/resources/grch37/gnomad_ld.rst", "api_reference/resources/grch37/index.rst", "api_reference/resources/grch37/reference_data.rst", "api_reference/resources/grch38/gnomad.rst", "api_reference/resources/grch38/index.rst", "api_reference/resources/grch38/reference_data.rst", "api_reference/resources/import_resources.rst", "api_reference/resources/index.rst", "api_reference/resources/resource_utils.rst", "api_reference/sample_qc/ancestry.rst", "api_reference/sample_qc/filtering.rst", "api_reference/sample_qc/index.rst", "api_reference/sample_qc/pipeline.rst", "api_reference/sample_qc/platform.rst", "api_reference/sample_qc/relatedness.rst", "api_reference/sample_qc/sex.rst", "api_reference/utils/annotations.rst", "api_reference/utils/constraint.rst", "api_reference/utils/file_utils.rst", "api_reference/utils/filtering.rst", "api_reference/utils/gen_stats.rst", "api_reference/utils/index.rst", "api_reference/utils/intervals.rst", "api_reference/utils/liftover.rst", "api_reference/utils/parse.rst", "api_reference/utils/plotting.rst", "api_reference/utils/reference_genome.rst", "api_reference/utils/release.rst", "api_reference/utils/slack.rst", "api_reference/utils/sparse_mt.rst", "api_reference/utils/transcript_annotation.rst", "api_reference/utils/vcf.rst", "api_reference/utils/vep.rst", "api_reference/variant_qc/evaluation.rst", "api_reference/variant_qc/index.rst", "api_reference/variant_qc/ld.rst", "api_reference/variant_qc/pipeline.rst", "api_reference/variant_qc/random_forest.rst", "api_reference/variant_qc/training.rst", "examples/index.rst", "examples/vep.rst", "getting_started.rst", "index.rst", "resource_sources.rst"], "titles": ["gnomad.assessment", "gnomad.assessment.summary_stats", "gnomad.assessment.validity_checks", "gnomad", "gnomad.resources.config", "gnomad.resources.grch37.gnomad", "gnomad.resources.grch37.gnomad_ld", "gnomad.resources.grch37", "gnomad.resources.grch37.reference_data", "gnomad.resources.grch38.gnomad", "gnomad.resources.grch38", "gnomad.resources.grch38.reference_data", "gnomad.resources.import_resources", "gnomad.resources", "gnomad.resources.resource_utils", "gnomad.sample_qc.ancestry", "gnomad.sample_qc.filtering", "gnomad.sample_qc", "gnomad.sample_qc.pipeline", "gnomad.sample_qc.platform", "gnomad.sample_qc.relatedness", "gnomad.sample_qc.sex", "gnomad.utils.annotations", "gnomad.utils.constraint", "gnomad.utils.file_utils", "gnomad.utils.filtering", "gnomad.utils.gen_stats", "gnomad.utils", "gnomad.utils.intervals", "gnomad.utils.liftover", "gnomad.utils.parse", "gnomad.utils.plotting", "gnomad.utils.reference_genome", "gnomad.utils.release", "gnomad.utils.slack", "gnomad.utils.sparse_mt", "gnomad.utils.transcript_annotation", "gnomad.utils.vcf", "gnomad.utils.vep", "gnomad.variant_qc.evaluation", "gnomad.variant_qc", "gnomad.variant_qc.ld", "gnomad.variant_qc.pipeline", "gnomad.variant_qc.random_forest", "gnomad.variant_qc.training", "Examples", "Variant Effect Predictor (VEP)", "Getting Started", "gnomad", "Resource Sources"], "terms": {"summary_stat": [0, 3, 48], "freq_bin_expr": [0, 1], "get_summary_counts_dict": [0, 1], "get_summary_ac_dict": [0, 1], "get_summary_count": [0, 1], "get_an_criteria": [0, 1], "get_tx_expression_expr": [0, 1], "get_summary_stats_variant_filter_expr": [0, 1], "get_summary_stats_csq_filter_expr": [0, 1], "generate_filter_combin": [0, 1], "get_summary_stats_filter_group_meta": [0, 1], "default_generate_gene_lof_matrix": [0, 1], "get_het_hom_summary_dict": [0, 1], "default_generate_gene_lof_summari": [0, 1], "validity_check": [0, 3, 48], "generic_field_check": [0, 2], "make_filters_expr_dict": [0, 2], "make_group_sum_expr_dict": [0, 2], "compare_row_count": [0, 2], "summarize_variant_filt": [0, 2], "generic_field_check_loop": [0, 2], "compare_subset_freq": [0, 2], "sum_group_callstat": [0, 2], "summarize_vari": [0, 2], "check_raw_and_adj_callstat": [0, 2], "check_sex_chr_metr": [0, 2], "compute_missing": [0, 2], "vcf_field_check": [0, 2], "check_global_and_row_annot_length": [0, 2], "pprint_global_ann": [0, 2], "validate_release_t": [0, 2], "count_vep_annotated_variants_per_interv": [0, 2], "check_missingness_of_struct": [0, 2], "flatten_missingness_struct": [0, 2], "unfurl_array_annot": [0, 2], "freq_expr": [1, 22, 23], "index": [1, 2, 16, 20, 22, 23, 25, 29, 35, 37, 38, 41, 43], "0": [1, 2, 8, 11, 14, 15, 16, 18, 19, 20, 22, 23, 25, 29, 31, 33, 36, 37, 38, 39, 41, 42, 43, 44, 47], "ac_cutoff": 1, "ac0": [1, 2, 37], "1": [1, 2, 4, 5, 8, 9, 11, 15, 16, 18, 19, 20, 22, 23, 24, 25, 26, 30, 31, 35, 36, 37, 39, 41, 42, 44, 47], "singleton": [1, 22, 23, 37, 39, 42], "2": [1, 8, 11, 15, 16, 18, 20, 22, 23, 25, 36, 37, 39, 42, 43, 47], "doubleton": [1, 22, 37], "af_cutoff": 1, "0001": [1, 22], "01": [1, 22], "001": [1, 18, 22, 23, 43], "10": [1, 9, 11, 15, 16, 18, 19, 20, 22, 31, 35, 36, 38], "upper_af": 1, "95": [1, 22, 23], "sourc": [1, 2, 4, 5, 6, 8, 9, 11, 12, 14, 15, 16, 18, 19, 20, 21, 22, 23, 24, 25, 26, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 41, 42, 43, 44, 48], "return": [1, 2, 4, 5, 6, 8, 9, 11, 12, 14, 15, 16, 18, 19, 20, 21, 22, 23, 24, 25, 26, 28, 29, 30, 31, 32, 33, 35, 36, 37, 38, 39, 41, 42, 43, 44], "frequenc": [1, 2, 9, 15, 18, 20, 22, 23, 25, 33, 37], "string": [1, 2, 9, 12, 14, 15, 20, 21, 22, 23, 24, 25, 30, 33, 37, 38, 41], "annot": [1, 2, 3, 8, 9, 11, 14, 15, 16, 18, 19, 20, 21, 23, 25, 27, 29, 33, 35, 36, 37, 38, 39, 41, 42, 43, 44, 46, 48], "base": [1, 14, 15, 18, 19, 20, 21, 22, 23, 25, 28, 32, 33, 35, 36, 37, 38, 39, 42, 43], "input": [1, 2, 9, 12, 15, 16, 18, 19, 20, 21, 22, 23, 24, 25, 26, 28, 29, 32, 35, 37, 38, 39, 41, 42, 43, 44], "ac": [1, 2, 9, 20, 22, 23, 25, 35, 37, 41, 42], "af": [1, 2, 9, 18, 22, 23, 25, 37], "default": [1, 2, 4, 5, 8, 9, 11, 14, 15, 16, 18, 19, 20, 21, 22, 23, 24, 25, 26, 31, 32, 33, 35, 36, 37, 38, 42, 43, 49], "i": [1, 2, 4, 5, 8, 9, 11, 12, 14, 15, 16, 18, 19, 20, 21, 22, 23, 24, 25, 30, 31, 32, 33, 35, 36, 37, 38, 39, 42, 43, 44, 46, 49], "becaus": [1, 16, 20, 22, 23, 36, 39], "function": [1, 2, 14, 15, 18, 20, 21, 22, 23, 25, 29, 30, 35, 36, 37, 38, 39, 42, 43, 44, 48, 49], "assum": [1, 16, 18, 20, 21, 22, 23, 25, 26, 28, 29, 30, 38], "wa": [1, 12, 20, 22, 23, 37, 38, 42, 43, 47], "calcul": [1, 2, 9, 15, 16, 18, 20, 22, 23, 24, 26, 35, 36, 37, 39], "annotate_freq": [1, 22, 23, 27], "from": [1, 2, 4, 5, 8, 11, 12, 14, 15, 16, 18, 20, 21, 22, 23, 24, 25, 29, 30, 31, 32, 35, 36, 37, 38, 39, 41, 42, 43, 44, 46, 47, 49], "all": [1, 2, 9, 12, 14, 15, 16, 19, 20, 21, 22, 23, 24, 25, 28, 31, 32, 35, 36, 37, 38, 39, 41, 42, 43, 44, 49], "pop": [1, 2, 6, 9, 10, 15, 22, 23, 25, 31, 33, 37, 41], "adj": [1, 2, 9, 18, 22, 23, 25, 33, 37, 42], "genotyp": [1, 9, 16, 18, 20, 21, 22, 23, 25, 35, 37, 39], "onli": [1, 2, 9, 12, 15, 16, 18, 19, 20, 21, 22, 23, 24, 25, 32, 35, 36, 37, 38, 39, 42, 43, 46], "paramet": [1, 2, 5, 6, 8, 9, 11, 12, 14, 15, 16, 18, 19, 20, 21, 22, 23, 24, 25, 26, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 41, 42, 43, 44], "union": [1, 2, 4, 8, 11, 15, 16, 18, 20, 21, 22, 23, 24, 25, 28, 29, 30, 31, 32, 34, 35, 36, 37, 38], "structexpress": [1, 2, 16, 20, 21, 22, 23, 25, 26, 29, 30, 35, 36, 38], "arrayexpress": [1, 15, 20, 22, 23, 25, 26, 29, 35, 38], "arrai": [1, 2, 8, 11, 15, 22, 23, 25, 26, 28, 32, 33, 35, 36, 37, 38], "struct": [1, 2, 9, 16, 18, 21, 22, 23, 25, 26, 29, 30, 31, 35, 36, 37, 38, 42, 43], "contain": [1, 2, 5, 9, 14, 15, 16, 19, 20, 21, 22, 23, 24, 25, 29, 30, 31, 32, 33, 35, 36, 37, 38, 39, 42, 43, 44, 48], "inform": [1, 9, 14, 16, 22, 23, 24, 36, 37, 41, 43, 47], "int": [1, 2, 5, 12, 15, 16, 18, 19, 20, 21, 22, 23, 24, 25, 28, 30, 31, 33, 35, 36, 37, 39, 42, 43], "which": [1, 2, 12, 15, 16, 18, 20, 21, 22, 23, 25, 29, 35, 36, 37, 38, 39, 41, 42, 46], "us": [1, 2, 4, 9, 14, 15, 16, 18, 19, 20, 21, 22, 23, 24, 25, 29, 31, 33, 34, 35, 36, 37, 38, 39, 42, 43, 44, 46, 47, 49], "option": [1, 2, 8, 9, 11, 12, 14, 15, 16, 18, 19, 20, 21, 22, 23, 24, 25, 30, 32, 33, 34, 35, 36, 37, 38, 39, 42, 43, 44], "list": [1, 2, 9, 12, 14, 15, 16, 18, 19, 20, 22, 23, 24, 25, 28, 30, 31, 32, 33, 35, 36, 37, 38, 42, 43, 49], "tupl": [1, 12, 15, 16, 19, 20, 21, 22, 23, 24, 25, 29, 35, 36, 41, 42, 43], "str": [1, 2, 4, 5, 6, 9, 12, 14, 15, 16, 18, 19, 20, 21, 22, 23, 24, 25, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 41, 42, 43], "cutoff": [1, 2, 11, 18, 20, 21, 23, 37], "bin": [1, 2, 22, 31, 37, 39, 42], "float": [1, 2, 15, 16, 18, 19, 20, 21, 22, 23, 24, 25, 26, 31, 37, 42, 43, 44], "none": [1, 2, 4, 8, 9, 11, 12, 14, 15, 16, 18, 19, 20, 21, 22, 23, 24, 25, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 41, 42, 43, 44], "upper": [1, 2, 16, 20, 21, 23, 31], "type": [1, 2, 4, 5, 6, 8, 9, 11, 12, 14, 15, 16, 18, 19, 20, 21, 22, 23, 24, 25, 26, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 42, 43, 44], "stringexpress": [1, 20, 21, 22, 23, 35, 36, 38], "name": [1, 2, 4, 9, 12, 14, 16, 18, 22, 23, 24, 25, 33, 35, 36, 37, 38, 42, 43, 46, 47, 49], "locus_expr": [1, 20, 21, 22, 25], "allele_expr": 1, "lof_expr": 1, "no_lof_flags_expr": 1, "most_severe_csq_expr": 1, "prefix_str": 1, "dictionari": [1, 2, 9, 14, 16, 20, 22, 23, 25, 33, 35, 37, 38, 39, 42, 43], "count": [1, 2, 20, 22, 23, 25, 31, 37, 39, 42, 43], "multipl": [1, 2, 15, 20, 24, 25, 26, 31, 38, 39, 49], "variant": [1, 2, 5, 9, 18, 20, 22, 23, 24, 25, 26, 30, 35, 36, 37, 38, 39, 41, 42, 43, 44, 45, 47, 48], "categori": [1, 20, 21, 23], "ar": [1, 2, 8, 9, 11, 12, 14, 15, 16, 18, 19, 20, 21, 22, 23, 24, 25, 28, 31, 32, 33, 35, 36, 37, 38, 39, 42, 43, 44, 47], "number": [1, 2, 9, 15, 16, 18, 19, 20, 21, 22, 23, 24, 25, 28, 35, 36, 37, 39, 41, 42, 43, 48], "indel": [1, 8, 11, 16, 20, 22, 23, 26, 35, 37, 39, 42, 44], "snv": [1, 5, 16, 18, 20, 22, 26, 37, 39, 42], "lof": [1, 23, 36, 38], "pass": [1, 2, 9, 14, 15, 18, 22, 25, 35, 37, 38, 41, 42], "lofte": [1, 23, 38], "without": [1, 20, 22, 23, 25], "ani": [1, 2, 8, 14, 15, 16, 20, 22, 23, 24, 25, 35, 38, 39, 42, 43], "flg": 1, "other": [1, 2, 14, 15, 16, 18, 19, 22, 23, 25, 31, 37, 38, 46], "splice": [1, 36, 37, 38], "o": [1, 38], "fail": [1, 2, 20, 24, 29], "missens": [1, 23, 37], "synonym": [1, 23, 38], "autosom": [1, 11, 15, 16, 19, 20, 22, 23, 25, 26, 35, 37, 42], "allosom": 1, "two": [1, 2, 5, 20, 21, 22, 23, 36, 37, 41, 42], "multi": [1, 16, 18, 20, 22, 26, 35, 37, 44], "allel": [1, 2, 9, 15, 16, 18, 19, 20, 22, 23, 25, 26, 29, 30, 32, 33, 35, 36, 37, 38, 39, 41, 42, 44], "have": [1, 2, 15, 16, 19, 20, 22, 23, 24, 35, 38, 39, 42], "been": [1, 20, 22, 38, 39], "split": [1, 9, 18, 20, 21, 22, 23, 25, 35, 36, 37], "locusexpress": [1, 20, 21, 22, 25, 29, 32], "booleanexpress": [1, 2, 16, 20, 22, 23, 25, 35, 38, 39, 42, 44], "indic": [1, 2, 6, 16, 20, 22, 23, 24, 33, 35, 36, 38, 42, 43, 44], "whether": [1, 2, 15, 16, 18, 20, 22, 23, 24, 25, 29, 31, 35, 36, 37, 38, 39, 43], "ha": [1, 2, 16, 18, 19, 20, 21, 22, 25, 30, 35, 36, 38, 39, 43, 44], "flag": [1, 2, 22, 23, 25, 38, 46, 47], "most": [1, 22, 25, 35, 38, 46, 47], "sever": [1, 23, 38, 46], "consequ": [1, 25, 36, 37, 38, 46], "desir": [1, 2, 9, 19, 20, 22, 24, 25, 29, 35, 37, 44], "prefix": [1, 2, 12, 16, 21, 34, 37, 43], "empti": [1, 2, 5, 23, 31, 37, 38, 43], "dict": [1, 2, 12, 14, 16, 20, 22, 23, 25, 31, 33, 35, 37, 39, 42, 43], "int64express": [1, 2, 22, 23], "per": [1, 2, 22, 23, 25, 35, 36, 37, 38, 42], "ac_expr": 1, "total": [1, 2, 22, 23, 28, 35, 37], "ht": [1, 2, 8, 9, 11, 14, 15, 16, 19, 20, 21, 22, 23, 24, 25, 29, 31, 35, 36, 37, 38, 39, 42, 43, 44, 47], "freq_field": 1, "freq": [1, 2, 9, 22, 23, 25, 33, 37], "filter_field": 1, "filter": [1, 2, 3, 17, 18, 20, 22, 23, 27, 32, 33, 35, 36, 37, 38, 41, 42, 48], "filter_decoi": [1, 18, 25], "fals": [1, 2, 9, 14, 16, 18, 20, 22, 23, 24, 25, 28, 31, 32, 35, 36, 37, 38, 39, 42, 43, 44], "canonical_onli": 1, "true": [1, 2, 14, 15, 16, 18, 19, 20, 22, 23, 24, 25, 29, 31, 32, 35, 36, 37, 38, 39, 42, 44, 46], "mane_select_onli": 1, "gener": [1, 2, 9, 12, 14, 15, 20, 22, 23, 28, 29, 30, 35, 36, 37, 38, 42], "summari": [1, 2, 38], "across": [1, 2, 5, 9, 22, 23, 29, 35, 36, 37], "includ": [1, 2, 8, 9, 11, 16, 18, 20, 21, 22, 23, 25, 32, 34, 35, 36, 37, 38, 42, 43, 46], "also": [1, 2, 9, 14, 15, 16, 18, 20, 22, 23, 25, 43, 44], "tabl": [1, 2, 5, 8, 9, 11, 14, 15, 16, 18, 19, 20, 21, 22, 23, 24, 25, 29, 31, 32, 35, 37, 38, 39, 41, 42, 43, 44, 46, 47], "": [1, 2, 4, 5, 9, 14, 16, 18, 20, 21, 22, 23, 25, 28, 30, 34, 35, 36, 37, 38, 39, 42, 43, 46, 49], "global": [1, 2, 5, 9, 20, 22, 23, 33, 35, 36, 37, 38, 39, 42, 43], "befor": [1, 9, 18, 20, 22, 23, 24, 35, 36, 37, 39, 42], "out": [1, 16, 18, 20, 22, 23, 24, 35, 36, 38, 42, 44], "low": [1, 20, 23, 25, 35, 37, 38], "confid": [1, 8, 11, 20, 23, 25, 37, 39, 42], "region": [1, 2, 18, 20, 22, 25, 36, 37], "canon": [1, 23, 37, 38], "transcript": [1, 5, 23, 25, 36, 37, 38], "mane": [1, 37, 38], "select": [1, 9, 12, 14, 16, 22, 23, 24, 37, 38], "vep": [1, 2, 3, 23, 25, 27, 36, 41, 45, 48], "multiallel": [1, 20, 23, 37], "bi": [1, 16, 18, 19, 20, 22, 25, 26, 39, 42], "field": [1, 2, 9, 14, 16, 18, 19, 20, 21, 22, 23, 25, 35, 36, 37, 38, 39, 42, 43, 46], "bool": [1, 2, 9, 14, 15, 16, 18, 19, 20, 22, 23, 24, 25, 28, 29, 31, 32, 35, 36, 37, 38, 39, 42, 43], "decoi": [1, 18, 22, 25, 37], "group": [1, 2, 9, 10, 16, 20, 21, 22, 23, 27, 33, 35, 36, 37, 39, 42, 48], "aggreg": [1, 2, 16, 22, 23, 31, 35, 36, 39, 42], "mt": [1, 14, 15, 18, 19, 20, 21, 22, 24, 25, 31, 35, 36, 38, 39, 42, 43], "samples_by_sex": 1, "meta_root": 1, "meta": [1, 22, 23, 25], "sex_field": 1, "sex_imput": 1, "sex_karyotyp": [1, 18, 21], "xy_str": [1, 22], "xy": [1, 2, 9, 18, 20, 21, 22, 33, 35, 37], "xx_str": 1, "xx": [1, 2, 9, 18, 20, 21, 22, 25, 33, 35, 37], "freq_index": 1, "an_proportion_cutoff": 1, "8": [1, 15, 18, 24, 36, 37], "criteria": [1, 18, 20, 22, 23, 25, 38], "sampl": [1, 2, 9, 15, 16, 18, 19, 20, 21, 22, 23, 25, 33, 35, 36, 37, 39, 42, 43, 44], "an": [1, 2, 4, 9, 14, 15, 16, 18, 19, 20, 21, 22, 23, 24, 25, 26, 28, 31, 32, 34, 35, 36, 37, 38, 39, 42, 43, 44, 47, 49], "proxi": 1, "call": [1, 2, 15, 16, 18, 19, 20, 22, 35, 37, 38], "rate": [1, 18, 19, 23], "matrixt": [1, 2, 14, 15, 16, 18, 19, 20, 21, 22, 23, 24, 25, 29, 31, 35, 36, 37, 38, 39, 42, 43], "valu": [1, 2, 4, 8, 12, 14, 15, 16, 18, 19, 20, 21, 22, 23, 24, 25, 26, 31, 33, 35, 36, 37, 38, 39, 41, 43], "each": [1, 2, 9, 15, 16, 19, 20, 22, 23, 24, 25, 28, 31, 33, 35, 36, 37, 38, 39, 42, 43, 46], "sex": [1, 2, 3, 9, 10, 16, 17, 18, 20, 22, 25, 26, 27, 33, 35, 37, 48], "kei": [1, 2, 12, 15, 16, 19, 20, 22, 23, 25, 33, 35, 36, 37, 38, 39], "metadata": [1, 22, 23, 25, 31, 33], "assign": [1, 15, 19, 21, 23, 37, 39], "defualt": 1, "mark": 1, "proport": [1, 5, 15, 23, 36], "key_expr": 1, "tx_ht": [1, 36], "csq_expr": [1, 38], "gene_field": 1, "ensg": 1, "csq_field": [1, 38], "csq": [1, 38], "tx_struct": 1, "tx_annot": 1, "pull": [1, 23, 36], "appropri": 1, "express": [1, 2, 5, 8, 11, 14, 15, 16, 20, 21, 22, 23, 25, 26, 29, 32, 35, 36, 39, 42, 43, 44], "given": [1, 2, 6, 15, 18, 20, 21, 22, 23, 26, 32, 35, 36, 37, 38, 41, 43, 44], "specif": [1, 2, 5, 9, 14, 16, 20, 22, 23, 25, 31, 35, 37, 39, 43, 46], "locu": [1, 2, 9, 19, 20, 21, 22, 25, 29, 30, 32, 35, 36, 37, 41], "provid": [1, 2, 4, 5, 8, 9, 11, 15, 16, 20, 22, 23, 25, 30, 31, 32, 35, 39, 44, 49], "both": [1, 20, 21, 22, 24, 25, 29, 35, 37, 39, 44], "row_key_expr": 1, "search": 1, "gene": [1, 2, 5, 9, 23, 25, 36, 38, 41], "id": [1, 15, 16, 20, 22, 23, 24, 25, 36, 37], "most_severe_consequ": [1, 23, 36, 38], "float64express": [1, 20, 22, 23, 26], "t": [1, 2, 14, 16, 18, 20, 22, 23, 24, 25, 29, 31, 35, 36, 37, 38, 41, 43], "filter_lcr": [1, 18, 25], "filter_expr": [1, 23, 25], "grpmax_expr": 1, "max_af": [1, 23], "max_grpmax": 1, "min_an_proport": 1, "stat": [1, 2, 18, 20, 21, 22, 23, 24, 26, 35, 42, 44], "The": [1, 2, 4, 8, 11, 12, 14, 15, 16, 19, 20, 21, 22, 23, 24, 25, 26, 30, 31, 33, 35, 36, 37, 38, 39, 42, 43, 44, 49], "possibl": [1, 2, 22, 23, 32, 37], "no_lcr": 1, "variant_qc_pass": 1, "grpmax": [1, 2, 33, 37], "min_an": 1, "an_proport": 1, "setexpress": [1, 20, 25], "max": [1, 20, 22, 37, 42], "maximum": [1, 12, 20, 22, 23, 25, 37, 43], "can": [1, 2, 5, 12, 15, 16, 19, 20, 21, 22, 23, 24, 25, 28, 31, 32, 35, 38, 39, 42, 49], "singl": [1, 5, 9, 15, 16, 18, 20, 21, 24, 25, 28, 35, 36, 38, 39], "genet": [1, 2, 20, 22, 23, 37], "ancestri": [1, 2, 3, 9, 17, 22, 23, 37, 48], "minimum": [1, 15, 18, 20, 22, 23, 25, 36, 37], "lof_csq_set": [1, 27, 38], "lof_label_set": 1, "lof_no_flag": 1, "lof_any_flag": 1, "additional_csq_set": 1, "additional_csq": 1, "requir": [1, 14, 18, 20, 22, 23, 24, 25, 29, 35, 36, 38, 39, 47], "request": [1, 2, 14, 22, 23, 24, 25, 36, 39, 46, 49], "no_lof_flag": [1, 38], "loftee_no_flag": 1, "loftee_with_flag": 1, "loftee_label": [1, 27, 38], "lof_label": 1, "csq_set": 1, "set": [1, 2, 4, 12, 14, 15, 16, 18, 19, 20, 21, 22, 23, 24, 25, 28, 32, 33, 35, 36, 37, 38, 39, 42, 43, 44, 49], "thi": [1, 2, 8, 11, 12, 14, 15, 16, 18, 19, 20, 21, 22, 23, 24, 25, 28, 29, 30, 32, 35, 36, 37, 38, 39, 41, 42, 43, 44, 46, 47, 48, 49], "label": [1, 2, 9, 15, 20, 22, 23, 31, 33, 37, 38, 43, 44], "addit": [1, 14, 15, 18, 20, 22, 23, 25, 31, 35, 36, 38, 39, 42, 43, 47], "keep": [1, 18, 20, 22, 23, 25, 35, 36, 37, 38, 43], "combo": [1, 37], "combo_opt": 1, "combin": [1, 2, 16, 20, 21, 22, 23, 24, 25, 33, 35, 36, 37, 38, 39, 42], "exampl": [1, 2, 4, 12, 20, 22, 23, 25, 26, 31, 33, 34, 36, 37, 39, 42, 43, 44, 47, 48, 49], "pass_filt": 1, "captur": [1, 23, 37], "ukb": 1, "broad": [1, 35, 37], "output": [1, 14, 15, 16, 20, 21, 22, 23, 24, 31, 35, 38, 43], "suppli": [1, 2, 15, 18, 21, 22, 23, 37, 38], "all_sum_stat_filt": 1, "common_filter_combo": 1, "common_filter_overrid": 1, "lof_filter_combo": 1, "lof_filter_overrid": 1, "filter_key_renam": 1, "variou": [1, 2], "statist": [1, 9, 22, 23, 35, 37, 38, 42], "It": [1, 21, 22, 23, 35, 36, 37, 38], "ensur": [1, 9, 20, 22, 37], "common": [1, 20, 29], "loss": [1, 23, 38], "variant_qc": [1, 3, 48], "remov": [1, 2, 9, 20, 22, 24, 25, 31, 35, 38], "lead": [1, 36], "renam": 1, "after": [1, 16, 22, 23, 34, 35, 38, 42, 43], "creat": [1, 16, 18, 20, 21, 22, 23, 25, 30, 32, 33, 35, 36, 37, 38, 39, 42, 43, 46, 47], "follow": [1, 2, 8, 11, 12, 16, 18, 20, 21, 22, 23, 24, 25, 35, 36, 37, 38, 39, 42, 44], "lof_csq": 1, "stop_gain": [1, 36, 38], "loftee_hc": 1, "loftee_flag": 1, "hc": [1, 38], "with_flag": 1, "overrid": 1, "lof_combo": 1, "final": [1, 21, 39], "high_expression_cutoff": 1, "9": [1, 14, 15, 36, 37], "low_expression_cutoff": 1, "missense_vari": [1, 36, 38], "synonymous_vari": [1, 36, 38], "all_transcript": 1, "filter_an": 1, "filter_to_rar": 1, "pre_lofte": 1, "frameshift_vari": [1, 36, 38], "splice_acceptor_vari": [1, 36, 38], "splice_donor_vari": [1, 36, 38], "remove_ultra_common": 1, "matrix": [1, 6, 20, 22, 35, 41, 42, 43], "metric": [1, 2, 16, 20, 22, 23, 37], "level": [1, 8, 11, 14, 22, 23, 35, 36, 37, 38], "mean": [1, 8, 18, 20, 21, 22, 23, 35, 36, 37, 49], "consid": [1, 2, 19, 20, 22, 23, 25], "highli": [1, 20, 22], "lowli": 1, "instead": [1, 9, 16, 18, 21, 22, 23, 25, 35, 36, 37, 38, 49], "just": [1, 25], "rare": 1, "5": [1, 2, 11, 15, 18, 20, 21, 22, 23, 35, 36, 37, 42, 43], "ultra": 1, "defined_sites_expr": 1, "num_homs_expr": 1, "num_hets_expr": 1, "pop_expr": [1, 22], "site": [1, 2, 5, 8, 9, 11, 14, 18, 19, 22, 23, 35, 37, 39, 42, 46], "defin": [1, 2, 16, 20, 22, 23, 25, 35, 36, 37, 38, 39, 42, 44, 49], "heterozyg": [1, 2, 20, 23, 37], "homozyg": [1, 2, 18, 20, 21, 23, 37], "popul": [1, 2, 5, 6, 9, 15, 20, 22, 23, 25, 33, 37, 41], "collapse_indel": 1, "tx": 1, "pop_field": 1, "filter_lofte": 1, "p": [1, 18, 20, 22, 31, 37], "haplotyp": [1, 37], "carri": 1, "put": 1, "plof": [1, 23], "observ": [1, 18, 20, 22, 23, 25, 37], "expect": [1, 2, 14, 16, 18, 20, 22, 23, 25, 35, 36, 37, 42], "oe": [1, 23], "ratio": [1, 20, 22, 23, 37, 42, 44], "abov": [1, 18, 20, 21, 22, 23, 25, 35], "were": [1, 11, 20, 22, 23, 25, 26, 37, 42, 43], "If": [1, 2, 4, 14, 15, 16, 18, 19, 20, 21, 22, 23, 24, 25, 28, 30, 31, 32, 34, 35, 36, 37, 38, 39, 42, 43, 44, 49], "row": [1, 2, 18, 20, 22, 23, 25, 31, 35, 36, 37, 38, 39, 42, 43, 44, 46], "collaps": [1, 23], "data": [1, 2, 4, 5, 8, 9, 11, 14, 15, 16, 18, 20, 22, 25, 26, 29, 31, 33, 34, 35, 36, 37, 38, 39, 42, 43, 46, 47, 49], "top": [1, 2, 36, 38], "indici": 1, "het": [1, 18, 20, 21, 22, 35], "hom": [1, 18, 21, 22], "check_descript": 2, "display_field": 2, "cond_expr": 2, "verbos": 2, "show_percent_sit": 2, "n_fail": 2, "ht_count": 2, "check": [2, 14, 22, 23, 24, 38, 43], "logic": [2, 37], "condit": [2, 20], "involv": 2, "hail": [2, 4, 8, 9, 11, 14, 15, 16, 20, 21, 22, 23, 24, 29, 30, 31, 32, 38, 39, 42, 43, 46, 47, 48, 49], "when": [2, 14, 15, 16, 18, 19, 20, 21, 22, 23, 24, 25, 33, 35, 36, 37, 38, 46, 47], "absent": 2, "print": [2, 24, 43], "result": [2, 15, 16, 19, 20, 22, 31, 35, 38, 42, 43], "stdout": 2, "displai": [2, 25, 31], "percent": [2, 24, 37], "either": [2, 9, 15, 20, 22, 23, 24, 25, 49], "previous": [2, 15, 20, 22], "comput": [2, 15, 16, 18, 19, 20, 22, 23, 26, 35, 37, 39, 42, 43], "match": [2, 19, 29, 36, 38, 43], "otherwis": [2, 4, 15, 20, 22, 23, 28, 30, 37, 49], "opposit": 2, "should": [2, 15, 16, 19, 20, 22, 23, 24, 25, 31, 32, 35, 36, 37, 38, 39, 42, 43, 44], "never": 2, "same": [2, 16, 19, 20, 22, 23, 24, 31, 35, 38, 39, 43], "e": [2, 5, 9, 12, 14, 15, 16, 19, 20, 22, 23, 25, 30, 33, 35, 37, 38, 39, 43], "g": [2, 9, 14, 16, 19, 20, 22, 23, 24, 25, 29, 30, 31, 33, 35, 37, 38, 39, 43, 49], "instanc": [2, 16], "where": [2, 15, 20, 22, 23, 25, 32, 33, 35, 36, 37, 38, 39, 43], "raw": [2, 9, 20, 22, 23, 33, 37, 42], "less": [2, 8, 16, 22, 23, 37], "than": [2, 8, 15, 16, 20, 21, 22, 23, 24, 25, 37, 44], "greater": [2, 8, 16, 23, 24, 25, 44], "equal": [2, 8, 22, 23, 25, 38, 39, 42], "describ": [2, 11, 20, 22, 23, 37, 47], "being": [2, 18, 20, 22, 23, 37, 49], "messag": [2, 24, 34], "case": [2, 14, 15, 20, 22, 25, 35, 39], "failur": [2, 24], "troubleshoot": 2, "purpos": [2, 16], "refer": [2, 16, 18, 20, 22, 23, 24, 25, 29, 30, 32, 35, 36, 37, 38, 41, 46, 47, 48], "show": [2, 22], "percentag": 2, "obtain": [2, 9, 23, 44], "within": [2, 16, 20, 22, 23, 31, 35, 37, 38, 39, 41], "perform": [2, 9, 20, 22, 23, 35, 37, 38, 39, 42], "extra_filter_check": 2, "variant_filter_field": 2, "rf": [2, 15, 37, 42, 43], "make": [2, 24, 33, 37, 42], "measur": 2, "under": [2, 20, 22, 23, 36, 37], "vari": [2, 23], "interest": [2, 24, 38], "fraction": [2, 18, 21, 35], "due": 2, "inbreed": [2, 18, 22, 37], "coeffici": [2, 18, 20, 22, 37], "examin": [2, 23], "extra": [2, 23, 38], "filtrat": [2, 42], "vqsr": [2, 35, 37, 42], "as_vqsr": [2, 37], "subset": [2, 9, 10, 16, 22, 23, 25, 33, 35, 37, 39, 44], "label_group": [2, 37], "sort_ord": [2, 27, 33, 37], "downsampl": [2, 9, 10, 22, 23, 25, 33, 37], "popmax": [2, 9, 22, 33, 37], "gen_anc": [2, 23, 33, 37], "subpop": [2, 5, 9, 25, 33, 37], "delimit": [2, 14, 33, 37, 38], "metric_first_field": 2, "nhomalt": [2, 37], "sum": [2, 20, 22, 23, 35, 36, 37], "specifi": [2, 9, 16, 18, 19, 20, 22, 23, 24, 25, 31, 35, 36, 37, 38, 42, 43], "compar": [2, 11, 39], "version": [2, 5, 8, 9, 11, 12, 14, 15, 20, 22, 25, 35, 36, 38, 47], "For": [2, 4, 11, 14, 16, 20, 22, 23, 24, 25, 33, 36, 37, 38, 39, 42, 47, 49], "subset1": 2, "consist": [2, 35], "pop1": 2, "pop2": 2, "pop3": 2, "info": [2, 9, 18, 20, 22, 35, 37, 38, 42, 43], "entri": [2, 16, 18, 19, 21, 22, 25, 27, 33, 35, 36, 37], "afr": [2, 9, 22, 23, 25, 33, 37], "nfe": [2, 5, 9, 22, 23, 33, 37, 41], "amr": [2, 9, 33, 37], "order": [2, 9, 18, 20, 22, 23, 25, 36, 37, 38, 39, 42, 43], "sort": [2, 9, 28, 33, 37, 38], "preced": [2, 37], "hgdp": [2, 9], "ht1": 2, "ht2": 2, "first": [2, 18, 20, 22, 23, 24, 35, 38, 42], "second": [2, 20, 22, 23, 35, 37], "problematic_region": 2, "lcr": [2, 18, 22, 25, 37], "segdup": [2, 18, 22, 25, 37], "nonpar": [2, 22, 37], "single_filter_count": 2, "site_gt_check_expr": 2, "n_row": 2, "50": [2, 9, 16, 35], "n_col": 2, "140": [2, 31], "summar": [2, 36], "problemat": [2, 18, 22], "run": [2, 4, 15, 16, 18, 19, 20, 22, 23, 34, 42, 43, 49], "explod": [2, 20, 23, 38], "column": [2, 15, 20, 22, 31, 35, 36, 39, 42, 43], "give": [2, 16, 20, 22, 44], "supplement": 2, "boolean": [2, 9, 16, 20, 22, 23, 24, 25, 39], "typic": [2, 14, 15, 18, 20, 22, 23, 24, 35], "log": [2, 14, 22, 31, 37, 48], "how": [2, 20, 25, 37], "mani": [2, 20], "monoallel": [2, 37], "100": [2, 9, 12, 15, 20, 23, 35, 36, 39, 42], "field_check_expr": 2, "loop": [2, 20], "through": [2, 22, 49], "allow": [2, 36, 49], "onc": [2, 9, 20, 22, 24], "oppos": 2, "dure": [2, 9, 16, 37, 49], "everi": [2, 35], "whose": [2, 20], "valid": 2, "callset": [2, 23, 37, 39], "eg": 2, "overal": [2, 37], "non_v2": [2, 9], "between": [2, 5, 8, 20, 23, 25, 36, 37, 38, 39, 41, 42], "entir": 2, "asj": [2, 9, 33, 37], "ea": [2, 5, 9, 33, 37], "fin": [2, 9, 22, 33, 37], "mid": [2, 9, 20, 33, 37], "remain": [2, 9, 20, 33, 37], "sa": [2, 9, 33, 37], "additional_subsets_and_pop": 2, "gen_anc_label_nam": 2, "present": [2, 12, 15, 20, 22, 23, 24, 30, 33, 35, 37, 38, 39, 42], "test": [2, 18, 22, 37, 42, 43], "current_major_releas": [2, 33], "exom": [2, 5, 9, 18, 19, 23, 24, 25, 29, 33, 37, 47], "callstat": [2, 22, 37], "doe": [2, 22, 25, 37, 44], "store": [2, 12, 14, 15, 16, 18, 22, 35, 37, 38, 43, 46], "male": [2, 20, 21, 37], "denot": [2, 14], "expected_contig": 2, "get": [2, 4, 5, 6, 18, 20, 21, 22, 23, 24, 35, 36, 38, 39, 41, 42, 48], "chromosom": [2, 9, 16, 18, 20, 21, 22, 23, 25, 26, 30, 35, 37], "contig": [2, 5, 9, 25, 28, 30, 32, 35, 37, 39], "found": [2, 22, 23, 25, 33, 43], "unexpect": 2, "nhomalt_metr": 2, "neg": [2, 20, 22, 23, 37, 42, 44], "correspond": [2, 8, 11, 16, 20, 22, 23, 33, 35, 36, 37, 38, 41, 44], "must": [2, 8, 11, 14, 18, 20, 21, 22, 23, 25, 35, 37, 38, 39, 42, 46, 49], "altern": [2, 18, 20, 21, 22, 23, 30, 35, 37, 49], "info_metr": 2, "That": 2, "chry": [2, 9, 22, 37], "na": [2, 14, 37, 43], "non": [2, 14, 16, 19, 20, 21, 22, 23, 25, 26, 35, 36, 37, 43], "par": [2, 20, 21, 22, 25, 37], "chrx": [2, 11, 22, 37], "non_info_metr": 2, "n_site": 2, "missingness_threshold": 2, "amount": [2, 23, 35], "missing": 2, "sdout": 2, "exce": 2, "minu": 2, "header_dict": 2, "row_annot": 2, "entry_annot": 2, "hist": [2, 22, 27, 31, 37], "gq_hist_alt": [2, 37], "gq_hist_al": [2, 37], "dp_hist_alt": [2, 37], "dp_hist_al": [2, 37], "ab_hist_alt": [2, 37], "vcf": [2, 3, 5, 9, 14, 16, 22, 24, 27, 33, 35, 38, 46, 48], "descript": [2, 37, 38], "header": [2, 24, 25, 37, 38], "export": [2, 9, 22, 24, 33, 37, 38], "histogram": [2, 22, 31, 37], "row_to_globals_check": 2, "check_all_row": 2, "length": [2, 22, 23, 28, 37, 42], "associ": [2, 22, 32, 37], "pretti": 2, "sum_metr": 2, "sample_sum_sets_and_pop": 2, "summarize_variants_check": 2, "filters_check": 2, "raw_adj_check": 2, "subset_freq_check": 2, "samples_sum_check": 2, "sex_chr_check": 2, "missingness_check": 2, "pprint_glob": 2, "check_all_rows_in_row_to_global_check": 2, "batteri": 2, "statu": [2, 20, 22, 23, 25], "differ": [2, 16, 20, 22, 37, 38, 49], "partit": [2, 18, 23, 24, 35], "outlier": [2, 16, 23, 25], "subgroup": [2, 9, 37], "add": [2, 9, 22, 23, 25, 28, 29, 32, 36, 38, 39, 42, 46], "up": [2, 33, 46, 47], "supergroup": 2, "main": [2, 12, 13, 20, 22], "relev": [2, 16, 22, 23, 25, 37], "regardless": 2, "violat": 2, "method": [2, 9, 14, 16, 20, 22, 35, 37], "check_sex_chr_metricss": 2, "vep_ht": [2, 41], "interval_ht": [2, 35], "interv": [2, 3, 18, 19, 23, 24, 25, 27, 30, 32, 35, 36, 38, 39, 42, 43, 48], "transcript_consequ": [2, 23, 36, 38], "biotyp": [2, 38], "determin": [2, 4, 11, 16, 18, 20, 21, 22, 23, 35, 37, 39, 42, 44, 49], "protein": [2, 23, 25, 36, 37, 38], "code": [2, 20, 22, 23, 25, 34, 36, 38], "gene_stable_id": 2, "ensembl": [2, 37, 38, 46], "releas": [2, 3, 5, 9, 16, 27, 36, 37, 38, 48], "ad": [2, 9, 20, 21, 22, 23, 25, 28, 29, 32, 36, 37, 38, 39, 42, 43], "n_total_vari": 2, "n_pcg_variant": 2, "struct_expr": 2, "recurs": [2, 22], "miss": [2, 14, 20, 21, 22, 23, 25, 31, 36, 43], "standalon": 2, "nest": [2, 23, 36], "element": [2, 15, 22, 23, 24, 25, 33, 36, 38, 39], "append": [2, 20, 37], "map": [2, 9, 22, 29, 31, 35, 37], "repres": [2, 21, 22, 23, 35, 37], "missingness_struct": 2, "flatten": [2, 20], "evalu": [2, 3, 20, 25, 40, 42, 48], "indexed_array_annot": 2, "faf": [2, 22, 33, 37], "faf_index_dict": 2, "freq_index_dict": [2, 22], "unfurl": 2, "structur": [2, 9, 22, 37], "int32": [2, 22, 35, 37], "float64": [2, 18, 22, 23, 36], "homozygote_count": [2, 22, 25], "int64": [2, 18, 37], "wherea": 2, "sample_qc": [3, 48], "pipelin": [3, 17, 23, 24, 40, 43, 48], "platform": [3, 17, 20, 22, 23, 35, 48], "related": [3, 17, 42, 48], "util": [3, 42, 46, 48], "constraint": [3, 5, 7, 9, 10, 27, 48], "file_util": [3, 27, 48], "gen_stat": [3, 27, 48], "liftov": [3, 5, 7, 11, 27, 48], "pars": [3, 25, 27, 48], "plot": [3, 27, 48], "reference_genom": [3, 9, 24, 25, 27, 30, 41, 46, 48], "slack": [3, 27, 48], "sparse_mt": [3, 27, 48], "transcript_annot": [3, 27, 48], "assess": [3, 48], "resourc": [3, 22, 25, 36, 37, 38, 47, 48], "config": [3, 13, 38, 48, 49], "grch37": [3, 9, 11, 12, 13, 25, 29, 30, 32, 38, 41, 46, 47, 48], "grch38": [3, 13, 18, 24, 25, 29, 30, 32, 37, 38, 47, 48], "import_resourc": [3, 13, 14, 48], "resource_util": [3, 13, 48], "ld": [3, 6, 18, 40, 48], "random_forest": [3, 40, 48], "train": [3, 15, 37, 40, 42, 43, 48], "configur": [4, 38, 46, 48], "load": [4, 5, 15, 20, 29, 36, 43, 49], "class": [4, 14, 34], "gnomadpublicresourcesourc": [4, 13, 49], "modul": [4, 12, 20, 23, 30, 36, 42], "qualnam": 4, "start": [4, 22, 23, 25, 28, 30, 31, 36, 37, 38, 46, 48, 49], "boundari": [4, 22, 31, 35, 38], "public": [4, 5, 9, 14, 47, 49], "google_cloud_public_dataset": 4, "googl": [4, 24, 46, 47, 49], "cloud": [4, 24, 47, 49], "dataset": [4, 9, 16, 20, 22, 23, 24, 25, 33, 35, 37, 46, 49], "registry_of_open_data_on_aw": 4, "registri": 4, "open": [4, 49], "aw": 4, "azure_open_dataset": 4, "azur": [4, 49], "get_default_public_resource_sourc": [4, 13], "gnomad_default_public_resource_sourc": [4, 49], "environ": [4, 38, 48], "variabl": [4, 12, 20, 38, 49], "spark": [4, 43, 49], "hdinsight": [4, 49], "cluster": [4, 9, 19, 38, 46, 47, 49], "public_releas": [5, 7, 9, 10, 47], "data_typ": [5, 9, 29], "retriev": [5, 9, 22], "publicli": [5, 9], "One": [5, 8, 9, 11, 22, 25, 29, 36, 37], "genom": [5, 8, 9, 11, 23, 24, 25, 29, 30, 32, 35, 36, 37, 38, 41, 42, 46, 47, 48], "versionedtableresourc": [5, 9, 13, 14, 38], "coverag": [5, 7, 9, 10, 18, 23, 35, 37], "38": 5, "v2": [5, 29, 47], "public_pca_load": [5, 7], "tableresourc": [5, 13, 14], "pca": [5, 15, 19], "gnomadpublictableresourc": [5, 6, 9, 13, 14], "release_vcf_path": [5, 7, 9, 10], "20": [5, 9, 11, 18, 20, 22, 35, 37, 38], "y": [5, 18, 20, 21, 22, 23, 25, 31, 35, 37], "path": [5, 9, 12, 14, 16, 18, 24, 25, 29, 31, 34, 38, 42, 43, 46], "pext": [5, 7, 9, 10, 36], "pext_typ": [5, 9], "base_level": [5, 9], "annotation_level": [5, 9], "mnv": [5, 7], "distanc": [5, 16, 38], "multinucleotid": 5, "browser_gen": [5, 7, 9, 10], "browser": [5, 9, 36], "ld_matrix": [6, 7], "gnomadpublicblockmatrixresourc": [6, 13, 14], "ld_index": [6, 7, 41], "ld_score": [6, 7], "score": [6, 8, 11, 15, 16, 19, 23, 25, 37, 38, 39, 42], "gnomad_ld": [7, 13], "reference_data": [7, 10, 13, 25], "get_truth_ht": [7, 8, 10, 11], "transform_grch37_methyl": [7, 8, 23], "latest": [8, 11, 38], "truth": [8, 11, 39, 42], "hapmap": [8, 11, 42], "kgp_omni": [8, 11], "1000": [8, 9, 11, 22, 25, 30, 38, 42], "intersect": [8, 11], "onni": [8, 11], "5m": [8, 11], "kgp_phase_1_hc": [8, 11], "high": [8, 9, 11, 18, 20, 22, 23, 25, 37, 38, 39, 42], "genonm": [8, 11], "mill": [8, 11, 42], "devin": [8, 11], "A": [8, 11, 12, 14, 15, 16, 19, 20, 22, 23, 24, 25, 35, 36, 37, 38, 41, 42, 43, 44], "popular": [8, 11], "methylation_expr": [8, 11, 23], "transform": [8, 11, 19, 22, 23, 35], "methyl": [8, 11, 23], "scale": [8, 11, 20, 22, 23, 26, 35, 37], "rang": [8, 9, 11, 23, 25, 35, 37], "we": [8, 20, 23, 35, 37, 39], "convert": [8, 15, 21, 22, 23, 34, 36, 37], "6": [8, 11, 18, 21, 36, 37], "numericexpress": [8, 11, 16, 18, 20, 21, 22, 23, 25, 26, 35, 36, 39], "v3": [9, 37, 47], "non_topm": 9, "non_canc": 9, "controls_and_biobank": 9, "non_neuro": 9, "tgp": 9, "v4": [9, 23, 37, 38], "non_ukb": [9, 33], "tgp_raw_ac_esn_xx": 9, "qualiti": [9, 20, 22, 23, 35, 37], "stratifi": [9, 20, 22, 23, 35, 37, 39, 42], "ami": [9, 37], "oth": [9, 15, 37], "cohorts_with_pop_stored_as_subpop": [9, 10], "broken": [9, 16, 43], "down": [9, 43, 47], "known": [9, 15, 18, 43], "tgp_pop": [9, 10], "esn": 9, "pur": 9, "pjl": 9, "clm": 9, "jpt": 9, "chb": 9, "stu": 9, "itu": 9, "tsi": 9, "mxl": 9, "ceu": 9, "msl": 9, "yri": 9, "beb": 9, "khv": 9, "cdx": 9, "lwk": 9, "acb": 9, "asw": 9, "ib": 9, "gbr": [9, 37], "pel": 9, "gih": 9, "ch": 9, "gwd": 9, "project": [9, 14, 15, 22, 24, 37, 46, 48, 49], "1kg": 9, "hgdp_pop": [9, 10], "japanes": [9, 37], "papuanhighland": 9, "papuansepik": 9, "adygei": 9, "orcadian": 9, "biaka": 9, "yakut": 9, "han": [9, 25], "northernhan": 9, "uygur": 9, "miao": 9, "mongolian": 9, "balochi": 9, "bedouin": 9, "russian": 9, "daur": 9, "pima": 9, "hezhen": 9, "sindhi": 9, "yi": 9, "oroqen": 9, "san": 9, "tuscan": 9, "tu": 9, "palestinian": 9, "tujia": 9, "druze": 9, "pathan": 9, "basqu": 9, "makrani": 9, "bergamoitalian": 9, "naxi": 9, "karitiana": 9, "sardinian": 9, "mbuti": 9, "mozabit": 9, "yoruba": 9, "lahu": 9, "dai": 9, "cambodian": 9, "bougainvil": 9, "french": 9, "brahui": 9, "hazara": 9, "bantusouthafrica": 9, "surui": 9, "mandenka": 9, "kalash": 9, "xibo": 9, "colombian": 9, "bantukenya": 9, "she": 9, "burusho": 9, "maya": 9, "human": [9, 23, 37], "divers": 9, "tgp_pop_nam": [9, 10], "african": [9, 22, 37], "caribbean": 9, "american": [9, 22, 37], "bengali": 9, "chines": 9, "utah": 9, "resid": 9, "european": [9, 37], "southern": [9, 37], "esan": 9, "finnish": [9, 37], "british": [9, 37], "gujarati": 9, "gambian": 9, "iberian": 9, "indian": 9, "telugu": 9, "kinh": 9, "luhya": 9, "mend": 9, "mexican": 9, "peruvian": 9, "punjabi": 9, "puerto": 9, "rican": 9, "sri": 9, "lankan": 9, "tamil": 9, "toscani": 9, "pops_to_remove_for_popmax": [9, 10], "200": [9, 22, 35], "500": [9, 16, 19, 38, 42, 43], "2000": [9, 25, 30], "5000": [9, 23, 35], "10000": [9, 23], "15000": 9, "20000": 9, "25000": 9, "30000": 9, "40000": 9, "50000": 9, "60000": 9, "70000": 9, "75000": 9, "80000": 9, "85000": 9, "90000": 9, "95000": 9, "100000": 9, "110000": 9, "120000": 9, "200000": 9, "500000": 9, "joint": [9, 37], "all_sites_an": [9, 10], "coverage_tsv_path": [9, 10], "chr20": [9, 18, 35, 37], "chr1": [9, 25, 30, 36, 37], "add_grpmaxfaf95_v4": [9, 10], "grpmaxfaf95": 9, "popmax_popul": 9, "jointgrpmaxfaf95": 9, "fafmax": [9, 37], "joint_fafmax": 9, "gnomad_gk": [9, 10], "locus_interv": 9, "by_ancestry_group": 9, "by_sex": [9, 22], "vrs_onli": 9, "custom_ht": 9, "skip_checkpoint": 9, "skip_coverag": 9, "custom_coverage_ht": 9, "gk": [9, 22], "intervalexpress": [9, 24, 25, 28, 30, 32], "hl": [9, 14, 16, 18, 20, 22, 23, 25, 31, 35, 36, 38, 41, 42, 46], "6424776": 9, "6461367": 9, "read": [9, 14, 22, 24, 35, 37, 46, 49], "cohort": [9, 22, 33, 37], "vr": [9, 22, 37], "what": [9, 23, 42], "would": [9, 20, 23, 33, 36, 43], "skip": [9, 22, 23], "checkpoint": [9, 18, 39], "mai": [9, 23, 25, 38, 47], "larg": [9, 16, 20, 22, 25, 35, 41], "reduc": 9, "copi": [9, 22], "custom": [9, 48], "browser_vari": [9, 10], "transform_grch38_methyl": [10, 11, 23], "15": [11, 20, 35], "chen": 11, "et": [11, 20, 23], "al": [11, 20, 23], "http": [11, 20, 22, 24, 36], "www": 11, "biorxiv": 11, "org": [11, 22], "content": [11, 34, 43], "1101": 11, "2022": [11, 47], "03": 11, "485034v2": 11, "full": [11, 20, 22, 23, 39], "12": [11, 23], "directli": 11, "analyz": 11, "separ": [11, 15, 16, 22, 23, 26, 37, 39], "rel": [11, 20, 21, 37, 43], "translat": [11, 38, 48], "correl": [11, 41], "propos": 11, "3": [11, 20, 22, 23, 25, 31, 36, 39, 43], "4": [11, 15, 16, 20, 22, 33, 36, 37, 39], "get_module_importable_resourc": [12, 13], "take": [12, 15, 20, 22, 23, 25, 31, 42], "import": [12, 14, 25, 42, 43, 46, 47, 49], "import_func": [12, 14], "produc": [12, 20, 22], "resource_nam": 12, "grch37_resourc": 12, "baseresourc": [12, 13, 14], "get_resources_descript": [12, 13], "width": [12, 31], "along": [12, 20], "thei": [12, 20, 22, 24, 26, 35, 36, 37, 39], "line": [12, 20, 24, 31, 36], "arg": 12, "gnomad_public_bucket": [13, 14], "matrixtableresourc": [13, 14], "variantdatasetresourc": [13, 14], "pedigreeresourc": [13, 14], "blockmatrixresourc": [13, 14], "expressionresourc": [13, 14], "baseversionedresourc": [13, 14], "versionedmatrixtableresourc": [13, 14], "versionedvariantdatasetresourc": [13, 14], "versionedpedigreeresourc": [13, 14], "versionedblockmatrixresourc": [13, 14], "resourcenotavail": [13, 14], "gnomadpublicresourc": [13, 14], "gnomadpublicmatrixtableresourc": [13, 14], "gnomadpublicpedigreeresourc": [13, 14], "dataexcept": [13, 14], "import_sites_vcf": [13, 14], "import_gencod": [13, 14], "pai": [14, 24, 46, 49], "bucket": [14, 24, 31, 46, 49], "stage": [14, 35, 43], "legaci": 14, "one": [14, 15, 16, 18, 20, 21, 22, 23, 25, 29, 31, 32, 35, 36, 37, 38, 39, 42, 49], "readm": 14, "text": [14, 20, 24, 34, 37], "file": [14, 22, 24, 25, 29, 31, 34, 37, 38, 43], "product": 14, "team": 14, "write": [14, 16, 24], "sync": 14, "gcp": 14, "import_arg": 14, "abstract": 14, "need": [14, 16, 22, 23, 35, 37, 39], "kept": [14, 18, 20, 25, 35, 36], "track": 14, "callabl": [14, 15, 22, 23, 25, 31, 35, 36, 38], "kwarg": [14, 25, 31, 35, 36, 38], "expected_file_extens": 14, "extens": [14, 24], "doesn": [14, 18, 24, 29, 35], "end": [14, 22, 23, 24, 25, 28, 30, 31, 35, 37], "warn": [14, 16, 41], "properti": 14, "overwrit": [14, 24, 43], "its": [14, 22, 39], "exist": [14, 18, 23, 24, 37, 43], "destin": [14, 29], "underli": 14, "accept": 14, "depend": [14, 15, 20, 23], "force_import": 14, "read_arg": 14, "forc": 14, "even": [14, 16, 22], "alreadi": [14, 22, 23, 24, 28, 29, 38], "argument": [14, 22, 25, 32, 35, 36, 43, 47], "read_tabl": 14, "noth": [14, 43], "read_matrix_t": 14, "overwritten": 14, "variantdataset": [14, 16, 18, 20, 22, 24, 25, 35], "vd": [14, 18, 24, 25, 35], "read_vd": 14, "quant_pheno": 14, "pedigre": [14, 20], "fam": 14, "ped": 14, "phenotyp": 14, "interpret": [14, 25, 37, 38], "quantit": 14, "regex": 14, "control": [14, 35], "numer": [14, 16, 20, 22, 25, 31, 43], "treat": [14, 35, 38], "famili": [14, 20, 37, 42], "import_fam": [14, 20], "current": [14, 18, 21, 22, 25, 37, 38], "implement": [14, 18, 22], "blockmatrix": [14, 41], "bm": [14, 24, 41], "he": [14, 24], "experiment": [14, 20, 42], "read_express": 14, "write_express": 14, "default_vers": 14, "attribut": [14, 37], "those": [14, 15, 18, 20, 22, 23, 35, 38, 44], "In": [14, 15, 18, 22, 23, 37, 39, 42], "resource_class": 14, "alia": 14, "param": [14, 25], "except": [14, 24, 31, 34], "rais": [14, 16, 20, 22, 24, 34], "avail": [14, 22, 49], "is_resource_avail": 14, "publish": 14, "gtf_path": 14, "gencod": [14, 23, 25, 36], "gtf": 14, "pc_project": [15, 17], "loadings_ht": [15, 19], "loading_loc": 15, "af_loc": 15, "pca_af": 15, "pre": [15, 49], "pc": [15, 16, 19, 20], "locat": [15, 23, 37, 43], "apply_onnx_classification_model": [15, 17], "data_pd": 15, "fit": [15, 23], "appli": [15, 18, 20, 22, 23, 25, 35, 38, 42, 43], "onnx": 15, "classif": 15, "model": [15, 18, 21, 23, 37, 42, 43], "panda": 15, "datafram": [15, 31, 43], "classifi": 15, "modelproto": 15, "ndarrai": 15, "probabl": [15, 20, 23, 37, 38, 43], "apply_sklearn_classification_model": [15, 17], "sklearn": 15, "convert_sklearn_rf_to_onnx": [15, 17], "target_opset": 15, "random": [15, 20, 37, 39, 42, 43], "forest": [15, 37, 42, 43], "target": [15, 23], "opset": 15, "assign_population_pc": [15, 17], "pop_pca_scor": 15, "pc_col": 15, "known_col": 15, "known_pop": 15, "seed": 15, "42": [15, 20], "prop_train": 15, "n_estim": 15, "min_prob": 15, "output_col": 15, "missing_label": 15, "pc_expr": 15, "convert_model_func": 15, "apply_model_func": 15, "As": 15, "hwe_normalized_pca": 15, "integ": [15, 22, 23, 37], "you": [15, 22, 36, 38, 47], "expand_pd_array_col": 15, "expand": 15, "pc1": 15, "pc2": 15, "pc4": 15, "pc5": 15, "previou": 15, "randomforestclassifi": [15, 43], "tree": [15, 16, 42, 43], "belong": [15, 22, 36], "smaller": [15, 20], "format": [15, 18, 20, 22, 23, 25, 30, 36, 37, 38, 47], "convers": [15, 37, 43], "work": [15, 16, 20, 38], "imput": [15, 18, 19, 35, 43], "run_pca_with_rel": [15, 17], "qc_mt": 15, "related_samples_to_drop": 15, "additional_samples_to_drop": 15, "n_pc": [15, 16, 19], "autosomes_onli": [15, 19, 42], "exclud": [15, 18, 20, 22, 23, 32, 35, 36, 44], "relat": [15, 20, 37, 42], "space": [15, 20, 37], "qc": [15, 16, 18, 37, 39, 42], "drop": [15, 20, 35, 37, 43], "analysi": [15, 23], "eigenvalu": [15, 19], "compute_qc_metrics_residu": [16, 17], "pc_score": 16, "qc_metric": 16, "use_pc_squar": 16, "regression_sample_inclusion_expr": 16, "strata": [16, 20, 22, 25, 35, 43], "residu": 16, "regress": [16, 23], "alwai": [16, 37], "arraynumericexpress": [16, 22, 23, 26, 35], "stratif": [16, 20, 22, 42, 43], "These": [16, 20, 21, 25, 38], "discret": 16, "compute_stratified_metrics_filt": [16, 17], "lower_threshold": 16, "upper_threshold": 16, "metric_threshold": 16, "filter_nam": 16, "qc_metrics_filt": 16, "comparison_sample_expr": 16, "median": [16, 23, 26, 35, 36, 43], "mad": [16, 26], "lower": [16, 20, 21, 22, 23, 31, 38], "threshold": [16, 18, 19, 20, 22, 23], "expr": [16, 22, 39], "critic": 16, "more": [16, 20, 21, 22, 23, 24, 25, 35, 36, 37, 38, 39, 47], "collectionexpress": 16, "well": [16, 22, 23, 29, 35, 37], "determine_nearest_neighbor": [16, 17], "compute_stratified_sample_qc": [16, 17], "mtd": [16, 18, 20, 25, 35], "tmp_ht_prefix": 16, "gt_col": 16, "merg": [16, 22, 26, 28], "overlap": [16, 19, 26, 28, 35], "v": [16, 20, 35, 37, 39], "intermedi": [16, 39], "recommend": [16, 21, 24, 35], "larger": [16, 20, 37], "gt": [16, 18, 20, 22, 25, 35, 37, 39], "strat": 16, "merge_sample_qc_expr": [16, 17], "sample_qc_expr": 16, "note": [16, 18, 20, 21, 22, 23, 26, 35, 37, 38, 39, 46, 47], "regard": 16, "dp_stat": 16, "gq_stat": 16, "n": [16, 20, 31, 32, 42, 43], "stdev": [16, 21], "n_call": [16, 18], "veri": 16, "standard": [16, 18, 20, 21, 23, 35, 37], "gatk": [16, 18, 22, 35], "essenti": 16, "dp": [16, 18, 20, 22, 35, 37], "gq": [16, 18, 20, 22, 37], "do": [16, 20, 22, 24, 35, 37], "assumpt": [16, 20], "some": [16, 20, 22, 23, 34, 35, 44], "shouldn": 16, "matter": 16, "too": [16, 18, 20, 25, 42], "much": [16, 20, 24], "scores_expr": 16, "n_neighbor": 16, "n_job": 16, "add_neighbor_dist": 16, "distance_metr": 16, "euclidean": 16, "use_approxim": 16, "n_tree": 16, "nearest": 16, "neighbor": 16, "limit": 16, "nearest_neighbor": 16, "nearest_neighbor_dist": 16, "identifi": [16, 20, 34, 37, 46], "thread": 16, "find": [16, 38], "cpu": 16, "head": 16, "node": 16, "scikit": 16, "learn": 16, "cityblock": 16, "cosin": 16, "haversin": 16, "l1": 16, "l2": 16, "manhattan": 16, "annoi": 16, "angular": 16, "ham": 16, "dot": 16, "packag": [16, 46, 47, 48, 49], "approxim": [16, 20, 35, 37, 39, 43], "nearestneighbor": 16, "faster": [16, 25, 37, 38], "000": 16, "approach": [16, 20], "build": [16, 22, 23, 25, 29, 30, 36, 37, 38], "time": [16, 20, 23, 35, 47], "affect": [16, 20], "size": [16, 18, 22, 23, 24, 31, 33, 35, 36, 42], "accur": [16, 35], "filter_rows_for_qc": [17, 18], "get_qc_mt": [17, 18], "infer_sex_karyotyp": [17, 18], "annotate_sex": [17, 18], "compute_callrate_mt": [17, 19], "run_platform_pca": [17, 19], "assign_platform_from_pc": [17, 19], "unrel": [17, 20, 37, 42], "second_degree_rel": [17, 20], "parent_child": [17, 20], "sibl": [17, 20, 37, 42], "duplicate_or_twin": [17, 20], "ambiguous_relationship": [17, 20], "get_duplicated_sampl": [17, 20], "get_duplicated_samples_ht": [17, 20], "explode_duplicate_samples_ht": [17, 20], "get_relationship_expr": [17, 20], "get_slope_int_relationship_expr": [17, 20], "infer_famili": [17, 20], "create_fake_pedigre": [17, 20], "compute_related_samples_to_drop": [17, 20], "filter_to_trio": [17, 20], "generate_trio_stats_expr": [17, 20, 42], "generate_sib_stats_expr": [17, 20, 42], "calculate_de_novo_post_prob": [17, 20], "default_get_de_novo_expr": [17, 20], "adjusted_sex_ploidy_expr": [17, 21, 22], "adjust_sex_ploidi": [17, 21], "gaussian_mixture_model_karyotype_assign": [17, 21], "get_ploidy_cutoff": [17, 18, 21], "get_chr_x_hom_alt_cutoff": [17, 21], "get_sex_expr": [17, 18, 21], "min_af": 18, "min_callr": 18, "99": [18, 20, 22, 37], "min_inbreeding_coeff_threshold": 18, "min_hardy_weinberg_threshold": 18, "1e": [18, 22], "08": [18, 20], "apply_hard_filt": 18, "bi_allelic_onli": [18, 19, 42], "snv_onli": 18, "sites_callr": 18, "site_inbreeding_coeff": 18, "callrat": [18, 19], "taken": [18, 23, 25], "coeff": 18, "mq": [18, 35, 37, 42], "f": [18, 21, 22, 37, 42], "qd": [18, 37, 42], "best": 18, "practic": 18, "hard": [18, 20, 37], "Not": 18, "hw": 18, "gakt": 18, "60": [18, 23, 42], "30": [18, 22, 23, 35, 42], "adj_onli": 18, "ld_r2": 18, "filter_segdup": [18, 25], "filter_exome_low_coverage_region": [18, 25], "high_conf_region": [18, 25], "checkpoint_path": [18, 39], "n_partit": [18, 35], "block_siz": 18, "readi": [18, 37, 43], "outsid": [18, 25, 31, 37], "maf": 18, "ab": [18, 20, 22], "prune": [18, 44], "r2": 18, "segment": [18, 37], "duplic": [18, 20, 37], "persist": 18, "repartit": 18, "written": 18, "reread": 18, "new": [18, 24, 37], "block": [18, 22, 34, 35, 37, 41], "ploidy_ht": 18, "f_stat_cutoff": [18, 21], "use_gaussian_mixture_model": 18, "normal_ploidy_cutoff": [18, 21], "aneuploidy_cutoff": [18, 21], "chr_x_frac_hom_alt_expr": [18, 21], "normal_chr_x_hom_alt_cutoff": 18, "x_karyotyp": [18, 21], "y_karyotyp": [18, 21], "x": [18, 20, 21, 22, 23, 25, 35, 37], "ploidi": [18, 20, 21, 22, 35], "karyotyp": [18, 20, 21, 22, 35], "By": [18, 23, 25, 35, 36, 37, 38, 42], "roughli": [18, 21], "gaussian": [18, 21, 37], "mixtur": [18, 21, 37], "divid": [18, 21, 39], "below": [18, 20, 21, 23, 25, 37, 47], "deviat": [18, 21, 23, 26], "aneuploidi": [18, 21, 22], "alt": [18, 21, 22, 23, 30, 37], "is_spars": 18, "excluded_interv": [18, 32], "included_interv": 18, "normalization_contig": [18, 35], "sites_ht": [18, 35], "aaf_expr": 18, "gt_expr": [18, 21, 22, 25], "aaf_threshold": 18, "variants_only_x_ploidi": 18, "variants_only_y_ploidi": 18, "variants_filter_lcr": 18, "variants_filter_segdup": 18, "variants_filter_decoi": 18, "variants_snv_onli": 18, "coverage_mt": 18, "compute_x_frac_variants_hom_alt": 18, "compute_fstat": 18, "infer_karyotyp": 18, "heterozygos": [18, 22, 37], "_mean_dp": 18, "float32": 18, "over": [18, 20, 23, 24, 29, 35, 36, 37], "chrx_mean_dp": 18, "chry_mean_dp": 18, "chrx_ploidi": [18, 21], "chry_ploidi": [18, 21], "f_stat": [18, 21], "impute_sex": 18, "expected_hom": 18, "homozygot": [18, 23], "observed_hom": 18, "infer": [18, 20, 22, 29, 30], "spars": [18, 20, 25, 35, 37, 42], "normal": [18, 22, 26, 31, 35, 37], "prior": [18, 20, 22, 23, 29, 35], "depth": [18, 20, 22, 35, 37, 42, 43], "estim": [18, 23, 36, 37], "nucleotid": 18, "precomput": [18, 46], "intervals_ht": 19, "sequenc": [19, 29, 32, 37], "interval_info": 19, "callrate_mt": 19, "binarization_threshold": 19, "25": [19, 20, 35], "binzarization_threshold": 19, "scores_ht": 19, "platform_pca_scores_ht": 19, "pc_scores_ann": 19, "hdbscan_min_cluster_s": 19, "hdbscan_min_sampl": 19, "hbdscan": 19, "hdbscan": 19, "min_cluster_s": 19, "smallest": 19, "n_sampl": 19, "min_sampl": [19, 36], "qc_platform": 19, "represent": [20, 38], "pair": [20, 23, 25, 28, 41, 42], "individu": [20, 22, 37, 42], "2nd": [20, 42], "degre": 20, "user": [20, 34, 36], "parent": [20, 37], "child": [20, 37], "twin": 20, "who": 20, "ident": 20, "mz": 20, "ambigu": [20, 22], "relationship": [20, 42], "kinship": 20, "ibd": 20, "biolog": 20, "relationship_ht": 20, "i_col": [20, 42], "j_col": [20, 42], "j": [20, 42], "rel_col": 20, "extract": [20, 23, 36, 43], "ouput": 20, "pc_relat": 20, "1st": [20, 42], "duplicated_sampl": 20, "samples_rankings_ht": 20, "rank_ann": 20, "rank": [20, 37, 39], "better": [20, 35, 38], "samples_ranking_ht": 20, "dups_ht": 20, "so": [20, 21, 22, 23, 24, 38], "dup_filt": 20, "kin_expr": 20, "ibd0_expr": 20, "ibd1_expr": 20, "ibd2_expr": 20, "first_degree_kin_threshold": 20, "19": 20, "second_degree_min_kin": 20, "ibd0_0_max": 20, "025": 20, "ibd0_25_threshold": 20, "425": 20, "ibd1_0_threshold": 20, "ibd1_50_threshold": 20, "275": 20, "75": 20, "ibd1_100_min": 20, "ibd2_0_max": 20, "125": 20, "ibd2_25_threshold": 20, "ibd2_100_threshold": 20, "kin": 20, "ibdo": 20, "ibd1": 20, "ibd2": 20, "doc": 20, "html": 20, "highlight": 20, "idb2": 20, "min": [20, 42], "ibd0": 20, "share": [20, 34, 42], "corner": 20, "00": 20, "constant": [20, 25, 38, 42], "y_expr": 20, "parent_child_max_i": 20, "second_degree_sibling_lower_cutoff_slop": 20, "second_degree_sibling_lower_cutoff_intercept": 20, "second_degree_upper_sibling_lower_cutoff_slop": 20, "second_degree_upper_sibling_lower_cutoff_intercept": 20, "duplicate_twin_min_kin": 20, "duplicate_twin_ibd1_min": 20, "duplicate_twin_ibd1_max": 20, "slope": [20, 23], "intercept": [20, 23], "axi": [20, 31], "independ": [20, 22], "ibs0": 20, "ibs2": 20, "No": [20, 22], "els": [20, 38], "second_degree_sibling_lower_cutoff": 20, "second_degree_upper_sibling_lower_cutoff": 20, "met": [20, 23, 25], "08838835": 20, "bycroft": 20, "2018": 20, "theoret": 20, "distribut": [20, 22, 23, 26, 35, 37], "duplicate_samples_ht": 20, "relationship_col": [20, 42], "trio": [20, 42], "ok": [20, 42], "member": 20, "sib": 20, "complet": [20, 22, 23], "father": 20, "mother": 20, "femal": [20, 21, 22, 37], "is_femal": 20, "TO": 20, "won": 20, "exactli": 20, "contatin": 20, "sample_list": 20, "exclude_real_proband": 20, "max_tri": 20, "real_pedigre": 20, "sample_list_stratif": 20, "made": 20, "children": 20, "fake": 20, "proband": 20, "real": 20, "ones": [20, 22, 32], "cannot": 20, "try": 20, "bail": 20, "prevent": 20, "infinit": 20, "w": 20, "r": [20, 37, 41], "pick": 20, "chosen": [20, 38], "relatedness_ht": [20, 42], "rank_ht": 20, "kin_threshold": 20, "filtered_sampl": 20, "min_related_hard_filt": 20, "keep_sampl": 20, "keep_samples_when_rel": 20, "maxim": [20, 23], "prefer": [20, 23], "appear": [20, 35], "error": [20, 24, 25, 43], "among": 20, "don": [20, 23, 36], "fam_ht": 20, "filter_col": 20, "howev": [20, 25], "filter_sampl": 20, "trio_mt": 20, "transmitted_strata": 20, "de_novo_strata": 20, "ac_strata": 20, "proband_is_female_expr": 20, "wise": [20, 22, 37], "transmiss": 20, "transmit": [20, 37, 42], "de": [20, 42], "novo": [20, 42], "mutat": [20, 23], "suffix": [20, 37], "dens": [20, 35, 42], "deal": [20, 42], "densifi": [20, 35, 37, 42], "trio_matrix": [20, 42], "dnm": 20, "sib_ht": 20, "stata": 20, "col": [20, 22], "anoth": [20, 22, 29, 37], "proband_pl_expr": 20, "father_pl_expr": 20, "mother_pl_expr": 20, "diploid_expr": [20, 22], "hemi_x_expr": [20, 22], "hemi_y_expr": [20, 22], "freq_prior_expr": 20, "min_pop_prior": 20, "3333333333333333e": 20, "06": 20, "de_novo_prior": 20, "3333333333333334e": 20, "posterior": 20, "p_dn": 20, "likelihood": [20, 37], "adapt": 20, "kaitlin": 20, "samocha": 20, "caller": [20, 22], "de_novo": 20, "neither": [20, 23], "explicitli": 20, "document": [20, 22, 47], "hemizyg": [20, 22], "To": [20, 46, 49], "address": [20, 47], "equat": 20, "docstr": 20, "event": 20, "truli": 20, "heterozygot": 20, "p_": 20, "dn": 20, "frac": 20, "term": [20, 23, 37, 38], "least": [20, 22, 23, 25, 35], "bayesian": 20, "theorem": 20, "denomin": 20, "yield": 20, "cdot": 20, "hom_ref": 20, "report": [20, 37], "adjust": [20, 21, 22, 23, 35, 38], "why": [20, 23], "formula": 20, "hom_alt": 20, "literatur": 20, "7": [20, 36], "freq_prior": 20, "phred": [20, 22, 26, 35, 37], "diploid": [20, 22], "3e7": 20, "alleles_expr": [20, 22], "proband_expr": 20, "father_expr": 20, "mother_expr": 20, "is_xx_expr": [20, 22], "min_dp_ratio": 20, "min_gq": 20, "min_proband_ab": 20, "max_parent_ab": 20, "05": [20, 22, 23, 37], "min_de_novo_p": 20, "high_conf_dp_ratio": 20, "dp_threshold_snp": 20, "high_med_conf_ab": 20, "low_conf_ab": 20, "high_conf_p": 20, "med_conf_p": 20, "dr": 20, "OR": 20, "medium": 20, "slightli": 20, "origin": [20, 25, 29, 39, 42, 47], "design": 20, "fill": [20, 22, 31], "gap": 20, "meet": [20, 37], "unclassifi": 20, "simplifi": 20, "github": [20, 22], "com": [20, 22, 36], "ksamocha": 20, "de_novo_script": 20, "ignore_in_sample_allele_frequ": 20, "mode": [20, 31], "might": 20, "pl": [20, 37], "balanc": [20, 22, 37], "our": 20, "intention": 20, "save": [20, 31, 43], "storag": [20, 24], "reason": [20, 25], "wai": [20, 22, 25, 35, 38], "them": [20, 22, 35, 38], "or_els": 20, "biallel": [20, 39, 42], "snp": [20, 23, 37], "karyotype_expr": [21, 22], "xy_karyotype_str": [21, 22], "xx_karyotype_str": [21, 22], "haploid": [21, 22], "callexpress": [21, 22, 39], "sex_expr": [21, 22], "male_str": [21, 22], "female_str": 21, "point": [21, 23, 24], "chang": [21, 22, 23, 36, 48], "fix": [21, 23], "sex_ht": 21, "chrx_ploidy_expr": 21, "chry_ploidy_expr": 21, "karyotype_output_prefix": 21, "gmm": 21, "compon": [21, 37], "_x_karyotyp": 21, "karyotype_output_prefix_y_karyotyp": 21, "_karyotyp": 21, "_y_karyotyp": 21, "yy": 21, "group_by_expr": 21, "x_ploidy_cutoff": 21, "y_ploidy_cutoff": 21, "doubl": [21, 22], "tripl": 21, "cutoff_stdev": 21, "chr_x_ploidi": 21, "chr_y_ploidi": 21, "chr_x_frac_hom_alt_cutoff": 21, "x0": 21, "annotate_with_ht": [22, 27], "annotation_ht": 22, "annotate_col": 22, "filter_miss": 22, "join": [22, 23, 24, 33, 35, 43], "pop_max_expr": [22, 27], "freq_meta": [22, 23, 25, 33, 35], "pops_to_exclud": 22, "pop_label": 22, "about": [22, 37, 43, 44], "highest": 22, "calcluat": 22, "project_max_expr": [22, 27], "project_expr": 22, "n_project": 22, "largest": [22, 37], "Will": 22, "itself": 22, "ti": [22, 44, 46], "guarante": 22, "projectmax": 22, "faf_expr": [22, 27], "faf_threshold": 22, "see": [22, 24, 25, 36, 38, 39, 47, 49], "cardiodb": 22, "allelefrequencyapp": 22, "agg": [22, 31, 36], "call_stat": 22, "bottleneck": 22, "consanguin": [22, 37], "gen_anc_faf_max_expr": [22, 27], "faf_meta": [22, 33], "faf95_max": 22, "faf95_max_gen_anc": 22, "faf99_max": 22, "faf99_max_gen_anc": 22, "faf95": [22, 37], "faf99": [22, 37], "qual_hist_expr": [22, 27], "gq_expr": 22, "dp_expr": 22, "ad_expr": 22, "adj_expr": 22, "ab_expr": 22, "split_adj_and_raw": 22, "here": [22, 23], "age_hists_expr": [22, 27], "age_expr": 22, "lowest_boundari": 22, "highest_boundari": 22, "80": 22, "n_bin": [22, 39, 42], "ag": [22, 37], "lowest": [22, 37], "younger": 22, "n_smaller": [22, 37], "older": 22, "n_larger": [22, 37], "age_hist_het": 22, "age_hist_hom": 22, "get_lowqual_expr": [22, 27], "qual_approx_expr": 22, "snv_phred_threshold": 22, "snv_phred_het_prior": 22, "indel_phred_threshold": 22, "indel_phred_het_prior": 22, "39": 22, "lowqual": [22, 35], "unsplit": 22, "qualapprox": [22, 35, 37], "as_qualapprox": [22, 37], "stringent": 22, "mix": [22, 37], "thu": 22, "certain": [22, 36, 38], "emiss": 22, "similar": [22, 23], "arraynumer": 22, "get_annotations_hist": [22, 27], "annotations_hist": 22, "log10_annot": 22, "json": [22, 24, 38, 43], "meric": 22, "create_frequency_bins_expr": [22, 27], "prepar": [22, 29, 36, 38], "qual": [22, 35, 37], "00005": 22, "0002": 22, "0005": 22, "002": 22, "005": 22, "02": 22, "annotate_and_index_source_mt_for_sex_ploidi": [22, 27], "downstream": [22, 25], "optim": 22, "get_is_haploid_expr": [22, 27], "in_non_par": 22, "x_nonpar": 22, "y_par": 22, "y_nonpar": 22, "get_gq_dp_adj_expr": [22, 27], "adj_gq": 22, "adj_dp": 22, "haploid_adj_dp": 22, "account": [22, 25], "int32express": [22, 23, 35, 37], "get_het_ab_adj_expr": [22, 27], "adj_ab": 22, "get_adj_expr": [22, 27], "annotate_adj": [22, 27], "add_variant_typ": [22, 27], "alt_allel": 22, "variant_typ": [22, 23, 37], "n_alt_allel": [22, 37], "annotate_allele_info": [22, 27], "allele_info": 22, "has_star": [22, 37], "star": [22, 25, 37], "allele_typ": [22, 37], "insert": [22, 37, 42], "delet": [22, 37, 42], "was_mix": [22, 37], "nonsplit_allel": 22, "annotation_type_is_numer": [22, 27], "annotation_type_in_vcf_info": [22, 27], "nativ": 22, "aren": 22, "bi_allelic_site_inbreeding_expr": [22, 27], "callstats_expr": 22, "inbreedingcoeff": [22, 37], "softwar": 22, "broadinstitut": 22, "articl": [22, 36], "php": 22, "8032": 22, "place": 22, "float32express": 22, "fs_from_sb": [22, 27], "sb": [22, 35, 37], "min_cell_count": 22, "min_count": 22, "min_p_valu": 22, "320": 22, "fisher": [22, 37], "strand": [22, 23, 37, 38], "side": [22, 28], "exact": [22, 37], "behavior": 22, "truncat": [22, 23, 31], "chi": 22, "squar": [22, 23, 37], "cell": [22, 36], "four": [22, 39], "forward": [22, 37], "revers": [22, 23, 37], "ref": [22, 23, 30, 32, 35, 37, 38], "fwd": 22, "rev": 22, "dimension": 22, "blob": 22, "master": 22, "src": 22, "java": 22, "hellbend": 22, "tool": [22, 31], "walker": 22, "fisherstrand": 22, "sq": 22, "fet": 22, "null": [22, 23], "sor_from_sb": [22, 27], "sor": [22, 37], "symmetr": [22, 37], "odd": [22, 37], "strandoddsratio": 22, "pab_max_expr": [22, 27], "la_expr": 22, "n_alleles_expr": 22, "binomi": [22, 37], "pab": 22, "local": [22, 24, 35, 38], "bi_allelic_expr": [22, 27], "unphase_call_expr": [22, 27], "call_expr": 22, "unphas": 22, "phase": [22, 37, 42], "region_flag_expr": [22, 27], "non_par": [22, 37], "prob_region": 22, "region_flag": 22, "hg38": [22, 37], "self": 22, "chain": [22, 29], "yet": 22, "loci": [22, 25], "occur": 22, "pseudoautosom": [22, 37], "missing_callstats_expr": [22, 27], "set_female_y_metrics_to_na_expr": [22, 27], "freq_meta_expr": [22, 23], "freq_index_dict_expr": 22, "dictexpress": [22, 23], "hemi_expr": [22, 27], "merge_freq_arrai": [22, 27], "farrai": 22, "fmeta": 22, "oper": [22, 25], "set_negatives_to_zero": 22, "count_arrai": 22, "freq1": 22, "freq2": 22, "fmeta1": 22, "fmeta2": 22, "diff": 22, "thier": 22, "subtract": 22, "primari": 22, "valueerror": 22, "descriptor": 22, "merge_histogram": [22, 27], "fashion": 22, "bin_edg": [22, 37], "bin_freq": 22, "edg": [22, 37], "subpop_expr": 22, "additional_strata_expr": 22, "downsampling_expr": 22, "ds_pop_count": 22, "entry_agg_func": [22, 35], "annotate_mt": 22, "freq_meta_sample_count": [22, 23, 25], "support": [22, 24, 32, 37, 38, 43, 44], "age_bin": 22, "replac": [22, 36, 43], "enough": 22, "randomli": 22, "higher": [22, 35, 36, 37, 38], "global_idx": 22, "pop_idx": 22, "annotate_downsampl": [22, 27], "adj_sampl": 22, "stratifict": 22, "sub": [22, 25, 42], "continent": 22, "build_freq_stratification_list": [22, 27], "primarili": 22, "generate_freq_group_membership_arrai": [22, 27], "strata_expr": [22, 25, 35], "remove_zero_sample_group": 22, "no_raw_group": 22, "group_membership": [22, 35], "pleas": [22, 47], "compute_freq_by_strata": [22, 27], "select_field": 22, "group_membership_includes_raw_group": 22, "like": [22, 25, 37], "still": [22, 44], "membership": [22, 35], "agg_by_strata": [22, 27], "group_membership_ht": [22, 35], "entry_agg_group_membership": [22, 35], "update_structured_annot": [22, 27], "annotation_update_expr": 22, "annotation_update_label": 22, "updat": [22, 37, 47], "fill_missing_key_combin": [22, 27], "fill_valu": 22, "key_valu": 22, "uniqu": [22, 37], "collect": [22, 23, 38], "unless": [22, 23, 35], "want": [22, 23, 36, 38, 39], "happen": 22, "python": [22, 24, 47], "parallel": 22, "key1": 22, "key2": 22, "b": 22, "tstruct": [22, 43], "tstr": 22, "tint32": 22, "filled_ht": 22, "missing_struct_expr": [22, 27], "dtype": 22, "add_gks_vr": [22, 27], "input_locu": 22, "input_vr": 22, "ga4gh": [22, 37], "conform": 22, "add_gks_va": [22, 27], "input_struct": 22, "label_nam": 22, "label_vers": 22, "ancestry_group": 22, "ancestry_groups_dict": 22, "va": 22, "schema": [22, 36], "subcohort": 22, "mean_depth": 22, "ancillaryresult": 22, "gks_va_freq_dict": 22, "focusallel": 22, "shorten": 22, "breakdown": 22, "get_copy_state_by_sex": [22, 27], "state": [22, 23, 24, 37], "karotyp": 22, "xxy": 22, "xyi": 22, "script": [23, 48], "coverage_cutoff": [23, 27], "differenti": 23, "calibr": 23, "variat": [23, 36], "get_mu_annotation_expr": [23, 27], "mutation_ht": 23, "mu_expr": 23, "mu_snp": 23, "context": [23, 25, 38], "methylation_level": 23, "avoid": 23, "unnecessari": 23, "shuffl": [23, 29, 39], "small": [23, 36], "annotate_with_mu": [23, 27], "mu_annot": 23, "count_variants_by_group": [23, 27], "count_singleton": 23, "count_downsampl": 23, "additional_group": 23, "partition_hint": 23, "omit_methyl": 23, "use_table_group_bi": 23, "singleton_expr": 23, "detail": [23, 25, 36, 38, 47], "variant_count": 23, "trinucleotid": 23, "featur": [23, 25, 38, 42, 43], "exome_coverag": 23, "omit": 23, "singleton_express": 23, "singleton_count": 23, "get_downsampling_freq_indic": [23, 27], "variant_qu": 23, "genetic_ancestry_label": 23, "downsampling_counts_expr": [23, 27], "explode_downsamplings_o": [23, 27], "downsampling_meta": 23, "syn": 23, "mi": 23, "gen_anc_exp": 23, "annotate_mutation_typ": [23, 27], "context_length": 23, "num_scan_context_length": 23, "cpg": 23, "transit": [23, 42], "mutation_typ": 23, "transvers": 23, "mutation_type_model": 23, "repo": 23, "trimer_from_heptam": [23, 27], "trim": 23, "heptam": 23, "trimer": 23, "collapse_strand": [23, 27], "dedupl": [23, 38], "dna": 23, "complement": 23, "was_flip": 23, "flip": 23, "transform_methylation_level": [23, 27], "methylation_cutoff": 23, "assemble_constraint_context_ht": [23, 27], "methylation_ht": 23, "gerp_ht": 23, "coverage_ht": 23, "an_ht": 23, "freq_ht": 23, "filter_ht": 23, "transformation_func": 23, "assembl": 23, "necessari": [23, 36], "was_split": [23, 39], "add_most_severe_csq_to_tc_within_vep_root": [23, 27, 38], "gerp": [23, 25], "etc": [23, 39], "uk": [23, 37], "biobank": [23, 37], "an_glob": 23, "strata_sample_count": 23, "strata_meta": 23, "freq_glob": 23, "build_model": [23, 27], "coverage_expr": 23, "weight": 23, "high_cov_definit": 23, "upper_cov_cutoff": 23, "skip_coverage_model": 23, "log10_coverag": 23, "plateau": 23, "plateau_model": 23, "linear": [23, 26], "against": [23, 31], "substitut": [23, 32, 37], "remaind": 23, "observed_vari": 23, "observed_": 23, "possible_vari": 23, "built": [23, 25], "factor": [23, 26], "correct": [23, 37], "log10": 23, "high_coverage_scale_factor": 23, "get_proportion_observed_by_coverag": 23, "downsampling_counts_": 23, "cover": 23, "build_plateau_model": [23, 27], "cpg_expr": 23, "mu_snp_expr": 23, "observed_variants_expr": 23, "possible_variants_expr": 23, "pops_observed_variants_array_expr": 23, "predict": [23, 25, 37, 38, 43], "build_coverage_model": [23, 27], "low_coverage_oe_expr": 23, "get_all_pop_length": [23, 27], "obs_expr": 23, "categor": 23, "get_constraint_grouping_expr": [23, 27], "vep_annotation_expr": 23, "include_transcript_group": 23, "include_canonical_group": 23, "include_mane_select_group": 23, "modifi": [23, 38], "classic": 23, "polyphen": [23, 36, 38], "polyphen_predict": 23, "gene_symbol": [23, 36, 38], "insid": [23, 38], "transcript_id": [23, 25, 36], "mane_select": [23, 38], "actual": 23, "annotate_exploded_vep_for_constraint_group": [23, 27], "vep_annot": [23, 36, 38], "worst_csq_by_gen": [23, 36, 38, 46], "ignor": [23, 36, 43], "compute_expected_vari": [23, 27], "plateau_models_expr": 23, "cov_corr_expr": 23, "oe_aggregation_expr": [23, 27], "exclude_mu_sum": 23, "ob": 23, "mu": 23, "exp": 23, "gen_anc_ob": 23, "expected_vari": 23, "expected_variants_": 23, "compute_pli": [23, 27], "exp_expr": 23, "expected_valu": 23, "min_diff_converg": 23, "pli": 23, "exac": [23, 37], "paper": 23, "lek": 23, "m": [23, 37], "karczewski": 23, "k": [23, 26], "minikel": 23, "706": 23, "natur": 23, "536": 23, "285": 23, "291": 23, "2016": 23, "intoler": 23, "em": 23, "algorithm": 23, "three": [23, 39, 42], "respect": [23, 25, 31], "sensit": 23, "toler": [23, 37], "recess": 23, "rec": 23, "haploinsuffici": 23, "li": 23, "deplet": 23, "blekhman": 23, "clingen": 23, "dosag": 23, "supplementari": 23, "463": 23, "deriv": 23, "empir": 23, "diseas": 23, "089": 23, "fall": [23, 31, 36, 37], "pnull": 23, "unconstrain": 23, "prec": 23, "iter": [23, 34], "converg": 23, "oe_confidence_interv": [23, 27], "alpha": [23, 31], "around": [23, 31], "densiti": [23, 31], "poisson": 23, "dpoi": 23, "lambda": 23, "lamb": 23, "extend": [23, 35, 42], "bound": 23, "cumul": [23, 25, 31, 35], "signific": [23, 25], "calculate_raw_z_scor": [23, 27], "sign": 23, "z": [23, 37], "posit": [23, 28, 29, 30, 35, 37, 42, 44], "had": 23, "fewer": 23, "get_constraint_flag": [23, 27], "raw_z_expr": 23, "raw_z_lower_threshold": 23, "raw_z_upper_threshold": 23, "flag_postfix": 23, "no_exp_": 23, "zero": [23, 25], "outlier_": 23, "postfix": 23, "calculate_raw_z_score_sd": [23, 27], "flag_expr": 23, "mirror_neg_raw_z": 23, "multipli": 23, "mirror": 23, "add_gencode_transcript_annot": [23, 27], "gencode_ht": [23, 25, 36], "transcript_typ": [23, 25], "cds_length": 23, "num_coding_exon": 23, "becom": 23, "file_exist": [24, 27], "fname": 24, "parquet": 24, "_success": 24, "check_file_exists_raise_error": [24, 27], "error_if_exist": 24, "error_if_not_exist": 24, "error_if_exists_msg": 24, "error_if_not_exists_msg": 24, "therefor": [24, 25], "write_temp_gc": [24, 27], "gcs_path": 24, "temp_path": 24, "select_primitives_from_ht": [24, 27], "primit": 24, "particularli": 24, "get_file_stat": [24, 27], "url": [24, 49], "project_id": 24, "md5": 24, "read_list_data": [24, 27], "input_file_path": 24, "gz": [24, 29, 46], "compress": 24, "repartition_for_join": [24, 27], "ht_path": 24, "new_partition_perc": 24, "via": 24, "_interv": 24, "effici": 24, "discuss": 24, "room": 24, "improv": 24, "2278": 24, "initi": 24, "increas": [24, 35], "create_vd": [24, 27], "output_path": 24, "vdse": 24, "gvcf": [24, 35, 37], "save_path": 24, "use_genome_default_interv": 24, "use_exome_default_interv": 24, "gvcf_batch_siz": 24, "directori": [24, 31], "temporari": 24, "life": 24, "cycl": 24, "polici": 24, "restart": 24, "combiner_plan": 24, "creation": [24, 49], "filter_to_adj": [25, 27], "filter_by_frequ": [25, 27], "direct": 25, "allele_count": 25, "At": [25, 35], "filter_row": 25, "combine_funct": [25, 27], "func_list": 25, "operator_func": 25, "iand": 25, "left": 25, "right": 25, "low_conf_regions_expr": [25, 27], "filter_telomeres_and_centromer": 25, "telomer": 25, "centromer": 25, "restrict": 25, "filter_low_conf_region": [25, 27], "keyword": [25, 35], "filter_to_autosom": [25, 27], "add_filters_expr": [25, 27], "current_filt": 25, "subset_samples_and_vari": [25, 27], "sample_path": 25, "table_kei": 25, "remove_dead_allel": 25, "filter_to_clinvar_pathogen": [25, 27], "clnrevstat_field": 25, "clnrevstat": 25, "clnsig_field": 25, "clnsig": 25, "clnsigconf_field": 25, "clnsigconf": 25, "remove_no_assert": 25, "remove_conflict": 25, "clinvar": [25, 42], "pathogen": [25, 37], "clinvar_ht": 25, "could": [25, 39], "review": 25, "clinic": 25, "signifc": 25, "conflict": 25, "undefin": 25, "assert": 25, "filter_gencode_ht": [25, 27], "protein_cod": [25, 38], "by_gene_symbol": 25, "cd": [25, 36], "symbol": [25, 36, 38, 41], "filter_by_interv": [25, 27], "padding_bp": [25, 28], "max_collect_interv": 25, "3000": 25, "pad": [25, 28], "filter_interv": [25, 35], "caus": 25, "memori": [25, 35], "filter_by_gencode_interv": [25, 27], "filter_to_gencode_cd": [25, 27], "filter_gencode_to_cd": 25, "analys": [25, 36], "step": [25, 47], "retain": [25, 35], "remove_fields_from_const": [25, 27], "fields_to_remov": 25, "filter_x_nonpar": [25, 27], "filter_y_nonpar": [25, 27], "filter_by_numeric_expr_rang": [25, 27], "filter_rang": 25, "keep_between": 25, "inclus": 25, "themselv": 25, "filter_for_mu": [25, 27], "gerp_lower_cutoff": 25, "9885": 25, "gerp_upper_cutoff": 25, "6607": 25, "precalcul": 25, "5th": [25, 33], "95th": 25, "percentil": 25, "intron": [25, 38], "intergen": 25, "split_vds_by_strata": [25, 27], "vdss": 25, "variant_data": [25, 35], "filter_arrays_by_meta": [25, 27], "meta_expr": 25, "meta_indexed_expr": 25, "items_to_filt": 25, "combine_oper": 25, "exact_match": 25, "item": [25, 39], "papuan": 25, "index_glob": [25, 27, 37], "often": [25, 38], "to_phr": [26, 27], "linear_expr": 26, "from_phr": [26, 27], "phred_score_expr": 26, "get_median_and_mad_expr": [26, 27], "metric_expr": 26, "4826": 26, "absolut": 26, "merge_stats_counters_expr": [26, 27], "counter": 26, "sort_interv": [27, 28], "union_interv": [27, 28], "interval_length": [27, 28], "pad_interv": [27, 28], "grch37_to_grch38_chain": [27, 29], "grch38_to_grch37_chain": [27, 29], "get_liftover_genom": [27, 29], "liftover_expr": [27, 29], "default_lift_data": [27, 29], "liftover_using_gnomad_map": [27, 29], "parse_vari": [27, 30], "parse_locus_interv": [27, 30], "new_show": [27, 31], "plot_hail_hist": [27, 31], "plot_multi_hail_hist": [27, 31], "plot_hail_hist_cumul": [27, 31], "plot_hail_hist_both": [27, 31], "set_font_s": [27, 31], "linear_and_log_tab": [27, 31], "plot_hail_file_metadata": [27, 31], "scale_file_s": [27, 31], "get_rows_data": [27, 31], "pair_plot": [27, 31], "get_reference_ht": [27, 32, 35], "add_reference_sequ": [27, 32], "get_reference_genom": [27, 32], "make_faf_index_dict": [27, 33], "make_freq_index_dict": [27, 33], "make_freq_index_dict_from_meta": [27, 33], "slackclient": [27, 34], "slack_notif": [27, 34], "compute_last_ref_block_end": [27, 35], "densify_sit": [27, 35], "get_as_info_expr": [27, 35], "get_site_info_expr": [27, 35], "default_compute_info": [27, 35], "split_info_annot": [27, 35], "split_lowqual_annot": [27, 35], "impute_sex_ploidi": [27, 35], "densify_all_reference_sit": [27, 35], "compute_stats_per_ref_sit": [27, 35], "compute_coverage_stat": [27, 35], "get_allele_number_agg_func": [27, 35], "compute_allele_number_per_ref_sit": [27, 35], "filter_ref_block": [27, 35], "reproductive_tissu": [27, 36], "cell_lin": [27, 36], "get_tissues_to_exclud": [27, 36], "summarize_transcript_express": [27, 36], "get_expression_proport": [27, 36], "filter_expression_ht_by_tissu": [27, 36], "tissue_expression_ht_to_arrai": [27, 36], "tx_filter_variants_by_csq": [27, 36], "tx_annotate_vari": [27, 36], "tx_aggregate_vari": [27, 36], "perform_tx_annotation_pipelin": [27, 36], "clean_tissue_name_for_brows": [27, 36], "create_tx_annotation_by_region": [27, 36], "faf_pop": [27, 37], "as_field": [27, 37], "site_field": [27, 37], "allele_type_field": [27, 37], "region_flag_field": [27, 37], "joint_region_flag_field": [27, 37], "rf_field": [27, 37], "as_vqsr_field": [27, 37], "vqsr_field": [27, 37], "info_dict": [27, 37], "in_silico_annotations_info_dict": [27, 37], "vrs_fields_dict": [27, 37], "sparse_entri": [27, 37], "format_dict": [27, 37], "adjust_vcf_incompatible_typ": [27, 37], "make_label_combo": [27, 37], "make_combo_header_text": [27, 37], "create_label_group": [27, 37], "make_info_dict": [27, 37], "add_as_info_dict": [27, 37], "make_vcf_filter_dict": [27, 37], "make_hist_bin_edges_expr": [27, 37], "make_hist_dict": [27, 37], "set_female_y_metrics_to_na": [27, 37], "build_vcf_export_refer": [27, 37], "rekey_new_refer": [27, 37], "current_vep_vers": [27, 38], "csq_code": [27, 38], "csq_splice": [27, 38], "possible_ref": [27, 38], "vep_config_path": [27, 38], "vep_csq_field": [27, 38], "vep_csq_head": [27, 38], "get_vep_help": [27, 38], "get_vep_context": [27, 38], "vep_or_lookup_vep": [27, 38, 46], "get_most_severe_consequence_expr": [27, 38], "add_most_severe_consequence_to_consequ": [27, 38], "process_consequ": [27, 38, 46], "filter_vep_to_canonical_transcript": [27, 38], "filter_vep_to_mane_select_transcript": [27, 38], "filter_vep_to_synonymous_vari": [27, 38], "filter_vep_to_gene_list": [27, 38], "vep_struct_to_csq": [27, 38], "get_most_severe_consequence_for_summari": [27, 38], "filter_vep_transcript_csq": [27, 38], "filter_vep_transcript_csqs_expr": [27, 38], "explode_by_vep_annot": [27, 38], "is_sort": 28, "grch37_to_grch38": 29, "lift": 29, "grch38_to_grch37": 29, "referencegenom": [29, 32, 37], "destination_refer": 29, "coordin": [29, 37], "original_locu": 29, "original_allel": [29, 37], "locus_fail_liftov": 29, "ref_allele_mismatch": 29, "remove_failed_sit": 29, "establish": 29, "variant_str": 30, "chr": [30, 37], "40": [31, 35], "hist_data": 31, "titl": 31, "fill_color": 31, "033649": 31, "outlier_fill_color": 31, "036564": 31, "line_color": 31, "hover_mod": 31, "mous": 31, "hide_zero": 31, "come": 31, "straight": 31, "color": 31, "bar": 31, "hover": 31, "vline": 31, "hline": 31, "figur": 31, "usag": [31, 34, 35, 36], "group_bi": 31, "len": [31, 35], "line_width": 31, "font_siz": 31, "12pt": 31, "plot_func": 31, "tab": 31, "t_path": 31, "grid": 31, "Or": 31, "unord": 31, "file_s": 31, "rows_fil": 31, "label_col": [31, 43], "pan": 31, "box_zoom": 31, "reset": [31, 37], "wheel_zoom": 31, "box_select": 31, "lasso_select": 31, "help": [31, 38, 47], "tooltip_col": 31, "diagon": 31, "scatter": 31, "color_dict": 31, "rgb": 31, "hex": 31, "tooltip": 31, "add_all_substitut": 32, "filter_n": 32, "obscur": 32, "slow": 32, "unknown": [32, 37], "fasta": 32, "add_sequ": 32, "label_delimit": [33, 37], "_": [33, 37], "look": [33, 39], "access": [33, 49], "nfe_xx": 33, "token": 34, "api": [34, 48], "client": 34, "send_fil": 34, "filenam": 34, "txt": 34, "filetyp": 34, "comment": 34, "send": 34, "channel": 34, "upload": 34, "send_messag": 34, "icon_emoji": 34, "long": 34, "snippet": 34, "emoji": 34, "icon": 34, "notif": 34, "wrap": 34, "stack": 34, "trace": 34, "usernam": 34, "run_analysi": 34, "upstream": 35, "sinc": 35, "beyond": 35, "last_end_posit": 35, "last_end_positions_ht": 35, "semi_join_row": 35, "minim": 35, "furthest": 35, "tag": 35, "semi": 35, "few": 35, "sum_agg_field": 35, "int32_sum_agg_field": 35, "vardp": [35, 37], "median_agg_field": 35, "readposranksum": [35, 37], "mqranksum": [35, 37], "array_sum_agg_field": 35, "raw_mqanddp": 35, "alt_alleles_range_array_field": 35, "alt_alleles_range_arrai": 35, "treat_fields_as_allele_specif": 35, "retain_cdf": 35, "cdf_k": 35, "as_sb_tabl": [35, 37], "accord": 35, "nomenclatur": 35, "raw_mq": 35, "mq_dp": 35, "gvcf_info": 35, "prioriti": 35, "clash": 35, "Then": 35, "cdf": 35, "later": [35, 38, 43, 47], "accuraci": [35, 43], "tradeoff": 35, "AS": 35, "site_annot": 35, "as_annot": 35, "quasi_as_annot": 35, "lowqual_indel_phred_het_prior": 35, "ac_filter_group": 35, "nonref": 35, "naive_coalesc": 35, "10k": 35, "scienc": 35, "info_expr": 35, "a_index": 35, "split_multi": 35, "split_multi_ht": 35, "lowqual_expr": 35, "particular": [35, 37], "excluded_calling_interv": 35, "included_calling_interv": 35, "chr_x": 35, "chr_y": 35, "use_only_vari": 35, "reference_ht": 35, "row_key_field": 35, "entry_keep_field": 35, "densif": 35, "row_keep_field": 35, "sex_karyotype_field": 35, "lgt": [35, 37], "coverage_over_x_bin": 35, "gt_field": 35, "awar": 36, "cervix_ectocervix": 36, "cervix_endocervix": 36, "fallopiantub": 36, "ovari": 36, "prostat": 36, "testi": 36, "uteru": 36, "vagina": 36, "reproduct": 36, "tissu": 36, "gtex": 36, "cells_ebv_transformedlymphocyt": 36, "cells_transformedfibroblast": 36, "cells_culturedfibroblast": 36, "reach": 36, "satur": 36, "stabl": 36, "studi": 36, "bmcgenom": 36, "biomedcentr": 36, "1186": 36, "s12864": 36, "021": 36, "08125": 36, "transcript_expression_expr": 36, "transcript_tpm": 36, "tissue_expr": 36, "summary_agg_func": 36, "transcript_express": 36, "expression_proport": 36, "gene_id": [36, 38], "tissue_1": 36, "tissue_2": 36, "quantif": 36, "tissues_to_keep": 36, "tissues_to_exclud": 36, "annotations_to_extract": 36, "tissue_express": 36, "filter_to_cd": 36, "filter_to_gen": 36, "match_by_gene_symbol": [36, 38], "filter_to_csq": 36, "ignore_spl": 36, "filter_to_protein_cod": 36, "vep_root": [36, 38], "include_polyphen_priorit": 36, "further": 36, "amino_acid": [36, 38], "stop_retained_vari": [36, 38], "And": 36, "root": [36, 37, 38], "priorit": [36, 37], "process": [36, 46], "preprocess": 36, "tissues_to_exclude_from_mean": 36, "worst_csq_for_vari": [36, 46], "worst_csq_by_gene_canon": 36, "worst_csq_for_variant_canon": 36, "worst": [36, 37, 38], "additional_group_bi": 36, "lof_flag": [36, 38], "transcript_abl": [36, 38], "stop_lost": [36, 38], "start_lost": [36, 38], "initiator_codon_vari": [36, 38], "transcript_amplif": [36, 38], "inframe_insert": [36, 38], "inframe_delet": [36, 38], "protein_altering_vari": [36, 38], "splice_donor_5th_base_vari": [36, 38], "splice_region_vari": [36, 38], "splice_donor_region_vari": [36, 38], "splice_polypyrimidine_tract_vari": [36, 38], "incomplete_terminal_codon_vari": [36, 38], "start_retained_vari": [36, 38], "coding_sequence_vari": [36, 38], "coding_transcript_vari": [36, 38], "stop": [36, 47], "tissue_nam": 36, "clean": 36, "compat": 36, "uppercas": 36, "letter": 36, "lowercas": 36, "underscor": 36, "word": 36, "addition": 36, "handl": 36, "organ": 36, "exp_prop_mean": 36, "tissue1": 36, "tissue2": 36, "gene1": 36, "gene2": 36, "chrom": 36, "raw_ac_afr_femal": 37, "earlier": 37, "as_f": 37, "as_mq": 37, "as_mqranksum": 37, "as_pab_max": 37, "as_qd": 37, "as_readposranksum": 37, "as_sor": 37, "as_vardp": 37, "fail_interval_qc": 37, "outside_broad_capture_region": 37, "outside_ukb_capture_region": 37, "outside_broad_calling_region": 37, "outside_ukb_calling_region": 37, "not_called_in_exom": 37, "not_called_in_genom": 37, "rf_positive_label": 37, "rf_negative_label": 37, "rf_label": [37, 42, 43], "rf_train": [37, 42], "rf_tp_probabl": 37, "as_culprit": 37, "as_vqslod": 37, "negative_train_sit": [37, 42], "positive_train_sit": [37, 42], "bia": 37, "baseqranksum": 37, "wilcoxon": 37, "excess": 37, "hardi": 37, "weinberg": 37, "equilibrium": 37, "vqslod": 37, "versu": 37, "culprit": 37, "85": 37, "20x": 37, "10x": 37, "coincid": 37, "span": 37, "elsewher": 37, "inbreeding_coeff": 37, "complex": 37, "only_het": 37, "sibling_singleton": 37, "wide": 37, "amongst": 37, "transmitted_singleton": 37, "cadd_phr": 37, "cadd": 37, "c": 37, "billion": 37, "deleteri": 37, "cadd_raw_scor": 37, "extent": 37, "profil": 37, "suggest": 37, "simul": 37, "pangolin_largest_d": 37, "pangolin": 37, "delta": 37, "reflect": 37, "alter": 37, "phylop": 37, "conserv": 37, "241": 37, "placent": 37, "mammal": 37, "zoonomia": 37, "28": 37, "acceler": 37, "evolut": 37, "neutral": 37, "drift": 37, "slower": 37, "polyphen_max": 37, "impact": [37, 38], "amino": 37, "acid": 37, "revel_max": 37, "revel": 37, "13": 37, "predictor": [37, 45, 48], "sift_max": 37, "spliceai_ds_max": 37, "illumina": 37, "spliceai": 37, "silico": 37, "vrs_allele_id": 37, "vrs_end": 37, "interresidu": 37, "vrs_start": 37, "vrs_state": 37, "liter": 37, "min_dp": 37, "pgt": 37, "pid": 37, "la": 37, "lad": 37, "lpgt": 37, "lpl": 37, "rgq": 37, "255": 37, "bad": 37, "mate": 37, "physic": 37, "connect": [37, 47], "record": 37, "compris": 37, "detect": 37, "pipe_delimited_annot": 37, "as_mq_dp": 37, "as_raw_mq": 37, "done": [37, 39, 42, 43], "coerc": 37, "pipe": 37, "afr_mal": 37, "afr_femal": 37, "nfe_mal": 37, "nfe_femal": 37, "amr_mal": 37, "amr_femal": 37, "globals_arrai": 37, "meta_arrai": 37, "preposit": 37, "combo_dict": 37, "pop_nam": 37, "programmat": 37, "group_typ": 37, "combo_field": 37, "automat": [37, 42], "all_group": 37, "pop_sex_group": 37, "prefix_before_metr": 37, "amish": 37, "admix": 37, "ashkenazi": 37, "jewish": 37, "bgr": 37, "bulgarian": 37, "eastern": 37, "south": 37, "asian": 37, "east": 37, "est": 37, "estonian": 37, "eur": 37, "jpn": 37, "kor": 37, "korean": 37, "mde": 37, "middl": 37, "nwe": 37, "north": 37, "western": 37, "oea": 37, "oeu": 37, "onf": 37, "sas_non_consang": 37, "seu": 37, "sgp": 37, "singaporean": 37, "swe": 37, "swedish": 37, "uniform": 37, "unk": 37, "freq_ctt": 37, "freq_cmh": 37, "freq_stat_union": 37, "description_text": 37, "age_hist_distribut": 37, "carrier": 37, "subpopul": 37, "auto": 37, "conting": 37, "ctt": 37, "cochran": 37, "mantel": 37, "haenszel": 37, "cmh": 37, "snp_cutoff": 37, "indel_cutoff": 37, "inbreeding_cutoff": 37, "variant_qc_filt": 37, "ann_with_hist": 37, "include_age_hist": 37, "reformat": 37, "hist_metric_list": 37, "drop_n_smaller_larg": 37, "keep_contig": 37, "chr2": 37, "chr3": 37, "chr4": 37, "chr5": 37, "chr6": 37, "chr7": 37, "chr8": 37, "chr9": 37, "chr10": 37, "chr11": 37, "chr12": 37, "chr13": 37, "chr14": 37, "chr15": 37, "chr16": 37, "chr17": 37, "chr18": 37, "chr19": 37, "chr21": 37, "chr22": 37, "keep_chrm": 37, "elimin": 37, "22": 37, "chr3_gl000221v1_random": 37, "155397": 37, "assembli": 37, "chrm": 37, "re": [37, 39], "105": 38, "vep_data": 38, "gcloud": 38, "101": 38, "feature_typ": 38, "exon": 38, "hgvsc": 38, "hgvsp": 38, "cdna_posit": 38, "cds_posit": 38, "protein_posit": 38, "codon": 38, "allele_num": 38, "variant_class": 38, "minimis": 38, "symbol_sourc": 38, "hgnc_id": 38, "tsl": 38, "appri": 38, "ccd": 38, "ensp": 38, "swissprot": 38, "trembl": 38, "uniparc": 38, "gene_pheno": 38, "sift": 38, "domain": 38, "hgvs_offset": 38, "motif_nam": 38, "motif_po": 38, "high_inf_po": 38, "motif_score_chang": 38, "lof_filt": 38, "lof_info": 38, "mane_plus_clin": 38, "uniprot_isoform": 38, "mirna": 38, "pubm": 38, "transcription_factor": 38, "lc": 38, "dataproc": [38, 46, 47, 49], "hailctl": [38, 46, 47, 49], "command": [38, 47], "vep_config_uri": 38, "default_refer": 38, "vepe": 38, "reference_vep_ht": 38, "vep_vers": 38, "lookup": 38, "databas": 38, "confirm": 38, "referenc": 38, "suitabl": 38, "grab": 38, "vep_config": 38, "vep_context": 38, "csq_order": 38, "intron_vari": 38, "tc": 38, "most_severe_csq_field": 38, "consequence_term": 38, "penalize_flag": 38, "has_polyphen": 38, "any_lof": 38, "csq_score": 38, "deduct": 38, "NO": 38, "everyth": 38, "penal": 38, "filter_empty_csq": 38, "upstream_gene_vari": 38, "downstream_gene_vari": 38, "vep_expr": 38, "has_polyphen_sift": 38, "flexibl": 38, "counterpart": 38, "while": 38, "Their": 38, "usual": 38, "mature_mirna_vari": 38, "5_prime_utr_vari": 38, "3_prime_utr_vari": 38, "non_coding_transcript_exon_vari": 38, "non_coding_exon_vari": 38, "nmd_transcript_vari": 38, "non_coding_transcript_vari": 38, "nc_transcript_vari": 38, "tfbs_ablat": 38, "tfbs_amplif": 38, "tf_binding_site_vari": 38, "regulatory_region_abl": 38, "regulatory_region_amplif": 38, "feature_elong": 38, "regulatory_region_vari": 38, "feature_trunc": 38, "intergenic_vari": 38, "sequence_vari": 38, "most_severe_csq": 38, "ensembl_onli": 38, "refseq": 38, "emsembl": 38, "keep_csq": 38, "keep_gen": 38, "additional_filtering_criteria": 38, "compute_ranked_bin": [39, 40, 42], "score_expr": 39, "bin_expr": 39, "compute_snv_indel_separ": 39, "desc": 39, "mutual": 39, "exclus": 39, "var1": [39, 41], "var2": [39, 41], "var3": 39, "var4": 39, "biallelic_bin": 39, "singleton_bin": 39, "descend": 39, "compute_grouped_binned_ht": [39, 40, 42], "bin_ht": 39, "create_binned_ht": [39, 40, 42], "bin_id": 39, "bi_allel": 39, "score_bin_agg": [39, 40, 42], "_parent": [39, 42], "groupedt": [39, 42], "compute_binned_truth_sample_concord": [39, 40], "binned_score_ht": 39, "add_bin": 39, "concord": 39, "tp": [39, 42, 44], "fp": [39, 42, 44], "fn": 39, "truth_gt": 39, "create_truth_sample_ht": [39, 40], "truth_mt": 39, "high_confidence_intervals_ht": 39, "add_rank": [39, 40], "subrank_expr": 39, "subrank": 39, "rank_variant_count": 39, "form": 39, "name_of_subrank": 39, "subrank_filtering_expr": 39, "get_r_human_read": [40, 41], "get_r_for_pair_of_vari": [40, 41], "get_r_within_gene_in_pop": [40, 41], "get_r_within_gen": [40, 41], "generate_trio_stat": [40, 42], "generate_sib_stat": [40, 42], "train_rf_model": [40, 42], "run_rf_test": [40, 43], "check_ht_fields_for_spark": [40, 43], "get_columns_quantil": [40, 43], "median_impute_featur": [40, 43], "ht_to_rf_df": [40, 43], "get_features_import": [40, 43], "get_label": [40, 43], "test_model": [40, 43], "apply_rf_model": [40, 43], "save_model": [40, 43], "load_model": [40, 43], "train_rf": [40, 42, 43], "get_rf_run": [40, 43], "get_run_data": [40, 43], "pretty_print_run": [40, 43], "sample_training_exampl": [40, 44], "ref_genom": 41, "get_ld_matrix": 41, "get_ld_index": 41, "parse_locu": 41, "10146": 41, "10151": 41, "ta": 41, "01789767935482124": 41, "tlocu": 41, "tarrai": 41, "quadrat": 41, "exercis": 41, "caution": 41, "get_gnomad_public_data": 41, "get_gene_interv": 41, "fast": 41, "add_substrat": 42, "meant": 42, "wrapper": [42, 43], "ac_raw": 42, "fam_stats_ht": 42, "easili": 42, "binned_ht": 42, "grouped_binned_ht": 42, "agg_ht": 42, "ac_qc_samples_unrelated_raw": 42, "fail_hard_filt": 42, "truth_ht": 42, "omni": 42, "kgp_phase1_hc": 42, "n_de_novos_adj": 42, "n_de_novos_raw": 42, "n_transmitted_raw": 42, "n_untransmitted_raw": 42, "min_scor": 42, "minimun": 42, "max_scor": 42, "maiximum": 42, "n_in": 42, "n_del": 42, "n_ti": 42, "n_tv": 42, "trnasvers": 42, "n_1bp_indel": 42, "n_mod3bp_indel": 42, "divis": 42, "n_singleton": 42, "n_vqsr_pos_train": 42, "n_vqsr_neg_train": 42, "n_clinvar": 42, "n_de_novos_singleton_adj": 42, "n_de_novo_singleton": 42, "unfilt": 42, "n_de_novo": 42, "n_trans_singleton": 42, "n_untrans_singleton": 42, "untransmit": 42, "n_omni": 42, "n_mill": 42, "n_hapmap": 42, "n_kgp_phase1_hc": 42, "rf_featur": 42, "tp_expr": [42, 44], "fp_expr": [42, 44], "fp_to_tp": [42, 44], "num_tre": [42, 43], "max_depth": [42, 43], "test_expr": [42, 44], "withheld": [42, 43], "rf_test": 42, "features_import": [42, 43], "test_result": [42, 43], "maxmimum": 42, "hold": 42, "pipelinemodel": [42, 43], "tmp": 43, "dummi": 43, "quantil": 43, "relative_error": 43, "approx_median": 43, "feature_imput": 43, "features_median": 43, "variants_by_strata": 43, "basic": 43, "back": 43, "rf_pipelin": 43, "rf_index": 43, "assembler_index": 43, "vectorassembl": 43, "stringindex": 43, "rf_model": 43, "prediction_col_nam": 43, "rf_predict": 43, "confus": 43, "probability_col_nam": 43, "rf_probabl": 43, "out_path": 43, "input_path": 43, "rf_json_fp": 43, "wasn": 43, "input_arg": 43, "test_interv": 43, "hash": 43, "tv": 44, "effect": [45, 48], "gnomad": [46, 47, 49], "bill": [46, 49], "charg": [46, 49], "d": 46, "import_vcf": 46, "drop_sampl": 46, "drastic": 46, "speed": 46, "worst_consequence_term": 46, "short": 47, "guid": 47, "date": 47, "septemb": 47, "2024": 47, "ip": 47, "131": 47, "instal": 47, "pip": 47, "jupyt": 47, "notebook": 47, "gnomad_v2_exom": 47, "exomes_ht": 47, "gnomad_v2_genom": 47, "genomes_ht": 47, "gnomad_v3_genom": 47, "shut": 47, "finish": 47, "program": 49, "gc": 49, "gnomad_public_resource_configur": 49, "transfer": 49, "your": 49, "my": 49}, "objects": {"gnomad.assessment": [[1, 0, 0, "-", "summary_stats"], [2, 0, 0, "-", "validity_checks"]], "gnomad.assessment.summary_stats": [[1, 1, 1, "", "default_generate_gene_lof_matrix"], [1, 1, 1, "", "default_generate_gene_lof_summary"], [1, 1, 1, "", "freq_bin_expr"], [1, 1, 1, "", "generate_filter_combinations"], [1, 1, 1, "", "get_an_criteria"], [1, 1, 1, "", "get_het_hom_summary_dict"], [1, 1, 1, "", "get_summary_ac_dict"], [1, 1, 1, "", "get_summary_counts"], [1, 1, 1, "", "get_summary_counts_dict"], [1, 1, 1, "", "get_summary_stats_csq_filter_expr"], [1, 1, 1, "", "get_summary_stats_filter_group_meta"], [1, 1, 1, "", "get_summary_stats_variant_filter_expr"], [1, 1, 1, "", "get_tx_expression_expr"]], "gnomad.assessment.validity_checks": [[2, 1, 1, "", "check_global_and_row_annot_lengths"], [2, 1, 1, "", "check_missingness_of_struct"], [2, 1, 1, "", "check_raw_and_adj_callstats"], [2, 1, 1, "", "check_sex_chr_metrics"], [2, 1, 1, "", "compare_row_counts"], [2, 1, 1, "", "compare_subset_freqs"], [2, 1, 1, "", "compute_missingness"], [2, 1, 1, "", "count_vep_annotated_variants_per_interval"], [2, 1, 1, "", "flatten_missingness_struct"], [2, 1, 1, "", "generic_field_check"], [2, 1, 1, "", "generic_field_check_loop"], [2, 1, 1, "", "make_filters_expr_dict"], [2, 1, 1, "", "make_group_sum_expr_dict"], [2, 1, 1, "", "pprint_global_anns"], [2, 1, 1, "", "sum_group_callstats"], [2, 1, 1, "", "summarize_variant_filters"], [2, 1, 1, "", "summarize_variants"], [2, 1, 1, "", "unfurl_array_annotations"], [2, 1, 1, "", "validate_release_t"], [2, 1, 1, "", "vcf_field_check"]], "gnomad.resources": [[4, 0, 0, "-", "config"], [12, 0, 0, "-", "import_resources"], [14, 0, 0, "-", "resource_utils"]], "gnomad.resources.config": [[4, 2, 1, "", "GnomadPublicResourceSource"], [4, 1, 1, "", "get_default_public_resource_source"]], "gnomad.resources.config.GnomadPublicResourceSource": [[4, 3, 1, "", "AZURE_OPEN_DATASETS"], [4, 3, 1, "", "GNOMAD"], [4, 3, 1, "", "GOOGLE_CLOUD_PUBLIC_DATASETS"], [4, 3, 1, "", "REGISTRY_OF_OPEN_DATA_ON_AWS"]], "gnomad.resources.grch37": [[5, 0, 0, "-", "gnomad"], [6, 0, 0, "-", "gnomad_ld"], [8, 0, 0, "-", "reference_data"]], "gnomad.resources.grch37.gnomad": [[5, 1, 1, "", "browser_gene"], [5, 1, 1, "", "constraint"], [5, 1, 1, "", "coverage"], [5, 1, 1, "", "liftover"], [5, 1, 1, "", "mnv"], [5, 1, 1, "", "pext"], [5, 1, 1, "", "public_pca_loadings"], [5, 1, 1, "", "public_release"], [5, 1, 1, "", "release_vcf_path"]], "gnomad.resources.grch37.gnomad_ld": [[6, 1, 1, "", "ld_index"], [6, 1, 1, "", "ld_matrix"], [6, 1, 1, "", "ld_scores"]], "gnomad.resources.grch37.reference_data": [[8, 1, 1, "", "get_truth_ht"], [8, 1, 1, "", "transform_grch37_methylation"]], "gnomad.resources.grch38": [[9, 0, 0, "-", "gnomad"], [11, 0, 0, "-", "reference_data"]], "gnomad.resources.grch38.gnomad": [[9, 4, 1, "", "COHORTS_WITH_POP_STORED_AS_SUBPOP"], [9, 4, 1, "", "DOWNSAMPLINGS"], [9, 4, 1, "", "GROUPS"], [9, 4, 1, "", "HGDP_POPS"], [9, 4, 1, "", "POPS"], [9, 4, 1, "", "POPS_TO_REMOVE_FOR_POPMAX"], [9, 4, 1, "", "SEXES"], [9, 4, 1, "", "SUBSETS"], [9, 4, 1, "", "TGP_POPS"], [9, 4, 1, "", "TGP_POP_NAMES"], [9, 1, 1, "", "add_grpMaxFAF95_v4"], [9, 1, 1, "", "all_sites_an"], [9, 1, 1, "", "browser_gene"], [9, 1, 1, "", "browser_variant"], [9, 1, 1, "", "constraint"], [9, 1, 1, "", "coverage"], [9, 1, 1, "", "coverage_tsv_path"], [9, 1, 1, "", "gnomad_gks"], [9, 1, 1, "", "pext"], [9, 1, 1, "", "public_release"], [9, 1, 1, "", "release_vcf_path"]], "gnomad.resources.grch38.reference_data": [[11, 1, 1, "", "get_truth_ht"], [11, 1, 1, "", "transform_grch38_methylation"]], "gnomad.resources.import_resources": [[12, 1, 1, "", "get_module_importable_resources"], [12, 1, 1, "", "get_resources_descriptions"], [12, 1, 1, "", "main"]], "gnomad.resources.resource_utils": [[14, 2, 1, "", "BaseResource"], [14, 2, 1, "", "BaseVersionedResource"], [14, 2, 1, "", "BlockMatrixResource"], [14, 7, 1, "", "DataException"], [14, 2, 1, "", "ExpressionResource"], [14, 4, 1, "", "GNOMAD_PUBLIC_BUCKETS"], [14, 2, 1, "", "GnomadPublicBlockMatrixResource"], [14, 2, 1, "", "GnomadPublicMatrixTableResource"], [14, 2, 1, "", "GnomadPublicPedigreeResource"], [14, 2, 1, "", "GnomadPublicResource"], [14, 2, 1, "", "GnomadPublicTableResource"], [14, 2, 1, "", "MatrixTableResource"], [14, 2, 1, "", "PedigreeResource"], [14, 7, 1, "", "ResourceNotAvailable"], [14, 2, 1, "", "TableResource"], [14, 2, 1, "", "VariantDatasetResource"], [14, 2, 1, "", "VersionedBlockMatrixResource"], [14, 2, 1, "", "VersionedMatrixTableResource"], [14, 2, 1, "", "VersionedPedigreeResource"], [14, 2, 1, "", "VersionedTableResource"], [14, 2, 1, "", "VersionedVariantDatasetResource"], [14, 1, 1, "", "import_gencode"], [14, 1, 1, "", "import_sites_vcf"]], "gnomad.resources.resource_utils.BaseResource": [[14, 3, 1, "", "expected_file_extensions"], [14, 5, 1, "", "import_resource"], [14, 6, 1, "", "path"]], "gnomad.resources.resource_utils.BaseVersionedResource": [[14, 3, 1, "", "default_version"], [14, 3, 1, "", "resource_class"], [14, 3, 1, "", "versions"]], "gnomad.resources.resource_utils.BlockMatrixResource": [[14, 5, 1, "", "bm"], [14, 3, 1, "", "expected_file_extensions"], [14, 5, 1, "", "import_resource"]], "gnomad.resources.resource_utils.ExpressionResource": [[14, 3, 1, "", "expected_file_extensions"], [14, 5, 1, "", "he"], [14, 5, 1, "", "import_resource"]], "gnomad.resources.resource_utils.GnomadPublicBlockMatrixResource": [[14, 5, 1, "", "bm"]], "gnomad.resources.resource_utils.GnomadPublicMatrixTableResource": [[14, 5, 1, "", "mt"]], "gnomad.resources.resource_utils.GnomadPublicPedigreeResource": [[14, 5, 1, "", "ht"], [14, 5, 1, "", "pedigree"]], "gnomad.resources.resource_utils.GnomadPublicResource": [[14, 5, 1, "", "is_resource_available"]], "gnomad.resources.resource_utils.GnomadPublicTableResource": [[14, 5, 1, "", "ht"]], "gnomad.resources.resource_utils.MatrixTableResource": [[14, 3, 1, "", "expected_file_extensions"], [14, 5, 1, "", "import_resource"], [14, 5, 1, "", "mt"]], "gnomad.resources.resource_utils.PedigreeResource": [[14, 3, 1, "", "expected_file_extensions"], [14, 5, 1, "", "ht"], [14, 5, 1, "", "import_resource"], [14, 5, 1, "", "pedigree"]], "gnomad.resources.resource_utils.TableResource": [[14, 3, 1, "", "expected_file_extensions"], [14, 5, 1, "", "ht"], [14, 5, 1, "", "import_resource"]], "gnomad.resources.resource_utils.VariantDatasetResource": [[14, 3, 1, "", "expected_file_extensions"], [14, 5, 1, "", "import_resource"], [14, 5, 1, "", "vds"]], "gnomad.resources.resource_utils.VersionedBlockMatrixResource": [[14, 3, 1, "", "resource_class"]], "gnomad.resources.resource_utils.VersionedMatrixTableResource": [[14, 3, 1, "", "resource_class"]], "gnomad.resources.resource_utils.VersionedPedigreeResource": [[14, 3, 1, "", "resource_class"]], "gnomad.resources.resource_utils.VersionedTableResource": [[14, 3, 1, "", "resource_class"]], "gnomad.resources.resource_utils.VersionedVariantDatasetResource": [[14, 3, 1, "", "resource_class"]], "gnomad.sample_qc": [[15, 0, 0, "-", "ancestry"], [16, 0, 0, "-", "filtering"], [18, 0, 0, "-", "pipeline"], [19, 0, 0, "-", "platform"], [20, 0, 0, "-", "relatedness"], [21, 0, 0, "-", "sex"]], "gnomad.sample_qc.ancestry": [[15, 1, 1, "", "apply_onnx_classification_model"], [15, 1, 1, "", "apply_sklearn_classification_model"], [15, 1, 1, "", "assign_population_pcs"], [15, 1, 1, "", "convert_sklearn_rf_to_onnx"], [15, 1, 1, "", "pc_project"], [15, 1, 1, "", "run_pca_with_relateds"]], "gnomad.sample_qc.filtering": [[16, 1, 1, "", "compute_qc_metrics_residuals"], [16, 1, 1, "", "compute_stratified_metrics_filter"], [16, 1, 1, "", "compute_stratified_sample_qc"], [16, 1, 1, "", "determine_nearest_neighbors"], [16, 1, 1, "", "merge_sample_qc_expr"]], "gnomad.sample_qc.pipeline": [[18, 1, 1, "", "annotate_sex"], [18, 1, 1, "", "filter_rows_for_qc"], [18, 1, 1, "", "get_qc_mt"], [18, 1, 1, "", "infer_sex_karyotype"]], "gnomad.sample_qc.platform": [[19, 1, 1, "", "assign_platform_from_pcs"], [19, 1, 1, "", "compute_callrate_mt"], [19, 1, 1, "", "run_platform_pca"]], "gnomad.sample_qc.relatedness": [[20, 4, 1, "", "AMBIGUOUS_RELATIONSHIP"], [20, 4, 1, "", "DUPLICATE_OR_TWINS"], [20, 4, 1, "", "PARENT_CHILD"], [20, 4, 1, "", "SECOND_DEGREE_RELATIVES"], [20, 4, 1, "", "SIBLINGS"], [20, 4, 1, "", "UNRELATED"], [20, 1, 1, "", "calculate_de_novo_post_prob"], [20, 1, 1, "", "compute_related_samples_to_drop"], [20, 1, 1, "", "create_fake_pedigree"], [20, 1, 1, "", "default_get_de_novo_expr"], [20, 1, 1, "", "explode_duplicate_samples_ht"], [20, 1, 1, "", "filter_to_trios"], [20, 1, 1, "", "generate_sib_stats_expr"], [20, 1, 1, "", "generate_trio_stats_expr"], [20, 1, 1, "", "get_duplicated_samples"], [20, 1, 1, "", "get_duplicated_samples_ht"], [20, 1, 1, "", "get_relationship_expr"], [20, 1, 1, "", "get_slope_int_relationship_expr"], [20, 1, 1, "", "infer_families"]], "gnomad.sample_qc.sex": [[21, 1, 1, "", "adjust_sex_ploidy"], [21, 1, 1, "", "adjusted_sex_ploidy_expr"], [21, 1, 1, "", "gaussian_mixture_model_karyotype_assignment"], [21, 1, 1, "", "get_chr_x_hom_alt_cutoffs"], [21, 1, 1, "", "get_ploidy_cutoffs"], [21, 1, 1, "", "get_sex_expr"]], "gnomad.utils": [[22, 0, 0, "-", "annotations"], [23, 0, 0, "-", "constraint"], [24, 0, 0, "-", "file_utils"], [25, 0, 0, "-", "filtering"], [26, 0, 0, "-", "gen_stats"], [28, 0, 0, "-", "intervals"], [29, 0, 0, "-", "liftover"], [30, 0, 0, "-", "parse"], [31, 0, 0, "-", "plotting"], [32, 0, 0, "-", "reference_genome"], [33, 0, 0, "-", "release"], [34, 0, 0, "-", "slack"], [35, 0, 0, "-", "sparse_mt"], [36, 0, 0, "-", "transcript_annotation"], [37, 0, 0, "-", "vcf"], [38, 0, 0, "-", "vep"]], "gnomad.utils.annotations": [[22, 1, 1, "", "add_gks_va"], [22, 1, 1, "", "add_gks_vrs"], [22, 1, 1, "", "add_variant_type"], [22, 1, 1, "", "age_hists_expr"], [22, 1, 1, "", "agg_by_strata"], [22, 1, 1, "", "annotate_adj"], [22, 1, 1, "", "annotate_allele_info"], [22, 1, 1, "", "annotate_and_index_source_mt_for_sex_ploidy"], [22, 1, 1, "", "annotate_downsamplings"], [22, 1, 1, "", "annotate_freq"], [22, 1, 1, "", "annotate_with_ht"], [22, 1, 1, "", "annotation_type_in_vcf_info"], [22, 1, 1, "", "annotation_type_is_numeric"], [22, 1, 1, "", "bi_allelic_expr"], [22, 1, 1, "", "bi_allelic_site_inbreeding_expr"], [22, 1, 1, "", "build_freq_stratification_list"], [22, 1, 1, "", "compute_freq_by_strata"], [22, 1, 1, "", "create_frequency_bins_expr"], [22, 1, 1, "", "faf_expr"], [22, 1, 1, "", "fill_missing_key_combinations"], [22, 1, 1, "", "fs_from_sb"], [22, 1, 1, "", "gen_anc_faf_max_expr"], [22, 1, 1, "", "generate_freq_group_membership_array"], [22, 1, 1, "", "get_adj_expr"], [22, 1, 1, "", "get_annotations_hists"], [22, 1, 1, "", "get_copy_state_by_sex"], [22, 1, 1, "", "get_gq_dp_adj_expr"], [22, 1, 1, "", "get_het_ab_adj_expr"], [22, 1, 1, "", "get_is_haploid_expr"], [22, 1, 1, "", "get_lowqual_expr"], [22, 1, 1, "", "hemi_expr"], [22, 1, 1, "", "merge_freq_arrays"], [22, 1, 1, "", "merge_histograms"], [22, 1, 1, "", "missing_callstats_expr"], [22, 1, 1, "", "missing_struct_expr"], [22, 1, 1, "", "pab_max_expr"], [22, 1, 1, "", "pop_max_expr"], [22, 1, 1, "", "project_max_expr"], [22, 1, 1, "", "qual_hist_expr"], [22, 1, 1, "", "region_flag_expr"], [22, 1, 1, "", "set_female_y_metrics_to_na_expr"], [22, 1, 1, "", "sor_from_sb"], [22, 1, 1, "", "unphase_call_expr"], [22, 1, 1, "", "update_structured_annotations"]], "gnomad.utils.constraint": [[23, 4, 1, "", "COVERAGE_CUTOFF"], [23, 1, 1, "", "add_gencode_transcript_annotations"], [23, 1, 1, "", "annotate_exploded_vep_for_constraint_groupings"], [23, 1, 1, "", "annotate_mutation_type"], [23, 1, 1, "", "annotate_with_mu"], [23, 1, 1, "", "assemble_constraint_context_ht"], [23, 1, 1, "", "build_coverage_model"], [23, 1, 1, "", "build_models"], [23, 1, 1, "", "build_plateau_models"], [23, 1, 1, "", "calculate_raw_z_score"], [23, 1, 1, "", "calculate_raw_z_score_sd"], [23, 1, 1, "", "collapse_strand"], [23, 1, 1, "", "compute_expected_variants"], [23, 1, 1, "", "compute_pli"], [23, 1, 1, "", "count_variants_by_group"], [23, 1, 1, "", "downsampling_counts_expr"], [23, 1, 1, "", "explode_downsamplings_oe"], [23, 1, 1, "", "get_all_pop_lengths"], [23, 1, 1, "", "get_constraint_flags"], [23, 1, 1, "", "get_constraint_grouping_expr"], [23, 1, 1, "", "get_downsampling_freq_indices"], [23, 1, 1, "", "get_mu_annotation_expr"], [23, 1, 1, "", "oe_aggregation_expr"], [23, 1, 1, "", "oe_confidence_interval"], [23, 1, 1, "", "transform_methylation_level"], [23, 1, 1, "", "trimer_from_heptamer"]], "gnomad.utils.file_utils": [[24, 1, 1, "", "check_file_exists_raise_error"], [24, 1, 1, "", "create_vds"], [24, 1, 1, "", "file_exists"], [24, 1, 1, "", "get_file_stats"], [24, 1, 1, "", "read_list_data"], [24, 1, 1, "", "repartition_for_join"], [24, 1, 1, "", "select_primitives_from_ht"], [24, 1, 1, "", "write_temp_gcs"]], "gnomad.utils.filtering": [[25, 1, 1, "", "add_filters_expr"], [25, 1, 1, "", "combine_functions"], [25, 1, 1, "", "filter_arrays_by_meta"], [25, 1, 1, "", "filter_by_frequency"], [25, 1, 1, "", "filter_by_gencode_intervals"], [25, 1, 1, "", "filter_by_intervals"], [25, 1, 1, "", "filter_by_numeric_expr_range"], [25, 1, 1, "", "filter_for_mu"], [25, 1, 1, "", "filter_gencode_ht"], [25, 1, 1, "", "filter_low_conf_regions"], [25, 1, 1, "", "filter_to_adj"], [25, 1, 1, "", "filter_to_autosomes"], [25, 1, 1, "", "filter_to_clinvar_pathogenic"], [25, 1, 1, "", "filter_to_gencode_cds"], [25, 1, 1, "", "filter_x_nonpar"], [25, 1, 1, "", "filter_y_nonpar"], [25, 1, 1, "", "low_conf_regions_expr"], [25, 1, 1, "", "remove_fields_from_constant"], [25, 1, 1, "", "split_vds_by_strata"], [25, 1, 1, "", "subset_samples_and_variants"]], "gnomad.utils.gen_stats": [[26, 1, 1, "", "from_phred"], [26, 1, 1, "", "get_median_and_mad_expr"], [26, 1, 1, "", "merge_stats_counters_expr"], [26, 1, 1, "", "to_phred"]], "gnomad.utils.intervals": [[28, 1, 1, "", "interval_length"], [28, 1, 1, "", "pad_intervals"], [28, 1, 1, "", "sort_intervals"], [28, 1, 1, "", "union_intervals"]], "gnomad.utils.liftover": [[29, 4, 1, "", "GRCH37_to_GRCH38_CHAIN"], [29, 4, 1, "", "GRCH38_TO_GRCH37_CHAIN"], [29, 1, 1, "", "default_lift_data"], [29, 1, 1, "", "get_liftover_genome"], [29, 1, 1, "", "liftover_expr"], [29, 1, 1, "", "liftover_using_gnomad_map"]], "gnomad.utils.parse": [[30, 1, 1, "", "parse_locus_intervals"], [30, 1, 1, "", "parse_variant"]], "gnomad.utils.plotting": [[31, 1, 1, "", "get_rows_data"], [31, 1, 1, "", "linear_and_log_tabs"], [31, 1, 1, "", "new_show"], [31, 1, 1, "", "pair_plot"], [31, 1, 1, "", "plot_hail_file_metadata"], [31, 1, 1, "", "plot_hail_hist"], [31, 1, 1, "", "plot_hail_hist_both"], [31, 1, 1, "", "plot_hail_hist_cumulative"], [31, 1, 1, "", "plot_multi_hail_hist"], [31, 1, 1, "", "scale_file_sizes"], [31, 1, 1, "", "set_font_size"]], "gnomad.utils.reference_genome": [[32, 1, 1, "", "add_reference_sequence"], [32, 1, 1, "", "get_reference_genome"], [32, 1, 1, "", "get_reference_ht"]], "gnomad.utils.release": [[33, 1, 1, "", "make_faf_index_dict"], [33, 1, 1, "", "make_freq_index_dict"], [33, 1, 1, "", "make_freq_index_dict_from_meta"]], "gnomad.utils.slack": [[34, 2, 1, "", "SlackClient"], [34, 1, 1, "", "slack_notifications"]], "gnomad.utils.slack.SlackClient": [[34, 5, 1, "", "send_file"], [34, 5, 1, "", "send_message"]], "gnomad.utils.sparse_mt": [[35, 1, 1, "", "compute_allele_number_per_ref_site"], [35, 1, 1, "", "compute_coverage_stats"], [35, 1, 1, "", "compute_last_ref_block_end"], [35, 1, 1, "", "compute_stats_per_ref_site"], [35, 1, 1, "", "default_compute_info"], [35, 1, 1, "", "densify_all_reference_sites"], [35, 1, 1, "", "densify_sites"], [35, 1, 1, "", "filter_ref_blocks"], [35, 1, 1, "", "get_allele_number_agg_func"], [35, 1, 1, "", "get_as_info_expr"], [35, 1, 1, "", "get_site_info_expr"], [35, 1, 1, "", "impute_sex_ploidy"], [35, 1, 1, "", "split_info_annotation"], [35, 1, 1, "", "split_lowqual_annotation"]], "gnomad.utils.transcript_annotation": [[36, 4, 1, "", "CELL_LINES"], [36, 4, 1, "", "REPRODUCTIVE_TISSUES"], [36, 1, 1, "", "clean_tissue_name_for_browser"], [36, 1, 1, "", "create_tx_annotation_by_region"], [36, 1, 1, "", "filter_expression_ht_by_tissues"], [36, 1, 1, "", "get_expression_proportion"], [36, 1, 1, "", "get_tissues_to_exclude"], [36, 1, 1, "", "perform_tx_annotation_pipeline"], [36, 1, 1, "", "summarize_transcript_expression"], [36, 1, 1, "", "tissue_expression_ht_to_array"], [36, 1, 1, "", "tx_aggregate_variants"], [36, 1, 1, "", "tx_annotate_variants"], [36, 1, 1, "", "tx_filter_variants_by_csqs"]], "gnomad.utils.vcf": [[37, 4, 1, "", "ALLELE_TYPE_FIELDS"], [37, 4, 1, "", "AS_FIELDS"], [37, 4, 1, "", "AS_VQSR_FIELDS"], [37, 4, 1, "", "ENTRIES"], [37, 4, 1, "", "FAF_POPS"], [37, 4, 1, "", "FORMAT_DICT"], [37, 4, 1, "", "GROUPS"], [37, 4, 1, "", "HISTS"], [37, 4, 1, "", "INFO_DICT"], [37, 4, 1, "", "IN_SILICO_ANNOTATIONS_INFO_DICT"], [37, 4, 1, "", "JOINT_REGION_FLAG_FIELDS"], [37, 4, 1, "", "REGION_FLAG_FIELDS"], [37, 4, 1, "", "RF_FIELDS"], [37, 4, 1, "", "SEXES"], [37, 4, 1, "", "SITE_FIELDS"], [37, 4, 1, "", "SORT_ORDER"], [37, 4, 1, "", "SPARSE_ENTRIES"], [37, 4, 1, "", "VQSR_FIELDS"], [37, 4, 1, "", "VRS_FIELDS_DICT"], [37, 1, 1, "", "add_as_info_dict"], [37, 1, 1, "", "adjust_vcf_incompatible_types"], [37, 1, 1, "", "build_vcf_export_reference"], [37, 1, 1, "", "create_label_groups"], [37, 1, 1, "", "index_globals"], [37, 1, 1, "", "make_combo_header_text"], [37, 1, 1, "", "make_hist_bin_edges_expr"], [37, 1, 1, "", "make_hist_dict"], [37, 1, 1, "", "make_info_dict"], [37, 1, 1, "", "make_label_combos"], [37, 1, 1, "", "make_vcf_filter_dict"], [37, 1, 1, "", "rekey_new_reference"], [37, 1, 1, "", "set_female_y_metrics_to_na"]], "gnomad.utils.vep": [[38, 4, 1, "", "CSQ_CODING"], [38, 4, 1, "", "CSQ_SPLICE"], [38, 4, 1, "", "CURRENT_VEP_VERSION"], [38, 4, 1, "", "LOFTEE_LABELS"], [38, 4, 1, "", "LOF_CSQ_SET"], [38, 4, 1, "", "POSSIBLE_REFS"], [38, 4, 1, "", "VEP_CONFIG_PATH"], [38, 4, 1, "", "VEP_CSQ_FIELDS"], [38, 4, 1, "", "VEP_CSQ_HEADER"], [38, 1, 1, "", "add_most_severe_consequence_to_consequence"], [38, 1, 1, "", "add_most_severe_csq_to_tc_within_vep_root"], [38, 1, 1, "", "explode_by_vep_annotation"], [38, 1, 1, "", "filter_vep_to_canonical_transcripts"], [38, 1, 1, "", "filter_vep_to_gene_list"], [38, 1, 1, "", "filter_vep_to_mane_select_transcripts"], [38, 1, 1, "", "filter_vep_to_synonymous_variants"], [38, 1, 1, "", "filter_vep_transcript_csqs"], [38, 1, 1, "", "filter_vep_transcript_csqs_expr"], [38, 1, 1, "", "get_most_severe_consequence_expr"], [38, 1, 1, "", "get_most_severe_consequence_for_summary"], [38, 1, 1, "", "get_vep_context"], [38, 1, 1, "", "get_vep_help"], [38, 1, 1, "", "process_consequences"], [38, 1, 1, "", "vep_or_lookup_vep"], [38, 1, 1, "", "vep_struct_to_csq"]], "gnomad.variant_qc": [[39, 0, 0, "-", "evaluation"], [41, 0, 0, "-", "ld"], [42, 0, 0, "-", "pipeline"], [43, 0, 0, "-", "random_forest"], [44, 0, 0, "-", "training"]], "gnomad.variant_qc.evaluation": [[39, 1, 1, "", "add_rank"], [39, 1, 1, "", "compute_binned_truth_sample_concordance"], [39, 1, 1, "", "compute_grouped_binned_ht"], [39, 1, 1, "", "compute_ranked_bin"], [39, 1, 1, "", "create_truth_sample_ht"]], "gnomad.variant_qc.ld": [[41, 1, 1, "", "get_r_for_pair_of_variants"], [41, 1, 1, "", "get_r_human_readable"], [41, 1, 1, "", "get_r_within_gene"], [41, 1, 1, "", "get_r_within_gene_in_pop"]], "gnomad.variant_qc.pipeline": [[42, 1, 1, "", "create_binned_ht"], [42, 1, 1, "", "generate_sib_stats"], [42, 1, 1, "", "generate_trio_stats"], [42, 1, 1, "", "score_bin_agg"], [42, 1, 1, "", "train_rf_model"]], "gnomad.variant_qc.random_forest": [[43, 1, 1, "", "apply_rf_model"], [43, 1, 1, "", "check_ht_fields_for_spark"], [43, 1, 1, "", "get_columns_quantiles"], [43, 1, 1, "", "get_features_importance"], [43, 1, 1, "", "get_labels"], [43, 1, 1, "", "get_rf_runs"], [43, 1, 1, "", "get_run_data"], [43, 1, 1, "", "ht_to_rf_df"], [43, 1, 1, "", "load_model"], [43, 1, 1, "", "median_impute_features"], [43, 1, 1, "", "pretty_print_runs"], [43, 1, 1, "", "run_rf_test"], [43, 1, 1, "", "save_model"], [43, 1, 1, "", "test_model"], [43, 1, 1, "", "train_rf"]], "gnomad.variant_qc.training": [[44, 1, 1, "", "sample_training_examples"]]}, "objtypes": {"0": "py:module", "1": "py:function", "2": "py:class", "3": "py:attribute", "4": "py:data", "5": "py:method", "6": "py:property", "7": "py:exception"}, "objnames": {"0": ["py", "module", "Python module"], "1": ["py", "function", "Python function"], "2": ["py", "class", "Python class"], "3": ["py", "attribute", "Python attribute"], "4": ["py", "data", "Python data"], "5": ["py", "method", "Python method"], "6": ["py", "property", "Python property"], "7": ["py", "exception", "Python exception"]}, "titleterms": {"gnomad": [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 48], "assess": [0, 1, 2], "summary_stat": 1, "validity_check": 2, "resourc": [4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 49], "config": 4, "grch37": [5, 6, 7, 8], "gnomad_ld": 6, "reference_data": [8, 11], "grch38": [9, 10, 11], "import_resourc": 12, "resource_util": 14, "sample_qc": [15, 16, 17, 18, 19, 20, 21], "ancestri": 15, "filter": [16, 25], "pipelin": [18, 42], "platform": 19, "related": 20, "sex": 21, "util": [22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38], "annot": 22, "constraint": 23, "file_util": 24, "gen_stat": 26, "interv": 28, "liftov": 29, "pars": 30, "plot": 31, "reference_genom": 32, "releas": 33, "slack": 34, "sparse_mt": 35, "transcript_annot": 36, "input": 36, "hail": 36, "tabl": 36, "output": 36, "vcf": 37, "vep": [38, 46], "variant_qc": [39, 40, 41, 42, 43, 44], "evalu": 39, "ld": 41, "random_forest": 43, "train": 44, "exampl": 45, "variant": 46, "effect": 46, "predictor": 46, "get": 47, "start": 47, "content": 48, "sourc": 49, "custom": 49, "environ": 49, "configur": 49}, "envversion": {"sphinx.domains.c": 2, "sphinx.domains.changeset": 1, "sphinx.domains.citation": 1, "sphinx.domains.cpp": 8, "sphinx.domains.index": 1, "sphinx.domains.javascript": 2, "sphinx.domains.math": 2, "sphinx.domains.python": 3, "sphinx.domains.rst": 2, "sphinx.domains.std": 2, "sphinx.ext.intersphinx": 1, "sphinx.ext.viewcode": 1, "sphinx": 57}, "alltitles": {"gnomad.assessment": [[0, "gnomad-assessment"]], "gnomad.assessment.summary_stats": [[1, "gnomad-assessment-summary-stats"]], "gnomad.assessment.validity_checks": [[2, "gnomad-assessment-validity-checks"]], "gnomad": [[3, "gnomad"], [48, "gnomad"]], "gnomad.resources.config": [[4, "gnomad-resources-config"]], "gnomad.resources.grch37.gnomad": [[5, "gnomad-resources-grch37-gnomad"]], "gnomad.resources.grch37.gnomad_ld": [[6, "gnomad-resources-grch37-gnomad-ld"]], "gnomad.resources.grch37": [[7, "gnomad-resources-grch37"]], "gnomad.resources.grch37.reference_data": [[8, "gnomad-resources-grch37-reference-data"]], "gnomad.resources.grch38.gnomad": [[9, "gnomad-resources-grch38-gnomad"]], "gnomad.resources.grch38": [[10, "gnomad-resources-grch38"]], "gnomad.resources.grch38.reference_data": [[11, "gnomad-resources-grch38-reference-data"]], "gnomad.resources.import_resources": [[12, "gnomad-resources-import-resources"]], "gnomad.resources": [[13, "gnomad-resources"]], "gnomad.resources.resource_utils": [[14, "gnomad-resources-resource-utils"]], "gnomad.sample_qc.ancestry": [[15, "gnomad-sample-qc-ancestry"]], "gnomad.sample_qc.filtering": [[16, "gnomad-sample-qc-filtering"]], "gnomad.sample_qc": [[17, "gnomad-sample-qc"]], "gnomad.sample_qc.pipeline": [[18, "gnomad-sample-qc-pipeline"]], "gnomad.sample_qc.platform": [[19, "gnomad-sample-qc-platform"]], "gnomad.sample_qc.relatedness": [[20, "gnomad-sample-qc-relatedness"]], "gnomad.sample_qc.sex": [[21, "gnomad-sample-qc-sex"]], "gnomad.utils.annotations": [[22, "gnomad-utils-annotations"]], "gnomad.utils.constraint": [[23, "gnomad-utils-constraint"]], "gnomad.utils.file_utils": [[24, "gnomad-utils-file-utils"]], "gnomad.utils.filtering": [[25, "gnomad-utils-filtering"]], "gnomad.utils.gen_stats": [[26, "gnomad-utils-gen-stats"]], "gnomad.utils": [[27, "gnomad-utils"]], "gnomad.utils.intervals": [[28, "gnomad-utils-intervals"]], "gnomad.utils.liftover": [[29, "gnomad-utils-liftover"]], "gnomad.utils.parse": [[30, "gnomad-utils-parse"]], "gnomad.utils.plotting": [[31, "gnomad-utils-plotting"]], "gnomad.utils.reference_genome": [[32, "gnomad-utils-reference-genome"]], "gnomad.utils.release": [[33, "gnomad-utils-release"]], "gnomad.utils.slack": [[34, "gnomad-utils-slack"]], "gnomad.utils.sparse_mt": [[35, "gnomad-utils-sparse-mt"]], "gnomad.utils.transcript_annotation": [[36, "gnomad-utils-transcript-annotation"]], "Input Hail Table": [[36, "id1"]], "Output Hail Table": [[36, "id2"]], "gnomad.utils.vcf": [[37, "gnomad-utils-vcf"]], "gnomad.utils.vep": [[38, "gnomad-utils-vep"]], "gnomad.variant_qc.evaluation": [[39, "gnomad-variant-qc-evaluation"]], "gnomad.variant_qc": [[40, "gnomad-variant-qc"]], "gnomad.variant_qc.ld": [[41, "gnomad-variant-qc-ld"]], "gnomad.variant_qc.pipeline": [[42, "gnomad-variant-qc-pipeline"]], "gnomad.variant_qc.random_forest": [[43, "gnomad-variant-qc-random-forest"]], "gnomad.variant_qc.training": [[44, "gnomad-variant-qc-training"]], "Examples": [[45, "examples"]], "Variant Effect Predictor (VEP)": [[46, "variant-effect-predictor-vep"]], "Getting Started": [[47, "getting-started"]], "Contents": [[48, "contents"]], "Resource Sources": [[49, "resource-sources"]], "Custom Sources": [[49, "custom-sources"]], "Environment Configuration": [[49, "environment-configuration"]]}, "indexentries": {"default_generate_gene_lof_matrix() (in module gnomad.assessment.summary_stats)": [[1, "gnomad.assessment.summary_stats.default_generate_gene_lof_matrix"]], "default_generate_gene_lof_summary() (in module gnomad.assessment.summary_stats)": [[1, "gnomad.assessment.summary_stats.default_generate_gene_lof_summary"]], "freq_bin_expr() (in module gnomad.assessment.summary_stats)": [[1, "gnomad.assessment.summary_stats.freq_bin_expr"]], "generate_filter_combinations() (in module gnomad.assessment.summary_stats)": [[1, "gnomad.assessment.summary_stats.generate_filter_combinations"]], "get_an_criteria() (in module gnomad.assessment.summary_stats)": [[1, "gnomad.assessment.summary_stats.get_an_criteria"]], "get_het_hom_summary_dict() (in module gnomad.assessment.summary_stats)": [[1, "gnomad.assessment.summary_stats.get_het_hom_summary_dict"]], "get_summary_ac_dict() (in module gnomad.assessment.summary_stats)": [[1, "gnomad.assessment.summary_stats.get_summary_ac_dict"]], "get_summary_counts() (in module gnomad.assessment.summary_stats)": [[1, "gnomad.assessment.summary_stats.get_summary_counts"]], "get_summary_counts_dict() (in module gnomad.assessment.summary_stats)": [[1, "gnomad.assessment.summary_stats.get_summary_counts_dict"]], "get_summary_stats_csq_filter_expr() (in module gnomad.assessment.summary_stats)": [[1, "gnomad.assessment.summary_stats.get_summary_stats_csq_filter_expr"]], "get_summary_stats_filter_group_meta() (in module gnomad.assessment.summary_stats)": [[1, "gnomad.assessment.summary_stats.get_summary_stats_filter_group_meta"]], "get_summary_stats_variant_filter_expr() (in module gnomad.assessment.summary_stats)": [[1, "gnomad.assessment.summary_stats.get_summary_stats_variant_filter_expr"]], "get_tx_expression_expr() (in module gnomad.assessment.summary_stats)": [[1, "gnomad.assessment.summary_stats.get_tx_expression_expr"]], "gnomad.assessment.summary_stats": [[1, "module-gnomad.assessment.summary_stats"]], "module": [[1, "module-gnomad.assessment.summary_stats"], [2, "module-gnomad.assessment.validity_checks"], [4, "module-gnomad.resources.config"], [5, "module-gnomad.resources.grch37.gnomad"], [6, "module-gnomad.resources.grch37.gnomad_ld"], [8, "module-gnomad.resources.grch37.reference_data"], [9, "module-gnomad.resources.grch38.gnomad"], [11, "module-gnomad.resources.grch38.reference_data"], [12, "module-gnomad.resources.import_resources"], [14, "module-gnomad.resources.resource_utils"], [15, "module-gnomad.sample_qc.ancestry"], [16, "module-gnomad.sample_qc.filtering"], [18, "module-gnomad.sample_qc.pipeline"], [19, "module-gnomad.sample_qc.platform"], [20, "module-gnomad.sample_qc.relatedness"], [21, "module-gnomad.sample_qc.sex"], [22, "module-gnomad.utils.annotations"], [23, "module-gnomad.utils.constraint"], [24, "module-gnomad.utils.file_utils"], [25, "module-gnomad.utils.filtering"], [26, "module-gnomad.utils.gen_stats"], [28, "module-gnomad.utils.intervals"], [29, "module-gnomad.utils.liftover"], [30, "module-gnomad.utils.parse"], [31, "module-gnomad.utils.plotting"], [32, "module-gnomad.utils.reference_genome"], [33, "module-gnomad.utils.release"], [34, "module-gnomad.utils.slack"], [35, "module-gnomad.utils.sparse_mt"], [36, "module-gnomad.utils.transcript_annotation"], [37, "module-gnomad.utils.vcf"], [38, "module-gnomad.utils.vep"], [39, "module-gnomad.variant_qc.evaluation"], [41, "module-gnomad.variant_qc.ld"], [42, "module-gnomad.variant_qc.pipeline"], [43, "module-gnomad.variant_qc.random_forest"], [44, "module-gnomad.variant_qc.training"]], "check_global_and_row_annot_lengths() (in module gnomad.assessment.validity_checks)": [[2, "gnomad.assessment.validity_checks.check_global_and_row_annot_lengths"]], "check_missingness_of_struct() (in module gnomad.assessment.validity_checks)": [[2, "gnomad.assessment.validity_checks.check_missingness_of_struct"]], "check_raw_and_adj_callstats() (in module gnomad.assessment.validity_checks)": [[2, "gnomad.assessment.validity_checks.check_raw_and_adj_callstats"]], "check_sex_chr_metrics() (in module gnomad.assessment.validity_checks)": [[2, "gnomad.assessment.validity_checks.check_sex_chr_metrics"]], "compare_row_counts() (in module gnomad.assessment.validity_checks)": [[2, "gnomad.assessment.validity_checks.compare_row_counts"]], "compare_subset_freqs() (in module gnomad.assessment.validity_checks)": [[2, "gnomad.assessment.validity_checks.compare_subset_freqs"]], "compute_missingness() (in module gnomad.assessment.validity_checks)": [[2, "gnomad.assessment.validity_checks.compute_missingness"]], "count_vep_annotated_variants_per_interval() (in module gnomad.assessment.validity_checks)": [[2, "gnomad.assessment.validity_checks.count_vep_annotated_variants_per_interval"]], "flatten_missingness_struct() (in module gnomad.assessment.validity_checks)": [[2, "gnomad.assessment.validity_checks.flatten_missingness_struct"]], "generic_field_check() (in module gnomad.assessment.validity_checks)": [[2, "gnomad.assessment.validity_checks.generic_field_check"]], "generic_field_check_loop() (in module gnomad.assessment.validity_checks)": [[2, "gnomad.assessment.validity_checks.generic_field_check_loop"]], "gnomad.assessment.validity_checks": [[2, "module-gnomad.assessment.validity_checks"]], "make_filters_expr_dict() (in module gnomad.assessment.validity_checks)": [[2, "gnomad.assessment.validity_checks.make_filters_expr_dict"]], "make_group_sum_expr_dict() (in module gnomad.assessment.validity_checks)": [[2, "gnomad.assessment.validity_checks.make_group_sum_expr_dict"]], "pprint_global_anns() (in module gnomad.assessment.validity_checks)": [[2, "gnomad.assessment.validity_checks.pprint_global_anns"]], "sum_group_callstats() (in module gnomad.assessment.validity_checks)": [[2, "gnomad.assessment.validity_checks.sum_group_callstats"]], "summarize_variant_filters() (in module gnomad.assessment.validity_checks)": [[2, "gnomad.assessment.validity_checks.summarize_variant_filters"]], "summarize_variants() (in module gnomad.assessment.validity_checks)": [[2, "gnomad.assessment.validity_checks.summarize_variants"]], "unfurl_array_annotations() (in module gnomad.assessment.validity_checks)": [[2, "gnomad.assessment.validity_checks.unfurl_array_annotations"]], "validate_release_t() (in module gnomad.assessment.validity_checks)": [[2, "gnomad.assessment.validity_checks.validate_release_t"]], "vcf_field_check() (in module gnomad.assessment.validity_checks)": [[2, "gnomad.assessment.validity_checks.vcf_field_check"]], "azure_open_datasets (gnomad.resources.config.gnomadpublicresourcesource attribute)": [[4, "gnomad.resources.config.GnomadPublicResourceSource.AZURE_OPEN_DATASETS"]], "gnomad (gnomad.resources.config.gnomadpublicresourcesource attribute)": [[4, "gnomad.resources.config.GnomadPublicResourceSource.GNOMAD"]], "google_cloud_public_datasets (gnomad.resources.config.gnomadpublicresourcesource attribute)": [[4, "gnomad.resources.config.GnomadPublicResourceSource.GOOGLE_CLOUD_PUBLIC_DATASETS"]], "gnomadpublicresourcesource (class in gnomad.resources.config)": [[4, "gnomad.resources.config.GnomadPublicResourceSource"]], "registry_of_open_data_on_aws (gnomad.resources.config.gnomadpublicresourcesource attribute)": [[4, "gnomad.resources.config.GnomadPublicResourceSource.REGISTRY_OF_OPEN_DATA_ON_AWS"]], "get_default_public_resource_source() (in module gnomad.resources.config)": [[4, "gnomad.resources.config.get_default_public_resource_source"]], "gnomad.resources.config": [[4, "module-gnomad.resources.config"]], "browser_gene() (in module gnomad.resources.grch37.gnomad)": [[5, "gnomad.resources.grch37.gnomad.browser_gene"]], "constraint() (in module gnomad.resources.grch37.gnomad)": [[5, "gnomad.resources.grch37.gnomad.constraint"]], "coverage() (in module gnomad.resources.grch37.gnomad)": [[5, "gnomad.resources.grch37.gnomad.coverage"]], "gnomad.resources.grch37.gnomad": [[5, "module-gnomad.resources.grch37.gnomad"]], "liftover() (in module gnomad.resources.grch37.gnomad)": [[5, "gnomad.resources.grch37.gnomad.liftover"]], "mnv() (in module gnomad.resources.grch37.gnomad)": [[5, "gnomad.resources.grch37.gnomad.mnv"]], "pext() (in module gnomad.resources.grch37.gnomad)": [[5, "gnomad.resources.grch37.gnomad.pext"]], "public_pca_loadings() (in module gnomad.resources.grch37.gnomad)": [[5, "gnomad.resources.grch37.gnomad.public_pca_loadings"]], "public_release() (in module gnomad.resources.grch37.gnomad)": [[5, "gnomad.resources.grch37.gnomad.public_release"]], "release_vcf_path() (in module gnomad.resources.grch37.gnomad)": [[5, "gnomad.resources.grch37.gnomad.release_vcf_path"]], "gnomad.resources.grch37.gnomad_ld": [[6, "module-gnomad.resources.grch37.gnomad_ld"]], "ld_index() (in module gnomad.resources.grch37.gnomad_ld)": [[6, "gnomad.resources.grch37.gnomad_ld.ld_index"]], "ld_matrix() (in module gnomad.resources.grch37.gnomad_ld)": [[6, "gnomad.resources.grch37.gnomad_ld.ld_matrix"]], "ld_scores() (in module gnomad.resources.grch37.gnomad_ld)": [[6, "gnomad.resources.grch37.gnomad_ld.ld_scores"]], "get_truth_ht() (in module gnomad.resources.grch37.reference_data)": [[8, "gnomad.resources.grch37.reference_data.get_truth_ht"]], "gnomad.resources.grch37.reference_data": [[8, "module-gnomad.resources.grch37.reference_data"]], "transform_grch37_methylation() (in module gnomad.resources.grch37.reference_data)": [[8, "gnomad.resources.grch37.reference_data.transform_grch37_methylation"]], "cohorts_with_pop_stored_as_subpop (in module gnomad.resources.grch38.gnomad)": [[9, "gnomad.resources.grch38.gnomad.COHORTS_WITH_POP_STORED_AS_SUBPOP"]], "downsamplings (in module gnomad.resources.grch38.gnomad)": [[9, "gnomad.resources.grch38.gnomad.DOWNSAMPLINGS"]], "groups (in module gnomad.resources.grch38.gnomad)": [[9, "gnomad.resources.grch38.gnomad.GROUPS"]], "hgdp_pops (in module gnomad.resources.grch38.gnomad)": [[9, "gnomad.resources.grch38.gnomad.HGDP_POPS"]], "pops (in module gnomad.resources.grch38.gnomad)": [[9, "gnomad.resources.grch38.gnomad.POPS"]], "pops_to_remove_for_popmax (in module gnomad.resources.grch38.gnomad)": [[9, "gnomad.resources.grch38.gnomad.POPS_TO_REMOVE_FOR_POPMAX"]], "sexes (in module gnomad.resources.grch38.gnomad)": [[9, "gnomad.resources.grch38.gnomad.SEXES"]], "subsets (in module gnomad.resources.grch38.gnomad)": [[9, "gnomad.resources.grch38.gnomad.SUBSETS"]], "tgp_pops (in module gnomad.resources.grch38.gnomad)": [[9, "gnomad.resources.grch38.gnomad.TGP_POPS"]], "tgp_pop_names (in module gnomad.resources.grch38.gnomad)": [[9, "gnomad.resources.grch38.gnomad.TGP_POP_NAMES"]], "add_grpmaxfaf95_v4() (in module gnomad.resources.grch38.gnomad)": [[9, "gnomad.resources.grch38.gnomad.add_grpMaxFAF95_v4"]], "all_sites_an() (in module gnomad.resources.grch38.gnomad)": [[9, "gnomad.resources.grch38.gnomad.all_sites_an"]], "browser_gene() (in module gnomad.resources.grch38.gnomad)": [[9, "gnomad.resources.grch38.gnomad.browser_gene"]], "browser_variant() (in module gnomad.resources.grch38.gnomad)": [[9, "gnomad.resources.grch38.gnomad.browser_variant"]], "constraint() (in module gnomad.resources.grch38.gnomad)": [[9, "gnomad.resources.grch38.gnomad.constraint"]], "coverage() (in module gnomad.resources.grch38.gnomad)": [[9, "gnomad.resources.grch38.gnomad.coverage"]], "coverage_tsv_path() (in module gnomad.resources.grch38.gnomad)": [[9, "gnomad.resources.grch38.gnomad.coverage_tsv_path"]], "gnomad.resources.grch38.gnomad": [[9, "module-gnomad.resources.grch38.gnomad"]], "gnomad_gks() (in module gnomad.resources.grch38.gnomad)": [[9, "gnomad.resources.grch38.gnomad.gnomad_gks"]], "pext() (in module gnomad.resources.grch38.gnomad)": [[9, "gnomad.resources.grch38.gnomad.pext"]], "public_release() (in module gnomad.resources.grch38.gnomad)": [[9, "gnomad.resources.grch38.gnomad.public_release"]], "release_vcf_path() (in module gnomad.resources.grch38.gnomad)": [[9, "gnomad.resources.grch38.gnomad.release_vcf_path"]], "get_truth_ht() (in module gnomad.resources.grch38.reference_data)": [[11, "gnomad.resources.grch38.reference_data.get_truth_ht"]], "gnomad.resources.grch38.reference_data": [[11, "module-gnomad.resources.grch38.reference_data"]], "transform_grch38_methylation() (in module gnomad.resources.grch38.reference_data)": [[11, "gnomad.resources.grch38.reference_data.transform_grch38_methylation"]], "get_module_importable_resources() (in module gnomad.resources.import_resources)": [[12, "gnomad.resources.import_resources.get_module_importable_resources"]], "get_resources_descriptions() (in module gnomad.resources.import_resources)": [[12, "gnomad.resources.import_resources.get_resources_descriptions"]], "gnomad.resources.import_resources": [[12, "module-gnomad.resources.import_resources"]], "main() (in module gnomad.resources.import_resources)": [[12, "gnomad.resources.import_resources.main"]], "baseresource (class in gnomad.resources.resource_utils)": [[14, "gnomad.resources.resource_utils.BaseResource"]], "baseversionedresource (class in gnomad.resources.resource_utils)": [[14, "gnomad.resources.resource_utils.BaseVersionedResource"]], "blockmatrixresource (class in gnomad.resources.resource_utils)": [[14, "gnomad.resources.resource_utils.BlockMatrixResource"]], "dataexception": [[14, "gnomad.resources.resource_utils.DataException"]], "expressionresource (class in gnomad.resources.resource_utils)": [[14, "gnomad.resources.resource_utils.ExpressionResource"]], "gnomad_public_buckets (in module gnomad.resources.resource_utils)": [[14, "gnomad.resources.resource_utils.GNOMAD_PUBLIC_BUCKETS"]], "gnomadpublicblockmatrixresource (class in gnomad.resources.resource_utils)": [[14, "gnomad.resources.resource_utils.GnomadPublicBlockMatrixResource"]], "gnomadpublicmatrixtableresource (class in gnomad.resources.resource_utils)": [[14, "gnomad.resources.resource_utils.GnomadPublicMatrixTableResource"]], "gnomadpublicpedigreeresource (class in gnomad.resources.resource_utils)": [[14, "gnomad.resources.resource_utils.GnomadPublicPedigreeResource"]], "gnomadpublicresource (class in gnomad.resources.resource_utils)": [[14, "gnomad.resources.resource_utils.GnomadPublicResource"]], "gnomadpublictableresource (class in gnomad.resources.resource_utils)": [[14, "gnomad.resources.resource_utils.GnomadPublicTableResource"]], "matrixtableresource (class in gnomad.resources.resource_utils)": [[14, "gnomad.resources.resource_utils.MatrixTableResource"]], "pedigreeresource (class in gnomad.resources.resource_utils)": [[14, "gnomad.resources.resource_utils.PedigreeResource"]], "resourcenotavailable": [[14, "gnomad.resources.resource_utils.ResourceNotAvailable"]], "tableresource (class in gnomad.resources.resource_utils)": [[14, "gnomad.resources.resource_utils.TableResource"]], "variantdatasetresource (class in gnomad.resources.resource_utils)": [[14, "gnomad.resources.resource_utils.VariantDatasetResource"]], "versionedblockmatrixresource (class in gnomad.resources.resource_utils)": [[14, "gnomad.resources.resource_utils.VersionedBlockMatrixResource"]], "versionedmatrixtableresource (class in gnomad.resources.resource_utils)": [[14, "gnomad.resources.resource_utils.VersionedMatrixTableResource"]], "versionedpedigreeresource (class in gnomad.resources.resource_utils)": [[14, "gnomad.resources.resource_utils.VersionedPedigreeResource"]], "versionedtableresource (class in gnomad.resources.resource_utils)": [[14, "gnomad.resources.resource_utils.VersionedTableResource"]], "versionedvariantdatasetresource (class in gnomad.resources.resource_utils)": [[14, "gnomad.resources.resource_utils.VersionedVariantDatasetResource"]], "bm() (gnomad.resources.resource_utils.blockmatrixresource method)": [[14, "gnomad.resources.resource_utils.BlockMatrixResource.bm"]], "bm() (gnomad.resources.resource_utils.gnomadpublicblockmatrixresource method)": [[14, "gnomad.resources.resource_utils.GnomadPublicBlockMatrixResource.bm"]], "default_version (gnomad.resources.resource_utils.baseversionedresource attribute)": [[14, "gnomad.resources.resource_utils.BaseVersionedResource.default_version"]], "expected_file_extensions (gnomad.resources.resource_utils.baseresource attribute)": [[14, "gnomad.resources.resource_utils.BaseResource.expected_file_extensions"]], "expected_file_extensions (gnomad.resources.resource_utils.blockmatrixresource attribute)": [[14, "gnomad.resources.resource_utils.BlockMatrixResource.expected_file_extensions"]], "expected_file_extensions (gnomad.resources.resource_utils.expressionresource attribute)": [[14, "gnomad.resources.resource_utils.ExpressionResource.expected_file_extensions"]], "expected_file_extensions (gnomad.resources.resource_utils.matrixtableresource attribute)": [[14, "gnomad.resources.resource_utils.MatrixTableResource.expected_file_extensions"]], "expected_file_extensions (gnomad.resources.resource_utils.pedigreeresource attribute)": [[14, "gnomad.resources.resource_utils.PedigreeResource.expected_file_extensions"]], "expected_file_extensions (gnomad.resources.resource_utils.tableresource attribute)": [[14, "gnomad.resources.resource_utils.TableResource.expected_file_extensions"]], "expected_file_extensions (gnomad.resources.resource_utils.variantdatasetresource attribute)": [[14, "gnomad.resources.resource_utils.VariantDatasetResource.expected_file_extensions"]], "gnomad.resources.resource_utils": [[14, "module-gnomad.resources.resource_utils"]], "he() (gnomad.resources.resource_utils.expressionresource method)": [[14, "gnomad.resources.resource_utils.ExpressionResource.he"]], "ht() (gnomad.resources.resource_utils.gnomadpublicpedigreeresource method)": [[14, "gnomad.resources.resource_utils.GnomadPublicPedigreeResource.ht"]], "ht() (gnomad.resources.resource_utils.gnomadpublictableresource method)": [[14, "gnomad.resources.resource_utils.GnomadPublicTableResource.ht"]], "ht() (gnomad.resources.resource_utils.pedigreeresource method)": [[14, "gnomad.resources.resource_utils.PedigreeResource.ht"]], "ht() (gnomad.resources.resource_utils.tableresource method)": [[14, "gnomad.resources.resource_utils.TableResource.ht"]], "import_gencode() (in module gnomad.resources.resource_utils)": [[14, "gnomad.resources.resource_utils.import_gencode"]], "import_resource() (gnomad.resources.resource_utils.baseresource method)": [[14, "gnomad.resources.resource_utils.BaseResource.import_resource"]], "import_resource() (gnomad.resources.resource_utils.blockmatrixresource method)": [[14, "gnomad.resources.resource_utils.BlockMatrixResource.import_resource"]], "import_resource() (gnomad.resources.resource_utils.expressionresource method)": [[14, "gnomad.resources.resource_utils.ExpressionResource.import_resource"]], "import_resource() (gnomad.resources.resource_utils.matrixtableresource method)": [[14, "gnomad.resources.resource_utils.MatrixTableResource.import_resource"]], "import_resource() (gnomad.resources.resource_utils.pedigreeresource method)": [[14, "gnomad.resources.resource_utils.PedigreeResource.import_resource"]], "import_resource() (gnomad.resources.resource_utils.tableresource method)": [[14, "gnomad.resources.resource_utils.TableResource.import_resource"]], "import_resource() (gnomad.resources.resource_utils.variantdatasetresource method)": [[14, "gnomad.resources.resource_utils.VariantDatasetResource.import_resource"]], "import_sites_vcf() (in module gnomad.resources.resource_utils)": [[14, "gnomad.resources.resource_utils.import_sites_vcf"]], "is_resource_available() (gnomad.resources.resource_utils.gnomadpublicresource method)": [[14, "gnomad.resources.resource_utils.GnomadPublicResource.is_resource_available"]], "mt() (gnomad.resources.resource_utils.gnomadpublicmatrixtableresource method)": [[14, "gnomad.resources.resource_utils.GnomadPublicMatrixTableResource.mt"]], "mt() (gnomad.resources.resource_utils.matrixtableresource method)": [[14, "gnomad.resources.resource_utils.MatrixTableResource.mt"]], "path (gnomad.resources.resource_utils.baseresource property)": [[14, "gnomad.resources.resource_utils.BaseResource.path"]], "pedigree() (gnomad.resources.resource_utils.gnomadpublicpedigreeresource method)": [[14, "gnomad.resources.resource_utils.GnomadPublicPedigreeResource.pedigree"]], "pedigree() (gnomad.resources.resource_utils.pedigreeresource method)": [[14, "gnomad.resources.resource_utils.PedigreeResource.pedigree"]], "resource_class (gnomad.resources.resource_utils.baseversionedresource attribute)": [[14, "gnomad.resources.resource_utils.BaseVersionedResource.resource_class"]], "resource_class (gnomad.resources.resource_utils.versionedblockmatrixresource attribute)": [[14, "gnomad.resources.resource_utils.VersionedBlockMatrixResource.resource_class"]], "resource_class (gnomad.resources.resource_utils.versionedmatrixtableresource attribute)": [[14, "gnomad.resources.resource_utils.VersionedMatrixTableResource.resource_class"]], "resource_class (gnomad.resources.resource_utils.versionedpedigreeresource attribute)": [[14, "gnomad.resources.resource_utils.VersionedPedigreeResource.resource_class"]], "resource_class (gnomad.resources.resource_utils.versionedtableresource attribute)": [[14, "gnomad.resources.resource_utils.VersionedTableResource.resource_class"]], "resource_class (gnomad.resources.resource_utils.versionedvariantdatasetresource attribute)": [[14, "gnomad.resources.resource_utils.VersionedVariantDatasetResource.resource_class"]], "vds() (gnomad.resources.resource_utils.variantdatasetresource method)": [[14, "gnomad.resources.resource_utils.VariantDatasetResource.vds"]], "versions (gnomad.resources.resource_utils.baseversionedresource attribute)": [[14, "gnomad.resources.resource_utils.BaseVersionedResource.versions"]], "apply_onnx_classification_model() (in module gnomad.sample_qc.ancestry)": [[15, "gnomad.sample_qc.ancestry.apply_onnx_classification_model"]], "apply_sklearn_classification_model() (in module gnomad.sample_qc.ancestry)": [[15, "gnomad.sample_qc.ancestry.apply_sklearn_classification_model"]], "assign_population_pcs() (in module gnomad.sample_qc.ancestry)": [[15, "gnomad.sample_qc.ancestry.assign_population_pcs"]], "convert_sklearn_rf_to_onnx() (in module gnomad.sample_qc.ancestry)": [[15, "gnomad.sample_qc.ancestry.convert_sklearn_rf_to_onnx"]], "gnomad.sample_qc.ancestry": [[15, "module-gnomad.sample_qc.ancestry"]], "pc_project() (in module gnomad.sample_qc.ancestry)": [[15, "gnomad.sample_qc.ancestry.pc_project"]], "run_pca_with_relateds() (in module gnomad.sample_qc.ancestry)": [[15, "gnomad.sample_qc.ancestry.run_pca_with_relateds"]], "compute_qc_metrics_residuals() (in module gnomad.sample_qc.filtering)": [[16, "gnomad.sample_qc.filtering.compute_qc_metrics_residuals"]], "compute_stratified_metrics_filter() (in module gnomad.sample_qc.filtering)": [[16, "gnomad.sample_qc.filtering.compute_stratified_metrics_filter"]], "compute_stratified_sample_qc() (in module gnomad.sample_qc.filtering)": [[16, "gnomad.sample_qc.filtering.compute_stratified_sample_qc"]], "determine_nearest_neighbors() (in module gnomad.sample_qc.filtering)": [[16, "gnomad.sample_qc.filtering.determine_nearest_neighbors"]], "gnomad.sample_qc.filtering": [[16, "module-gnomad.sample_qc.filtering"]], "merge_sample_qc_expr() (in module gnomad.sample_qc.filtering)": [[16, "gnomad.sample_qc.filtering.merge_sample_qc_expr"]], "annotate_sex() (in module gnomad.sample_qc.pipeline)": [[18, "gnomad.sample_qc.pipeline.annotate_sex"]], "filter_rows_for_qc() (in module gnomad.sample_qc.pipeline)": [[18, "gnomad.sample_qc.pipeline.filter_rows_for_qc"]], "get_qc_mt() (in module gnomad.sample_qc.pipeline)": [[18, "gnomad.sample_qc.pipeline.get_qc_mt"]], "gnomad.sample_qc.pipeline": [[18, "module-gnomad.sample_qc.pipeline"]], "infer_sex_karyotype() (in module gnomad.sample_qc.pipeline)": [[18, "gnomad.sample_qc.pipeline.infer_sex_karyotype"]], "assign_platform_from_pcs() (in module gnomad.sample_qc.platform)": [[19, "gnomad.sample_qc.platform.assign_platform_from_pcs"]], "compute_callrate_mt() (in module gnomad.sample_qc.platform)": [[19, "gnomad.sample_qc.platform.compute_callrate_mt"]], "gnomad.sample_qc.platform": [[19, "module-gnomad.sample_qc.platform"]], "run_platform_pca() (in module gnomad.sample_qc.platform)": [[19, "gnomad.sample_qc.platform.run_platform_pca"]], "ambiguous_relationship (in module gnomad.sample_qc.relatedness)": [[20, "gnomad.sample_qc.relatedness.AMBIGUOUS_RELATIONSHIP"]], "duplicate_or_twins (in module gnomad.sample_qc.relatedness)": [[20, "gnomad.sample_qc.relatedness.DUPLICATE_OR_TWINS"]], "parent_child (in module gnomad.sample_qc.relatedness)": [[20, "gnomad.sample_qc.relatedness.PARENT_CHILD"]], "second_degree_relatives (in module gnomad.sample_qc.relatedness)": [[20, "gnomad.sample_qc.relatedness.SECOND_DEGREE_RELATIVES"]], "siblings (in module gnomad.sample_qc.relatedness)": [[20, "gnomad.sample_qc.relatedness.SIBLINGS"]], "unrelated (in module gnomad.sample_qc.relatedness)": [[20, "gnomad.sample_qc.relatedness.UNRELATED"]], "calculate_de_novo_post_prob() (in module gnomad.sample_qc.relatedness)": [[20, "gnomad.sample_qc.relatedness.calculate_de_novo_post_prob"]], "compute_related_samples_to_drop() (in module gnomad.sample_qc.relatedness)": [[20, "gnomad.sample_qc.relatedness.compute_related_samples_to_drop"]], "create_fake_pedigree() (in module gnomad.sample_qc.relatedness)": [[20, "gnomad.sample_qc.relatedness.create_fake_pedigree"]], "default_get_de_novo_expr() (in module gnomad.sample_qc.relatedness)": [[20, "gnomad.sample_qc.relatedness.default_get_de_novo_expr"]], "explode_duplicate_samples_ht() (in module gnomad.sample_qc.relatedness)": [[20, "gnomad.sample_qc.relatedness.explode_duplicate_samples_ht"]], "filter_to_trios() (in module gnomad.sample_qc.relatedness)": [[20, "gnomad.sample_qc.relatedness.filter_to_trios"]], "generate_sib_stats_expr() (in module gnomad.sample_qc.relatedness)": [[20, "gnomad.sample_qc.relatedness.generate_sib_stats_expr"]], "generate_trio_stats_expr() (in module gnomad.sample_qc.relatedness)": [[20, "gnomad.sample_qc.relatedness.generate_trio_stats_expr"]], "get_duplicated_samples() (in module gnomad.sample_qc.relatedness)": [[20, "gnomad.sample_qc.relatedness.get_duplicated_samples"]], "get_duplicated_samples_ht() (in module gnomad.sample_qc.relatedness)": [[20, "gnomad.sample_qc.relatedness.get_duplicated_samples_ht"]], "get_relationship_expr() (in module gnomad.sample_qc.relatedness)": [[20, "gnomad.sample_qc.relatedness.get_relationship_expr"]], "get_slope_int_relationship_expr() (in module gnomad.sample_qc.relatedness)": [[20, "gnomad.sample_qc.relatedness.get_slope_int_relationship_expr"]], "gnomad.sample_qc.relatedness": [[20, "module-gnomad.sample_qc.relatedness"]], "infer_families() (in module gnomad.sample_qc.relatedness)": [[20, "gnomad.sample_qc.relatedness.infer_families"]], "adjust_sex_ploidy() (in module gnomad.sample_qc.sex)": [[21, "gnomad.sample_qc.sex.adjust_sex_ploidy"]], "adjusted_sex_ploidy_expr() (in module gnomad.sample_qc.sex)": [[21, "gnomad.sample_qc.sex.adjusted_sex_ploidy_expr"]], "gaussian_mixture_model_karyotype_assignment() (in module gnomad.sample_qc.sex)": [[21, "gnomad.sample_qc.sex.gaussian_mixture_model_karyotype_assignment"]], "get_chr_x_hom_alt_cutoffs() (in module gnomad.sample_qc.sex)": [[21, "gnomad.sample_qc.sex.get_chr_x_hom_alt_cutoffs"]], "get_ploidy_cutoffs() (in module gnomad.sample_qc.sex)": [[21, "gnomad.sample_qc.sex.get_ploidy_cutoffs"]], "get_sex_expr() (in module gnomad.sample_qc.sex)": [[21, "gnomad.sample_qc.sex.get_sex_expr"]], "gnomad.sample_qc.sex": [[21, "module-gnomad.sample_qc.sex"]], "add_gks_va() (in module gnomad.utils.annotations)": [[22, "gnomad.utils.annotations.add_gks_va"]], "add_gks_vrs() (in module gnomad.utils.annotations)": [[22, "gnomad.utils.annotations.add_gks_vrs"]], "add_variant_type() (in module gnomad.utils.annotations)": [[22, "gnomad.utils.annotations.add_variant_type"]], "age_hists_expr() (in module gnomad.utils.annotations)": [[22, "gnomad.utils.annotations.age_hists_expr"]], "agg_by_strata() (in module gnomad.utils.annotations)": [[22, "gnomad.utils.annotations.agg_by_strata"]], "annotate_adj() (in module gnomad.utils.annotations)": [[22, "gnomad.utils.annotations.annotate_adj"]], "annotate_allele_info() (in module gnomad.utils.annotations)": [[22, "gnomad.utils.annotations.annotate_allele_info"]], "annotate_and_index_source_mt_for_sex_ploidy() (in module gnomad.utils.annotations)": [[22, "gnomad.utils.annotations.annotate_and_index_source_mt_for_sex_ploidy"]], "annotate_downsamplings() (in module gnomad.utils.annotations)": [[22, "gnomad.utils.annotations.annotate_downsamplings"]], "annotate_freq() (in module gnomad.utils.annotations)": [[22, "gnomad.utils.annotations.annotate_freq"]], "annotate_with_ht() (in module gnomad.utils.annotations)": [[22, "gnomad.utils.annotations.annotate_with_ht"]], "annotation_type_in_vcf_info() (in module gnomad.utils.annotations)": [[22, "gnomad.utils.annotations.annotation_type_in_vcf_info"]], "annotation_type_is_numeric() (in module gnomad.utils.annotations)": [[22, "gnomad.utils.annotations.annotation_type_is_numeric"]], "bi_allelic_expr() (in module gnomad.utils.annotations)": [[22, "gnomad.utils.annotations.bi_allelic_expr"]], "bi_allelic_site_inbreeding_expr() (in module gnomad.utils.annotations)": [[22, "gnomad.utils.annotations.bi_allelic_site_inbreeding_expr"]], "build_freq_stratification_list() (in module gnomad.utils.annotations)": [[22, "gnomad.utils.annotations.build_freq_stratification_list"]], "compute_freq_by_strata() (in module gnomad.utils.annotations)": [[22, "gnomad.utils.annotations.compute_freq_by_strata"]], "create_frequency_bins_expr() (in module gnomad.utils.annotations)": [[22, "gnomad.utils.annotations.create_frequency_bins_expr"]], "faf_expr() (in module gnomad.utils.annotations)": [[22, "gnomad.utils.annotations.faf_expr"]], "fill_missing_key_combinations() (in module gnomad.utils.annotations)": [[22, "gnomad.utils.annotations.fill_missing_key_combinations"]], "fs_from_sb() (in module gnomad.utils.annotations)": [[22, "gnomad.utils.annotations.fs_from_sb"]], "gen_anc_faf_max_expr() (in module gnomad.utils.annotations)": [[22, "gnomad.utils.annotations.gen_anc_faf_max_expr"]], "generate_freq_group_membership_array() (in module gnomad.utils.annotations)": [[22, "gnomad.utils.annotations.generate_freq_group_membership_array"]], "get_adj_expr() (in module gnomad.utils.annotations)": [[22, "gnomad.utils.annotations.get_adj_expr"]], "get_annotations_hists() (in module gnomad.utils.annotations)": [[22, "gnomad.utils.annotations.get_annotations_hists"]], "get_copy_state_by_sex() (in module gnomad.utils.annotations)": [[22, "gnomad.utils.annotations.get_copy_state_by_sex"]], "get_gq_dp_adj_expr() (in module gnomad.utils.annotations)": [[22, "gnomad.utils.annotations.get_gq_dp_adj_expr"]], "get_het_ab_adj_expr() (in module gnomad.utils.annotations)": [[22, "gnomad.utils.annotations.get_het_ab_adj_expr"]], "get_is_haploid_expr() (in module gnomad.utils.annotations)": [[22, "gnomad.utils.annotations.get_is_haploid_expr"]], "get_lowqual_expr() (in module gnomad.utils.annotations)": [[22, "gnomad.utils.annotations.get_lowqual_expr"]], "gnomad.utils.annotations": [[22, "module-gnomad.utils.annotations"]], "hemi_expr() (in module gnomad.utils.annotations)": [[22, "gnomad.utils.annotations.hemi_expr"]], "merge_freq_arrays() (in module gnomad.utils.annotations)": [[22, "gnomad.utils.annotations.merge_freq_arrays"]], "merge_histograms() (in module gnomad.utils.annotations)": [[22, "gnomad.utils.annotations.merge_histograms"]], "missing_callstats_expr() (in module gnomad.utils.annotations)": [[22, "gnomad.utils.annotations.missing_callstats_expr"]], "missing_struct_expr() (in module gnomad.utils.annotations)": [[22, "gnomad.utils.annotations.missing_struct_expr"]], "pab_max_expr() (in module gnomad.utils.annotations)": [[22, "gnomad.utils.annotations.pab_max_expr"]], "pop_max_expr() (in module gnomad.utils.annotations)": [[22, "gnomad.utils.annotations.pop_max_expr"]], "project_max_expr() (in module gnomad.utils.annotations)": [[22, "gnomad.utils.annotations.project_max_expr"]], "qual_hist_expr() (in module gnomad.utils.annotations)": [[22, "gnomad.utils.annotations.qual_hist_expr"]], "region_flag_expr() (in module gnomad.utils.annotations)": [[22, "gnomad.utils.annotations.region_flag_expr"]], "set_female_y_metrics_to_na_expr() (in module gnomad.utils.annotations)": [[22, "gnomad.utils.annotations.set_female_y_metrics_to_na_expr"]], "sor_from_sb() (in module gnomad.utils.annotations)": [[22, "gnomad.utils.annotations.sor_from_sb"]], "unphase_call_expr() (in module gnomad.utils.annotations)": [[22, "gnomad.utils.annotations.unphase_call_expr"]], "update_structured_annotations() (in module gnomad.utils.annotations)": [[22, "gnomad.utils.annotations.update_structured_annotations"]], "coverage_cutoff (in module gnomad.utils.constraint)": [[23, "gnomad.utils.constraint.COVERAGE_CUTOFF"]], "add_gencode_transcript_annotations() (in module gnomad.utils.constraint)": [[23, "gnomad.utils.constraint.add_gencode_transcript_annotations"]], "annotate_exploded_vep_for_constraint_groupings() (in module gnomad.utils.constraint)": [[23, "gnomad.utils.constraint.annotate_exploded_vep_for_constraint_groupings"]], "annotate_mutation_type() (in module gnomad.utils.constraint)": [[23, "gnomad.utils.constraint.annotate_mutation_type"]], "annotate_with_mu() (in module gnomad.utils.constraint)": [[23, "gnomad.utils.constraint.annotate_with_mu"]], "assemble_constraint_context_ht() (in module gnomad.utils.constraint)": [[23, "gnomad.utils.constraint.assemble_constraint_context_ht"]], "build_coverage_model() (in module gnomad.utils.constraint)": [[23, "gnomad.utils.constraint.build_coverage_model"]], "build_models() (in module gnomad.utils.constraint)": [[23, "gnomad.utils.constraint.build_models"]], "build_plateau_models() (in module gnomad.utils.constraint)": [[23, "gnomad.utils.constraint.build_plateau_models"]], "calculate_raw_z_score() (in module gnomad.utils.constraint)": [[23, "gnomad.utils.constraint.calculate_raw_z_score"]], "calculate_raw_z_score_sd() (in module gnomad.utils.constraint)": [[23, "gnomad.utils.constraint.calculate_raw_z_score_sd"]], "collapse_strand() (in module gnomad.utils.constraint)": [[23, "gnomad.utils.constraint.collapse_strand"]], "compute_expected_variants() (in module gnomad.utils.constraint)": [[23, "gnomad.utils.constraint.compute_expected_variants"]], "compute_pli() (in module gnomad.utils.constraint)": [[23, "gnomad.utils.constraint.compute_pli"]], "count_variants_by_group() (in module gnomad.utils.constraint)": [[23, "gnomad.utils.constraint.count_variants_by_group"]], "downsampling_counts_expr() (in module gnomad.utils.constraint)": [[23, "gnomad.utils.constraint.downsampling_counts_expr"]], "explode_downsamplings_oe() (in module gnomad.utils.constraint)": [[23, "gnomad.utils.constraint.explode_downsamplings_oe"]], "get_all_pop_lengths() (in module gnomad.utils.constraint)": [[23, "gnomad.utils.constraint.get_all_pop_lengths"]], "get_constraint_flags() (in module gnomad.utils.constraint)": [[23, "gnomad.utils.constraint.get_constraint_flags"]], "get_constraint_grouping_expr() (in module gnomad.utils.constraint)": [[23, "gnomad.utils.constraint.get_constraint_grouping_expr"]], "get_downsampling_freq_indices() (in module gnomad.utils.constraint)": [[23, "gnomad.utils.constraint.get_downsampling_freq_indices"]], "get_mu_annotation_expr() (in module gnomad.utils.constraint)": [[23, "gnomad.utils.constraint.get_mu_annotation_expr"]], "gnomad.utils.constraint": [[23, "module-gnomad.utils.constraint"]], "oe_aggregation_expr() (in module gnomad.utils.constraint)": [[23, "gnomad.utils.constraint.oe_aggregation_expr"]], "oe_confidence_interval() (in module gnomad.utils.constraint)": [[23, "gnomad.utils.constraint.oe_confidence_interval"]], "transform_methylation_level() (in module gnomad.utils.constraint)": [[23, "gnomad.utils.constraint.transform_methylation_level"]], "trimer_from_heptamer() (in module gnomad.utils.constraint)": [[23, "gnomad.utils.constraint.trimer_from_heptamer"]], "check_file_exists_raise_error() (in module gnomad.utils.file_utils)": [[24, "gnomad.utils.file_utils.check_file_exists_raise_error"]], "create_vds() (in module gnomad.utils.file_utils)": [[24, "gnomad.utils.file_utils.create_vds"]], "file_exists() (in module gnomad.utils.file_utils)": [[24, "gnomad.utils.file_utils.file_exists"]], "get_file_stats() (in module gnomad.utils.file_utils)": [[24, "gnomad.utils.file_utils.get_file_stats"]], "gnomad.utils.file_utils": [[24, "module-gnomad.utils.file_utils"]], "read_list_data() (in module gnomad.utils.file_utils)": [[24, "gnomad.utils.file_utils.read_list_data"]], "repartition_for_join() (in module gnomad.utils.file_utils)": [[24, "gnomad.utils.file_utils.repartition_for_join"]], "select_primitives_from_ht() (in module gnomad.utils.file_utils)": [[24, "gnomad.utils.file_utils.select_primitives_from_ht"]], "write_temp_gcs() (in module gnomad.utils.file_utils)": [[24, "gnomad.utils.file_utils.write_temp_gcs"]], "add_filters_expr() (in module gnomad.utils.filtering)": [[25, "gnomad.utils.filtering.add_filters_expr"]], "combine_functions() (in module gnomad.utils.filtering)": [[25, "gnomad.utils.filtering.combine_functions"]], "filter_arrays_by_meta() (in module gnomad.utils.filtering)": [[25, "gnomad.utils.filtering.filter_arrays_by_meta"]], "filter_by_frequency() (in module gnomad.utils.filtering)": [[25, "gnomad.utils.filtering.filter_by_frequency"]], "filter_by_gencode_intervals() (in module gnomad.utils.filtering)": [[25, "gnomad.utils.filtering.filter_by_gencode_intervals"]], "filter_by_intervals() (in module gnomad.utils.filtering)": [[25, "gnomad.utils.filtering.filter_by_intervals"]], "filter_by_numeric_expr_range() (in module gnomad.utils.filtering)": [[25, "gnomad.utils.filtering.filter_by_numeric_expr_range"]], "filter_for_mu() (in module gnomad.utils.filtering)": [[25, "gnomad.utils.filtering.filter_for_mu"]], "filter_gencode_ht() (in module gnomad.utils.filtering)": [[25, "gnomad.utils.filtering.filter_gencode_ht"]], "filter_low_conf_regions() (in module gnomad.utils.filtering)": [[25, "gnomad.utils.filtering.filter_low_conf_regions"]], "filter_to_adj() (in module gnomad.utils.filtering)": [[25, "gnomad.utils.filtering.filter_to_adj"]], "filter_to_autosomes() (in module gnomad.utils.filtering)": [[25, "gnomad.utils.filtering.filter_to_autosomes"]], "filter_to_clinvar_pathogenic() (in module gnomad.utils.filtering)": [[25, "gnomad.utils.filtering.filter_to_clinvar_pathogenic"]], "filter_to_gencode_cds() (in module gnomad.utils.filtering)": [[25, "gnomad.utils.filtering.filter_to_gencode_cds"]], "filter_x_nonpar() (in module gnomad.utils.filtering)": [[25, "gnomad.utils.filtering.filter_x_nonpar"]], "filter_y_nonpar() (in module gnomad.utils.filtering)": [[25, "gnomad.utils.filtering.filter_y_nonpar"]], "gnomad.utils.filtering": [[25, "module-gnomad.utils.filtering"]], "low_conf_regions_expr() (in module gnomad.utils.filtering)": [[25, "gnomad.utils.filtering.low_conf_regions_expr"]], "remove_fields_from_constant() (in module gnomad.utils.filtering)": [[25, "gnomad.utils.filtering.remove_fields_from_constant"]], "split_vds_by_strata() (in module gnomad.utils.filtering)": [[25, "gnomad.utils.filtering.split_vds_by_strata"]], "subset_samples_and_variants() (in module gnomad.utils.filtering)": [[25, "gnomad.utils.filtering.subset_samples_and_variants"]], "from_phred() (in module gnomad.utils.gen_stats)": [[26, "gnomad.utils.gen_stats.from_phred"]], "get_median_and_mad_expr() (in module gnomad.utils.gen_stats)": [[26, "gnomad.utils.gen_stats.get_median_and_mad_expr"]], "gnomad.utils.gen_stats": [[26, "module-gnomad.utils.gen_stats"]], "merge_stats_counters_expr() (in module gnomad.utils.gen_stats)": [[26, "gnomad.utils.gen_stats.merge_stats_counters_expr"]], "to_phred() (in module gnomad.utils.gen_stats)": [[26, "gnomad.utils.gen_stats.to_phred"]], "gnomad.utils.intervals": [[28, "module-gnomad.utils.intervals"]], "interval_length() (in module gnomad.utils.intervals)": [[28, "gnomad.utils.intervals.interval_length"]], "pad_intervals() (in module gnomad.utils.intervals)": [[28, "gnomad.utils.intervals.pad_intervals"]], "sort_intervals() (in module gnomad.utils.intervals)": [[28, "gnomad.utils.intervals.sort_intervals"]], "union_intervals() (in module gnomad.utils.intervals)": [[28, "gnomad.utils.intervals.union_intervals"]], "grch37_to_grch38_chain (in module gnomad.utils.liftover)": [[29, "gnomad.utils.liftover.GRCH37_to_GRCH38_CHAIN"]], "grch38_to_grch37_chain (in module gnomad.utils.liftover)": [[29, "gnomad.utils.liftover.GRCH38_TO_GRCH37_CHAIN"]], "default_lift_data() (in module gnomad.utils.liftover)": [[29, "gnomad.utils.liftover.default_lift_data"]], "get_liftover_genome() (in module gnomad.utils.liftover)": [[29, "gnomad.utils.liftover.get_liftover_genome"]], "gnomad.utils.liftover": [[29, "module-gnomad.utils.liftover"]], "liftover_expr() (in module gnomad.utils.liftover)": [[29, "gnomad.utils.liftover.liftover_expr"]], "liftover_using_gnomad_map() (in module gnomad.utils.liftover)": [[29, "gnomad.utils.liftover.liftover_using_gnomad_map"]], "gnomad.utils.parse": [[30, "module-gnomad.utils.parse"]], "parse_locus_intervals() (in module gnomad.utils.parse)": [[30, "gnomad.utils.parse.parse_locus_intervals"]], "parse_variant() (in module gnomad.utils.parse)": [[30, "gnomad.utils.parse.parse_variant"]], "get_rows_data() (in module gnomad.utils.plotting)": [[31, "gnomad.utils.plotting.get_rows_data"]], "gnomad.utils.plotting": [[31, "module-gnomad.utils.plotting"]], "linear_and_log_tabs() (in module gnomad.utils.plotting)": [[31, "gnomad.utils.plotting.linear_and_log_tabs"]], "new_show() (in module gnomad.utils.plotting)": [[31, "gnomad.utils.plotting.new_show"]], "pair_plot() (in module gnomad.utils.plotting)": [[31, "gnomad.utils.plotting.pair_plot"]], "plot_hail_file_metadata() (in module gnomad.utils.plotting)": [[31, "gnomad.utils.plotting.plot_hail_file_metadata"]], "plot_hail_hist() (in module gnomad.utils.plotting)": [[31, "gnomad.utils.plotting.plot_hail_hist"]], "plot_hail_hist_both() (in module gnomad.utils.plotting)": [[31, "gnomad.utils.plotting.plot_hail_hist_both"]], "plot_hail_hist_cumulative() (in module gnomad.utils.plotting)": [[31, "gnomad.utils.plotting.plot_hail_hist_cumulative"]], "plot_multi_hail_hist() (in module gnomad.utils.plotting)": [[31, "gnomad.utils.plotting.plot_multi_hail_hist"]], "scale_file_sizes() (in module gnomad.utils.plotting)": [[31, "gnomad.utils.plotting.scale_file_sizes"]], "set_font_size() (in module gnomad.utils.plotting)": [[31, "gnomad.utils.plotting.set_font_size"]], "add_reference_sequence() (in module gnomad.utils.reference_genome)": [[32, "gnomad.utils.reference_genome.add_reference_sequence"]], "get_reference_genome() (in module gnomad.utils.reference_genome)": [[32, "gnomad.utils.reference_genome.get_reference_genome"]], "get_reference_ht() (in module gnomad.utils.reference_genome)": [[32, "gnomad.utils.reference_genome.get_reference_ht"]], "gnomad.utils.reference_genome": [[32, "module-gnomad.utils.reference_genome"]], "gnomad.utils.release": [[33, "module-gnomad.utils.release"]], "make_faf_index_dict() (in module gnomad.utils.release)": [[33, "gnomad.utils.release.make_faf_index_dict"]], "make_freq_index_dict() (in module gnomad.utils.release)": [[33, "gnomad.utils.release.make_freq_index_dict"]], "make_freq_index_dict_from_meta() (in module gnomad.utils.release)": [[33, "gnomad.utils.release.make_freq_index_dict_from_meta"]], "slackclient (class in gnomad.utils.slack)": [[34, "gnomad.utils.slack.SlackClient"]], "gnomad.utils.slack": [[34, "module-gnomad.utils.slack"]], "send_file() (gnomad.utils.slack.slackclient method)": [[34, "gnomad.utils.slack.SlackClient.send_file"]], "send_message() (gnomad.utils.slack.slackclient method)": [[34, "gnomad.utils.slack.SlackClient.send_message"]], "slack_notifications() (in module gnomad.utils.slack)": [[34, "gnomad.utils.slack.slack_notifications"]], "compute_allele_number_per_ref_site() (in module gnomad.utils.sparse_mt)": [[35, "gnomad.utils.sparse_mt.compute_allele_number_per_ref_site"]], "compute_coverage_stats() (in module gnomad.utils.sparse_mt)": [[35, "gnomad.utils.sparse_mt.compute_coverage_stats"]], "compute_last_ref_block_end() (in module gnomad.utils.sparse_mt)": [[35, "gnomad.utils.sparse_mt.compute_last_ref_block_end"]], "compute_stats_per_ref_site() (in module gnomad.utils.sparse_mt)": [[35, "gnomad.utils.sparse_mt.compute_stats_per_ref_site"]], "default_compute_info() (in module gnomad.utils.sparse_mt)": [[35, "gnomad.utils.sparse_mt.default_compute_info"]], "densify_all_reference_sites() (in module gnomad.utils.sparse_mt)": [[35, "gnomad.utils.sparse_mt.densify_all_reference_sites"]], "densify_sites() (in module gnomad.utils.sparse_mt)": [[35, "gnomad.utils.sparse_mt.densify_sites"]], "filter_ref_blocks() (in module gnomad.utils.sparse_mt)": [[35, "gnomad.utils.sparse_mt.filter_ref_blocks"]], "get_allele_number_agg_func() (in module gnomad.utils.sparse_mt)": [[35, "gnomad.utils.sparse_mt.get_allele_number_agg_func"]], "get_as_info_expr() (in module gnomad.utils.sparse_mt)": [[35, "gnomad.utils.sparse_mt.get_as_info_expr"]], "get_site_info_expr() (in module gnomad.utils.sparse_mt)": [[35, "gnomad.utils.sparse_mt.get_site_info_expr"]], "gnomad.utils.sparse_mt": [[35, "module-gnomad.utils.sparse_mt"]], "impute_sex_ploidy() (in module gnomad.utils.sparse_mt)": [[35, "gnomad.utils.sparse_mt.impute_sex_ploidy"]], "split_info_annotation() (in module gnomad.utils.sparse_mt)": [[35, "gnomad.utils.sparse_mt.split_info_annotation"]], "split_lowqual_annotation() (in module gnomad.utils.sparse_mt)": [[35, "gnomad.utils.sparse_mt.split_lowqual_annotation"]], "cell_lines (in module gnomad.utils.transcript_annotation)": [[36, "gnomad.utils.transcript_annotation.CELL_LINES"]], "reproductive_tissues (in module gnomad.utils.transcript_annotation)": [[36, "gnomad.utils.transcript_annotation.REPRODUCTIVE_TISSUES"]], "clean_tissue_name_for_browser() (in module gnomad.utils.transcript_annotation)": [[36, "gnomad.utils.transcript_annotation.clean_tissue_name_for_browser"]], "create_tx_annotation_by_region() (in module gnomad.utils.transcript_annotation)": [[36, "gnomad.utils.transcript_annotation.create_tx_annotation_by_region"]], "filter_expression_ht_by_tissues() (in module gnomad.utils.transcript_annotation)": [[36, "gnomad.utils.transcript_annotation.filter_expression_ht_by_tissues"]], "get_expression_proportion() (in module gnomad.utils.transcript_annotation)": [[36, "gnomad.utils.transcript_annotation.get_expression_proportion"]], "get_tissues_to_exclude() (in module gnomad.utils.transcript_annotation)": [[36, "gnomad.utils.transcript_annotation.get_tissues_to_exclude"]], "gnomad.utils.transcript_annotation": [[36, "module-gnomad.utils.transcript_annotation"]], "perform_tx_annotation_pipeline() (in module gnomad.utils.transcript_annotation)": [[36, "gnomad.utils.transcript_annotation.perform_tx_annotation_pipeline"]], "summarize_transcript_expression() (in module gnomad.utils.transcript_annotation)": [[36, "gnomad.utils.transcript_annotation.summarize_transcript_expression"]], "tissue_expression_ht_to_array() (in module gnomad.utils.transcript_annotation)": [[36, "gnomad.utils.transcript_annotation.tissue_expression_ht_to_array"]], "tx_aggregate_variants() (in module gnomad.utils.transcript_annotation)": [[36, "gnomad.utils.transcript_annotation.tx_aggregate_variants"]], "tx_annotate_variants() (in module gnomad.utils.transcript_annotation)": [[36, "gnomad.utils.transcript_annotation.tx_annotate_variants"]], "tx_filter_variants_by_csqs() (in module gnomad.utils.transcript_annotation)": [[36, "gnomad.utils.transcript_annotation.tx_filter_variants_by_csqs"]], "allele_type_fields (in module gnomad.utils.vcf)": [[37, "gnomad.utils.vcf.ALLELE_TYPE_FIELDS"]], "as_fields (in module gnomad.utils.vcf)": [[37, "gnomad.utils.vcf.AS_FIELDS"]], "as_vqsr_fields (in module gnomad.utils.vcf)": [[37, "gnomad.utils.vcf.AS_VQSR_FIELDS"]], "entries (in module gnomad.utils.vcf)": [[37, "gnomad.utils.vcf.ENTRIES"]], "faf_pops (in module gnomad.utils.vcf)": [[37, "gnomad.utils.vcf.FAF_POPS"]], "format_dict (in module gnomad.utils.vcf)": [[37, "gnomad.utils.vcf.FORMAT_DICT"]], "groups (in module gnomad.utils.vcf)": [[37, "gnomad.utils.vcf.GROUPS"]], "hists (in module gnomad.utils.vcf)": [[37, "gnomad.utils.vcf.HISTS"]], "info_dict (in module gnomad.utils.vcf)": [[37, "gnomad.utils.vcf.INFO_DICT"]], "in_silico_annotations_info_dict (in module gnomad.utils.vcf)": [[37, "gnomad.utils.vcf.IN_SILICO_ANNOTATIONS_INFO_DICT"]], "joint_region_flag_fields (in module gnomad.utils.vcf)": [[37, "gnomad.utils.vcf.JOINT_REGION_FLAG_FIELDS"]], "region_flag_fields (in module gnomad.utils.vcf)": [[37, "gnomad.utils.vcf.REGION_FLAG_FIELDS"]], "rf_fields (in module gnomad.utils.vcf)": [[37, "gnomad.utils.vcf.RF_FIELDS"]], "sexes (in module gnomad.utils.vcf)": [[37, "gnomad.utils.vcf.SEXES"]], "site_fields (in module gnomad.utils.vcf)": [[37, "gnomad.utils.vcf.SITE_FIELDS"]], "sort_order (in module gnomad.utils.vcf)": [[37, "gnomad.utils.vcf.SORT_ORDER"]], "sparse_entries (in module gnomad.utils.vcf)": [[37, "gnomad.utils.vcf.SPARSE_ENTRIES"]], "vqsr_fields (in module gnomad.utils.vcf)": [[37, "gnomad.utils.vcf.VQSR_FIELDS"]], "vrs_fields_dict (in module gnomad.utils.vcf)": [[37, "gnomad.utils.vcf.VRS_FIELDS_DICT"]], "add_as_info_dict() (in module gnomad.utils.vcf)": [[37, "gnomad.utils.vcf.add_as_info_dict"]], "adjust_vcf_incompatible_types() (in module gnomad.utils.vcf)": [[37, "gnomad.utils.vcf.adjust_vcf_incompatible_types"]], "build_vcf_export_reference() (in module gnomad.utils.vcf)": [[37, "gnomad.utils.vcf.build_vcf_export_reference"]], "create_label_groups() (in module gnomad.utils.vcf)": [[37, "gnomad.utils.vcf.create_label_groups"]], "gnomad.utils.vcf": [[37, "module-gnomad.utils.vcf"]], "index_globals() (in module gnomad.utils.vcf)": [[37, "gnomad.utils.vcf.index_globals"]], "make_combo_header_text() (in module gnomad.utils.vcf)": [[37, "gnomad.utils.vcf.make_combo_header_text"]], "make_hist_bin_edges_expr() (in module gnomad.utils.vcf)": [[37, "gnomad.utils.vcf.make_hist_bin_edges_expr"]], "make_hist_dict() (in module gnomad.utils.vcf)": [[37, "gnomad.utils.vcf.make_hist_dict"]], "make_info_dict() (in module gnomad.utils.vcf)": [[37, "gnomad.utils.vcf.make_info_dict"]], "make_label_combos() (in module gnomad.utils.vcf)": [[37, "gnomad.utils.vcf.make_label_combos"]], "make_vcf_filter_dict() (in module gnomad.utils.vcf)": [[37, "gnomad.utils.vcf.make_vcf_filter_dict"]], "rekey_new_reference() (in module gnomad.utils.vcf)": [[37, "gnomad.utils.vcf.rekey_new_reference"]], "set_female_y_metrics_to_na() (in module gnomad.utils.vcf)": [[37, "gnomad.utils.vcf.set_female_y_metrics_to_na"]], "csq_coding (in module gnomad.utils.vep)": [[38, "gnomad.utils.vep.CSQ_CODING"]], "csq_splice (in module gnomad.utils.vep)": [[38, "gnomad.utils.vep.CSQ_SPLICE"]], "current_vep_version (in module gnomad.utils.vep)": [[38, "gnomad.utils.vep.CURRENT_VEP_VERSION"]], "loftee_labels (in module gnomad.utils.vep)": [[38, "gnomad.utils.vep.LOFTEE_LABELS"]], "lof_csq_set (in module gnomad.utils.vep)": [[38, "gnomad.utils.vep.LOF_CSQ_SET"]], "possible_refs (in module gnomad.utils.vep)": [[38, "gnomad.utils.vep.POSSIBLE_REFS"]], "vep_config_path (in module gnomad.utils.vep)": [[38, "gnomad.utils.vep.VEP_CONFIG_PATH"]], "vep_csq_fields (in module gnomad.utils.vep)": [[38, "gnomad.utils.vep.VEP_CSQ_FIELDS"]], "vep_csq_header (in module gnomad.utils.vep)": [[38, "gnomad.utils.vep.VEP_CSQ_HEADER"]], "add_most_severe_consequence_to_consequence() (in module gnomad.utils.vep)": [[38, "gnomad.utils.vep.add_most_severe_consequence_to_consequence"]], "add_most_severe_csq_to_tc_within_vep_root() (in module gnomad.utils.vep)": [[38, "gnomad.utils.vep.add_most_severe_csq_to_tc_within_vep_root"]], "explode_by_vep_annotation() (in module gnomad.utils.vep)": [[38, "gnomad.utils.vep.explode_by_vep_annotation"]], "filter_vep_to_canonical_transcripts() (in module gnomad.utils.vep)": [[38, "gnomad.utils.vep.filter_vep_to_canonical_transcripts"]], "filter_vep_to_gene_list() (in module gnomad.utils.vep)": [[38, "gnomad.utils.vep.filter_vep_to_gene_list"]], "filter_vep_to_mane_select_transcripts() (in module gnomad.utils.vep)": [[38, "gnomad.utils.vep.filter_vep_to_mane_select_transcripts"]], "filter_vep_to_synonymous_variants() (in module gnomad.utils.vep)": [[38, "gnomad.utils.vep.filter_vep_to_synonymous_variants"]], "filter_vep_transcript_csqs() (in module gnomad.utils.vep)": [[38, "gnomad.utils.vep.filter_vep_transcript_csqs"]], "filter_vep_transcript_csqs_expr() (in module gnomad.utils.vep)": [[38, "gnomad.utils.vep.filter_vep_transcript_csqs_expr"]], "get_most_severe_consequence_expr() (in module gnomad.utils.vep)": [[38, "gnomad.utils.vep.get_most_severe_consequence_expr"]], "get_most_severe_consequence_for_summary() (in module gnomad.utils.vep)": [[38, "gnomad.utils.vep.get_most_severe_consequence_for_summary"]], "get_vep_context() (in module gnomad.utils.vep)": [[38, "gnomad.utils.vep.get_vep_context"]], "get_vep_help() (in module gnomad.utils.vep)": [[38, "gnomad.utils.vep.get_vep_help"]], "gnomad.utils.vep": [[38, "module-gnomad.utils.vep"]], "process_consequences() (in module gnomad.utils.vep)": [[38, "gnomad.utils.vep.process_consequences"]], "vep_or_lookup_vep() (in module gnomad.utils.vep)": [[38, "gnomad.utils.vep.vep_or_lookup_vep"]], "vep_struct_to_csq() (in module gnomad.utils.vep)": [[38, "gnomad.utils.vep.vep_struct_to_csq"]], "add_rank() (in module gnomad.variant_qc.evaluation)": [[39, "gnomad.variant_qc.evaluation.add_rank"]], "compute_binned_truth_sample_concordance() (in module gnomad.variant_qc.evaluation)": [[39, "gnomad.variant_qc.evaluation.compute_binned_truth_sample_concordance"]], "compute_grouped_binned_ht() (in module gnomad.variant_qc.evaluation)": [[39, "gnomad.variant_qc.evaluation.compute_grouped_binned_ht"]], "compute_ranked_bin() (in module gnomad.variant_qc.evaluation)": [[39, "gnomad.variant_qc.evaluation.compute_ranked_bin"]], "create_truth_sample_ht() (in module gnomad.variant_qc.evaluation)": [[39, "gnomad.variant_qc.evaluation.create_truth_sample_ht"]], "gnomad.variant_qc.evaluation": [[39, "module-gnomad.variant_qc.evaluation"]], "get_r_for_pair_of_variants() (in module gnomad.variant_qc.ld)": [[41, "gnomad.variant_qc.ld.get_r_for_pair_of_variants"]], "get_r_human_readable() (in module gnomad.variant_qc.ld)": [[41, "gnomad.variant_qc.ld.get_r_human_readable"]], "get_r_within_gene() (in module gnomad.variant_qc.ld)": [[41, "gnomad.variant_qc.ld.get_r_within_gene"]], "get_r_within_gene_in_pop() (in module gnomad.variant_qc.ld)": [[41, "gnomad.variant_qc.ld.get_r_within_gene_in_pop"]], "gnomad.variant_qc.ld": [[41, "module-gnomad.variant_qc.ld"]], "create_binned_ht() (in module gnomad.variant_qc.pipeline)": [[42, "gnomad.variant_qc.pipeline.create_binned_ht"]], "generate_sib_stats() (in module gnomad.variant_qc.pipeline)": [[42, "gnomad.variant_qc.pipeline.generate_sib_stats"]], "generate_trio_stats() (in module gnomad.variant_qc.pipeline)": [[42, "gnomad.variant_qc.pipeline.generate_trio_stats"]], "gnomad.variant_qc.pipeline": [[42, "module-gnomad.variant_qc.pipeline"]], "score_bin_agg() (in module gnomad.variant_qc.pipeline)": [[42, "gnomad.variant_qc.pipeline.score_bin_agg"]], "train_rf_model() (in module gnomad.variant_qc.pipeline)": [[42, "gnomad.variant_qc.pipeline.train_rf_model"]], "apply_rf_model() (in module gnomad.variant_qc.random_forest)": [[43, "gnomad.variant_qc.random_forest.apply_rf_model"]], "check_ht_fields_for_spark() (in module gnomad.variant_qc.random_forest)": [[43, "gnomad.variant_qc.random_forest.check_ht_fields_for_spark"]], "get_columns_quantiles() (in module gnomad.variant_qc.random_forest)": [[43, "gnomad.variant_qc.random_forest.get_columns_quantiles"]], "get_features_importance() (in module gnomad.variant_qc.random_forest)": [[43, "gnomad.variant_qc.random_forest.get_features_importance"]], "get_labels() (in module gnomad.variant_qc.random_forest)": [[43, "gnomad.variant_qc.random_forest.get_labels"]], "get_rf_runs() (in module gnomad.variant_qc.random_forest)": [[43, "gnomad.variant_qc.random_forest.get_rf_runs"]], "get_run_data() (in module gnomad.variant_qc.random_forest)": [[43, "gnomad.variant_qc.random_forest.get_run_data"]], "gnomad.variant_qc.random_forest": [[43, "module-gnomad.variant_qc.random_forest"]], "ht_to_rf_df() (in module gnomad.variant_qc.random_forest)": [[43, "gnomad.variant_qc.random_forest.ht_to_rf_df"]], "load_model() (in module gnomad.variant_qc.random_forest)": [[43, "gnomad.variant_qc.random_forest.load_model"]], "median_impute_features() (in module gnomad.variant_qc.random_forest)": [[43, "gnomad.variant_qc.random_forest.median_impute_features"]], "pretty_print_runs() (in module gnomad.variant_qc.random_forest)": [[43, "gnomad.variant_qc.random_forest.pretty_print_runs"]], "run_rf_test() (in module gnomad.variant_qc.random_forest)": [[43, "gnomad.variant_qc.random_forest.run_rf_test"]], "save_model() (in module gnomad.variant_qc.random_forest)": [[43, "gnomad.variant_qc.random_forest.save_model"]], "test_model() (in module gnomad.variant_qc.random_forest)": [[43, "gnomad.variant_qc.random_forest.test_model"]], "train_rf() (in module gnomad.variant_qc.random_forest)": [[43, "gnomad.variant_qc.random_forest.train_rf"]], "gnomad.variant_qc.training": [[44, "module-gnomad.variant_qc.training"]], "sample_training_examples() (in module gnomad.variant_qc.training)": [[44, "gnomad.variant_qc.training.sample_training_examples"]]}})